// @aws-amplify/graphql-api-construct/lib/index.js
"use strict";
// ############################################################################
// Note that sections of this file are excluded from test coverage metrics with
// 'c8 ignore' comments. If you add code to this file, suppress non-executable
// code from coverage metrics by using a `c8 ignore`. If the code is
// executable, you MUST NOT suppress it from coverage metrics.
//
// `ignore` statements must only span a single block of code. Do not apply an
// `ignore` statement to multiple blocks, or the entire file.
// ############################################################################
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RetentionDays = exports.FieldLogLevel = exports.SQLLambdaModelDataSourceStrategyFactory = exports.SSEType = exports.AmplifyDynamoDbTableWrapper = exports.AmplifyGraphqlDefinition = exports.AmplifyGraphqlApi = void 0;
/* c8 ignore stop */
/* c8 ignore start */
var amplify_graphql_api_1 = require("./amplify-graphql-api");
Object.defineProperty(exports, "AmplifyGraphqlApi", { enumerable: true, get: function () { return amplify_graphql_api_1.AmplifyGraphqlApi; } });
/* c8 ignore stop */
/* c8 ignore start */
var amplify_graphql_definition_1 = require("./amplify-graphql-definition");
Object.defineProperty(exports, "AmplifyGraphqlDefinition", { enumerable: true, get: function () { return amplify_graphql_definition_1.AmplifyGraphqlDefinition; } });
/* c8 ignore stop */
/* c8 ignore start */
var amplify_dynamodb_table_wrapper_1 = require("./amplify-dynamodb-table-wrapper");
Object.defineProperty(exports, "AmplifyDynamoDbTableWrapper", { enumerable: true, get: function () { return amplify_dynamodb_table_wrapper_1.AmplifyDynamoDbTableWrapper; } });
Object.defineProperty(exports, "SSEType", { enumerable: true, get: function () { return amplify_dynamodb_table_wrapper_1.SSEType; } });
/* c8 ignore stop */
/* c8 ignore start */
var sql_model_datasource_strategy_1 = require("./sql-model-datasource-strategy");
Object.defineProperty(exports, "SQLLambdaModelDataSourceStrategyFactory", { enumerable: true, get: function () { return sql_model_datasource_strategy_1.SQLLambdaModelDataSourceStrategyFactory; } });
/* c8 ignore stop */
/* c8 ignore start */
__exportStar(require("./model-datasource-strategy-types"), exports);
/* c8 ignore stop */
/* c8 ignore start */
var log_config_types_1 = require("./log-config-types");
Object.defineProperty(exports, "FieldLogLevel", { enumerable: true, get: function () { return log_config_types_1.FieldLogLevel; } });
Object.defineProperty(exports, "RetentionDays", { enumerable: true, get: function () { return log_config_types_1.RetentionDays; } });
/* c8 ignore stop */
