// @aws-amplify/cli-core/lib/package-manager-controller/npm_package_manager_controller.js
import { existsSync as _existsSync } from 'fs';
import _fsp from 'fs/promises';
import { execa as _execa } from 'execa';
import * as _path from 'path';
import { executeWithDebugLogger as _executeWithDebugLogger } from './execute_with_debugger_logger.js';
import { PackageManagerControllerBase } from './package_manager_controller_base.js';
import { NpmLockFileReader } from './lock-file-reader/npm_lock_file_reader.js';
/**
 * NpmPackageManagerController is an abstraction around npm commands that are needed to initialize a project and install dependencies
 */
export class NpmPackageManagerController extends PackageManagerControllerBase {
    cwd;
    fsp;
    path;
    execa;
    executeWithDebugLogger;
    existsSync;
    lockFileReader;
    /**
     * constructor
     */
    constructor(cwd, fsp = _fsp, path = _path, execa = _execa, executeWithDebugLogger = _executeWithDebugLogger, existsSync = _existsSync, lockFileReader = new NpmLockFileReader()) {
        super(cwd, 'npm', ['init', '--yes'], 'install', lockFileReader, fsp, path, execa, executeWithDebugLogger, existsSync);
        this.cwd = cwd;
        this.fsp = fsp;
        this.path = path;
        this.execa = execa;
        this.executeWithDebugLogger = executeWithDebugLogger;
        this.existsSync = existsSync;
        this.lockFileReader = lockFileReader;
    }
}
