// @aws-amplify/cli-core/lib/package-manager-controller/yarn_modern_package_manager_controller.js
import { existsSync as _existsSync } from 'fs';
import _fsp from 'fs/promises';
import { execa as _execa } from 'execa';
import * as _path from 'path';
import { LogLevel } from '../printer/printer.js';
import { format } from '../format/format.js';
import { executeWithDebugLogger as _executeWithDebugLogger } from './execute_with_debugger_logger.js';
import { PackageManagerControllerBase } from './package_manager_controller_base.js';
import { YarnModernLockFileReader } from './lock-file-reader/yarn_modern_lock_file_reader.js';
/**
 * YarnModernPackageManagerController is an abstraction around yarn modern (yarn v2+) commands that are needed to initialize a project and install dependencies
 */
export class YarnModernPackageManagerController extends PackageManagerControllerBase {
    cwd;
    printer;
    fsp;
    path;
    execa;
    executeWithDebugLogger;
    existsSync;
    lockFileReader;
    /**
     * constructor
     */
    constructor(cwd, printer, fsp = _fsp, path = _path, execa = _execa, executeWithDebugLogger = _executeWithDebugLogger, existsSync = _existsSync, lockFileReader = new YarnModernLockFileReader()) {
        super(cwd, 'yarn', ['init', '--yes'], 'add', lockFileReader, fsp, path, execa, executeWithDebugLogger, existsSync);
        this.cwd = cwd;
        this.printer = printer;
        this.fsp = fsp;
        this.path = path;
        this.execa = execa;
        this.executeWithDebugLogger = executeWithDebugLogger;
        this.existsSync = existsSync;
        this.lockFileReader = lockFileReader;
    }
    initializeTsConfig = async (targetDir) => {
        await this.addLockFile(targetDir);
        await this.addTypescript(targetDir);
        await super.initializeTsConfig(targetDir);
    };
    /**
     * addLockFile - adds a yarn.lock file to the project root for yarn v2+
     */
    addLockFile = async (targetDir) => {
        try {
            await this.fsp.writeFile(this.path.resolve(targetDir, 'yarn.lock'), '');
        }
        catch (error) {
            if (typeof error === 'string') {
                this.printer.log(format.error(`Error creating ${targetDir}/yarn.lock ${error}}`), LogLevel.ERROR);
            }
            else if (error instanceof Error) {
                this.printer.log(format.error(`Error creating ${targetDir}/yarn.lock ${error.message}}`), LogLevel.ERROR);
            }
        }
    };
    addTypescript = async (targetDir) => {
        await this.executeWithDebugLogger(targetDir, 'yarn', ['add', 'typescript@^5'], this.execa);
    };
}
