// @aws-cdk/toolkit-lib/lib/actions/refactor/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MappingSource = void 0;
const refactoring_1 = require("../../api/refactoring");
/**
 * The source of the resource mappings to be used for refactoring.
 */
class MappingSource {
    /**
     * The mapping will be automatically generated based on a comparison of
     * the deployed stacks and the local stacks.
     *
     * @param exclude - A list of resource locations to exclude from the mapping.
     */
    static auto(exclude = []) {
        const excludeList = new refactoring_1.InMemoryExcludeList(exclude);
        return new MappingSource('auto', [], excludeList);
    }
    /**
     * An explicitly provided list of mappings, which will be used for refactoring.
     */
    static explicit(groups) {
        return new MappingSource('explicit', groups, new refactoring_1.NeverExclude());
    }
    /**
     * An explicitly provided list of mappings, which will be used for refactoring,
     * but in reverse, that is, the source locations will become the destination
     * locations and vice versa.
     */
    static reverse(groups) {
        const reverseGroups = groups.map((group) => ({
            ...group,
            resources: Object.fromEntries(Object.entries(group.resources).map(([src, dst]) => [dst, src])),
        }));
        return MappingSource.explicit(reverseGroups);
    }
    /**
     * @internal
     */
    source;
    /**
     * @internal
     */
    groups;
    /**
     * @internal
     */
    exclude;
    constructor(source, groups, exclude) {
        this.source = source;
        this.groups = groups;
        this.exclude = exclude;
    }
}
exports.MappingSource = MappingSource;
