// @aws-amplify/backend-cli/lib/notices/notices_manifest_fetcher.js
import { LogLevel, NoticesManifestValidator, noticesManifestSchema, printer, } from '@aws-amplify/cli-core';
import { noticesMetadataFileInstance } from './notices_files.js';
import { AmplifyFault } from '@aws-amplify/platform-core';
/**
 * Notices manifest fetcher.
 */
export class NoticesManifestFetcher {
    noticesMetadataFile;
    noticeManifestValidator;
    noticesManifestUrl;
    cacheTTLMs;
    _fetch;
    _printer;
    // In memory cache
    cachedManifest = { notices: [] };
    refreshedAt = undefined;
    /**
     * Creates new notices manifest fetcher.
     */
    constructor(noticesMetadataFile = noticesMetadataFileInstance, noticeManifestValidator = new NoticesManifestValidator({
        checkLinksWithGitHubApi: false,
    }), noticesManifestUrl = process.env
        .AMPLIFY_BACKEND_NOTICES_ENDPOINT ||
        'https://notices.cli.amplify.aws/notices.json', cacheTTLMs = 60 * 60 * 1000, // one hour
    _fetch = fetch, _printer = printer) {
        this.noticesMetadataFile = noticesMetadataFile;
        this.noticeManifestValidator = noticeManifestValidator;
        this.noticesManifestUrl = noticesManifestUrl;
        this.cacheTTLMs = cacheTTLMs;
        this._fetch = _fetch;
        this._printer = _printer;
    }
    fetchNoticesManifest = async () => {
        if (this.isStale()) {
            await this.tryLoadManifestFromDisk();
        }
        if (this.isStale()) {
            await this.loadManifestFromWebsiteAndCache();
        }
        return this.cachedManifest;
    };
    isStale = () => {
        if (!this.refreshedAt) {
            return true;
        }
        return Date.now() - this.refreshedAt > this.cacheTTLMs;
    };
    loadManifestFromWebsiteAndCache = async () => {
        const response = await this._fetch(this.noticesManifestUrl);
        if (!response.ok) {
            throw new AmplifyFault('NoticeManifestFetchFault', {
                message: `Attempt to fetch notices manifest failed, url=${this.noticesManifestUrl}, statusCode=${response.status}`,
            });
        }
        const noticesManifest = noticesManifestSchema.parse(await response.json());
        await this.noticeManifestValidator.validate(noticesManifest);
        this.cachedManifest = noticesManifest;
        this.refreshedAt = Date.now();
        const metadata = await this.noticesMetadataFile.read();
        metadata.manifestCache = {
            noticesManifest: noticesManifest,
            refreshedAt: this.refreshedAt,
        };
        await this.noticesMetadataFile.write(metadata);
    };
    tryLoadManifestFromDisk = async () => {
        try {
            const metadata = await this.noticesMetadataFile.read();
            await this.noticeManifestValidator.validate(metadata.manifestCache.noticesManifest);
            this.cachedManifest = metadata.manifestCache.noticesManifest;
            this.refreshedAt = metadata.manifestCache.refreshedAt;
        }
        catch (e) {
            this._printer.log('Unable to read cached notices manifest', LogLevel.DEBUG);
            if (e instanceof Error) {
                this._printer.log(e.message, LogLevel.DEBUG);
            }
        }
    };
}
