// @aws-amplify/client-config/lib/client-config-writer/client_config_writer.js
import _fsp from 'fs/promises';
import path from 'path';
import { ClientConfigFormat, } from '../client-config-types/client_config.js';
import { AmplifyUserError } from '@aws-amplify/platform-core';
/**
 * A class that persists client config to disk.
 */
export class ClientConfigWriter {
    pathResolver;
    nameResolver;
    formatter;
    fsp;
    /**
     * Creates client config writer
     */
    constructor(pathResolver, nameResolver, formatter, fsp = _fsp) {
        this.pathResolver = pathResolver;
        this.nameResolver = nameResolver;
        this.formatter = formatter;
        this.fsp = fsp;
    }
    /**
     * Persists provided client config as json file to target path.
     */
    writeClientConfig = async (clientConfig, version, outDir, format = ClientConfigFormat.JSON) => {
        const targetPath = await this.pathResolver(this.nameResolver(version), outDir, format);
        const fileContent = this.formatter.format(clientConfig, format);
        try {
            await this.fsp.writeFile(targetPath, fileContent);
        }
        catch (err) {
            const error = err;
            if (error.message.includes('EACCES')) {
                throw new AmplifyUserError('PermissionsError', {
                    message: `You do not have the permissions to write to this file: ${targetPath}`,
                    resolution: `Ensure that you have the right permissions to write to ${targetPath}.`,
                }, error);
            }
            else {
                throw error;
            }
        }
        return {
            filesWritten: [path.relative(process.cwd(), targetPath)],
        };
    };
}
