// @aws-amplify/backend-auth/lib/auth_access_policy_arbiter.js
/**
 * Middleman between creating bucket policies and attaching those policies to corresponding roles
 */
export class AuthAccessPolicyArbiter {
    accessDefinition;
    getInstanceProps;
    ssmEnvironmentEntries;
    userPoolAccessPolicyFactory;
    /**
     * Instantiate with context from the auth factory
     */
    constructor(accessDefinition, getInstanceProps, ssmEnvironmentEntries, userPoolAccessPolicyFactory) {
        this.accessDefinition = accessDefinition;
        this.getInstanceProps = getInstanceProps;
        this.ssmEnvironmentEntries = ssmEnvironmentEntries;
        this.userPoolAccessPolicyFactory = userPoolAccessPolicyFactory;
    }
    /**
     * Responsible for creating policies corresponding to the definition,
     * then invoking the corresponding ResourceAccessAcceptor to accept the policies
     */
    arbitratePolicies = () => {
        this.accessDefinition.forEach(this.acceptResourceAccess);
    };
    acceptResourceAccess = (accessDefinition) => {
        const accessAcceptor = accessDefinition.getResourceAccessAcceptor(this.getInstanceProps);
        const policy = this.userPoolAccessPolicyFactory.createPolicy(accessDefinition.actions);
        accessAcceptor.acceptResourceAccess(policy, this.ssmEnvironmentEntries);
    };
}
/**
 *
 */
export class AuthAccessPolicyArbiterFactory {
    getInstance = (accessDefinition, getInstanceProps, ssmEnvironmentEntries, userpoolAccessPolicyFactory) => new AuthAccessPolicyArbiter(accessDefinition, getInstanceProps, ssmEnvironmentEntries, userpoolAccessPolicyFactory);
}
