// @aws-amplify/platform-core/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.telemetrySpanAttributeCountLimit = exports.TELEMETRY_ENABLED = exports.telemetryPayloadSchema = exports.TagName = exports.CDKContextKey = exports.USAGE_DATA_TRACKING_ENABLED = void 0;
__exportStar(require("./backend_identifier_conversions.js"), exports);
__exportStar(require("./backend_entry_point_locator.js"), exports);
__exportStar(require("./caller_directory_extractor.js"), exports);
__exportStar(require("./extract_file_path_from_stack_trace_line.js"), exports);
__exportStar(require("./package_json_reader.js"), exports);
__exportStar(require("./usage-data/usage_data_emitter_factory.js"), exports);
__exportStar(require("./config/local_configuration_controller_factory.js"), exports);
__exportStar(require("./config/typed_configuration_file_factory.js"), exports);
__exportStar(require("./errors"), exports);
var constants_js_1 = require("./usage-data/constants.js");
Object.defineProperty(exports, "USAGE_DATA_TRACKING_ENABLED", { enumerable: true, get: function () { return constants_js_1.USAGE_DATA_TRACKING_ENABLED; } });
var cdk_context_key_js_1 = require("./cdk_context_key.js");
Object.defineProperty(exports, "CDKContextKey", { enumerable: true, get: function () { return cdk_context_key_js_1.CDKContextKey; } });
__exportStar(require("./parameter_path_conversions.js"), exports);
__exportStar(require("./object_accumulator.js"), exports);
var tag_name_js_1 = require("./tag_name.js");
Object.defineProperty(exports, "TagName", { enumerable: true, get: function () { return tag_name_js_1.TagName; } });
__exportStar(require("./naming_convention_conversions.js"), exports);
__exportStar(require("./telemetry/telemetry_span_processor_factory.js"), exports);
var telemetry_payload_js_1 = require("./telemetry/telemetry_payload.js");
Object.defineProperty(exports, "telemetryPayloadSchema", { enumerable: true, get: function () { return telemetry_payload_js_1.telemetryPayloadSchema; } });
__exportStar(require("./telemetry/set_span_attributes.js"), exports);
__exportStar(require("./telemetry/translate_error_to_telemetry_error_details.js"), exports);
var constants_js_2 = require("./telemetry/constants.js");
Object.defineProperty(exports, "TELEMETRY_ENABLED", { enumerable: true, get: function () { return constants_js_2.TELEMETRY_ENABLED; } });
Object.defineProperty(exports, "telemetrySpanAttributeCountLimit", { enumerable: true, get: function () { return constants_js_2.telemetrySpanAttributeCountLimit; } });
