// @aws-amplify/platform-core/lib/object_accumulator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectAccumulator = exports.ObjectAccumulatorVersionMismatchError = exports.ObjectAccumulatorPropertyAlreadyExistsError = void 0;
const lodash_mergewith_1 = __importDefault(require("lodash.mergewith"));
const semver_1 = __importDefault(require("semver"));
/**
 * This error is thrown when there's a collision in the object keys
 */
class ObjectAccumulatorPropertyAlreadyExistsError extends Error {
    key;
    existingValue;
    incomingValue;
    /**
     * Creates property already exists error.
     */
    constructor(key, existingValue, incomingValue) {
        super(`Property ${key} already exists`);
        this.key = key;
        this.existingValue = existingValue;
        this.incomingValue = incomingValue;
    }
}
exports.ObjectAccumulatorPropertyAlreadyExistsError = ObjectAccumulatorPropertyAlreadyExistsError;
/**
 * This error is thrown when partial objects with different versions are being accumulated
 */
class ObjectAccumulatorVersionMismatchError extends Error {
    existingVersion;
    newVersion;
    /**
     * Creates property already exists error.
     */
    constructor(existingVersion, newVersion) {
        super(`Version mismatch: Cannot accumulate new objects with version ${newVersion} with existing accumulated object with version ${existingVersion}`);
        this.existingVersion = existingVersion;
        this.newVersion = newVersion;
    }
}
exports.ObjectAccumulatorVersionMismatchError = ObjectAccumulatorVersionMismatchError;
/**
 * A class that can accumulate (squash merge) objects into single instance.
 */
class ObjectAccumulator {
    accumulator;
    versionKey;
    /**
     * creates object accumulator.
     */
    constructor(accumulator, versionKey = 'version') {
        this.accumulator = accumulator;
        this.versionKey = versionKey;
    }
    /**
     * Accumulate a new object part with accumulator.
     * This method throws if there is any intersection between the object parts
     * except for the versionKey, which should be the same across all object parts (nested objects included)
     * @param part a new object part to accumulate
     * @returns the accumulator object for easy chaining
     */
    accumulate = (part) => {
        (0, lodash_mergewith_1.default)(this.accumulator, part, (existingValue, incomingValue, key) => {
            if (Array.isArray(existingValue)) {
                return existingValue.concat(incomingValue);
            }
            if (existingValue && typeof existingValue !== 'object') {
                if (key === this.versionKey) {
                    const incomingVersion = semver_1.default.coerce(incomingValue);
                    const existingVersion = semver_1.default.coerce(existingValue);
                    if (incomingVersion && existingVersion) {
                        // Only throw if the major version is not equal
                        if (incomingVersion.major !== existingVersion.major) {
                            throw new ObjectAccumulatorVersionMismatchError(existingValue, incomingValue);
                        }
                        else {
                            // We always get the max version to persist in the accumulated object
                            return semver_1.default.gte(incomingVersion, existingVersion)
                                ? incomingValue
                                : existingValue;
                        }
                    }
                }
                else if (key !== this.versionKey) {
                    throw new ObjectAccumulatorPropertyAlreadyExistsError(key, existingValue, incomingValue);
                }
            }
            // returning undefined falls back to default merge algorithm
            return undefined;
        });
        return this;
    };
    getAccumulatedObject = () => {
        return this.accumulator;
    };
}
exports.ObjectAccumulator = ObjectAccumulator;
