// @aws-amplify/cli-core/lib/loggers/cfn-deployment-progress/rewritable_block.js
import wrapAnsi from 'wrap-ansi';
import { EOL } from 'os';
/**
 * A class representing re-writable display lines
 */
export class RewritableBlock {
    getBlockWidth;
    getBlockHeight;
    ioHost;
    lastHeight = 0;
    /**
     * Constructor for RewritableBlock
     * @param getBlockWidth A function that returns the width of the block
     * @param getBlockHeight A function that returns the height of the block
     * @param ioHost to send new updates to display
     */
    constructor(getBlockWidth, getBlockHeight, ioHost) {
        this.getBlockWidth = getBlockWidth;
        this.getBlockHeight = getBlockHeight;
        this.ioHost = ioHost;
    }
    /**
     * Display the given lines in this rewritable block. It expands to make room for more lines
     * and compress when not needed. It also takes care of not overflowing the block.
     */
    async displayLines(lines) {
        lines = this.terminalWrap(this.getBlockWidth(), this.expandNewlines(lines));
        lines = lines.slice(0, this.getMaxBlockHeight(this.getBlockHeight(), this.lastHeight, lines));
        const progressUpdate = [];
        for (const line of lines) {
            progressUpdate.push(this.cll() + line + EOL);
        }
        await this.ioHost.notify({
            message: progressUpdate.join(''),
            code: 'AMPLIFY_CFN_PROGRESS_UPDATE',
            time: new Date(),
            action: 'amplify',
            level: 'info',
            data: undefined,
        });
        this.lastHeight = Math.max(this.lastHeight, lines.length);
    }
    /**
     * Clear to the end of line
     */
    cll = () => {
        return '\x1B[K';
    };
    /**
     * Wrap extra long lines while still being in the rewritable block
     */
    terminalWrap = (width, lines) => {
        if (width === undefined) {
            return lines;
        }
        return lines.flatMap((line) => wrapAnsi(line, width - 1, {
            hard: true,
            trim: true,
            wordWrap: false,
        }).split(EOL));
    };
    /**
     * Make sure there are no hidden newlines in the input strings
     */
    expandNewlines = (lines) => {
        return lines.flatMap((line) => line.split(EOL));
    };
    getMaxBlockHeight = (windowHeight, lastHeight, lines) => {
        if (windowHeight === undefined) {
            return Math.max(lines.length, lastHeight);
        }
        return lines.length < windowHeight ? lines.length : windowHeight - 1;
    };
}
