// @aws-amplify/backend-deployer/lib/cdk_deployer_singleton_factory.js
import { CDKDeployer } from './cdk_deployer.js';
import { CdkErrorMapper } from './cdk_error_mapper.js';
import { BackendLocator } from '@aws-amplify/platform-core';
import { BaseCredentials, Toolkit } from '@aws-cdk/toolkit-lib';
/**
 * Factory to create a backend deployer
 */
export class BackendDeployerFactory {
    packageManagerController;
    formatter;
    backendDeployerIOHost;
    sdkProfileResolver;
    static instance;
    /**
     * constructor - sets the packageManagerController
     */
    constructor(packageManagerController, formatter, backendDeployerIOHost, sdkProfileResolver) {
        this.packageManagerController = packageManagerController;
        this.formatter = formatter;
        this.backendDeployerIOHost = backendDeployerIOHost;
        this.sdkProfileResolver = sdkProfileResolver;
    }
    /**
     * Returns a single instance of BackendDeployer
     */
    getInstance() {
        if (!BackendDeployerFactory.instance) {
            BackendDeployerFactory.instance = new CDKDeployer(new CdkErrorMapper(this.formatter), new BackendLocator(), this.packageManagerController, new Toolkit({
                ioHost: this.backendDeployerIOHost,
                emojis: false,
                color: false,
                sdkConfig: {
                    baseCredentials: BaseCredentials.awsCliCompatible({
                        profile: this.sdkProfileResolver(),
                    }),
                },
            }), this.backendDeployerIOHost);
        }
        return BackendDeployerFactory.instance;
    }
}
