// @aws-amplify/backend-cli/lib/commands/sandbox/sandbox-secret/sandbox_secret_set_command.js
import { AmplifyPrompter, printer } from '@aws-amplify/cli-core';
import { once } from 'events';
import { AmplifyUserError } from '@aws-amplify/platform-core';
/**
 * Command to set sandbox secret.
 */
export class SandboxSecretSetCommand {
    sandboxIdResolver;
    secretClient;
    readStream;
    /**
     * @inheritDoc
     */
    command;
    /**
     * @inheritDoc
     */
    describe;
    /**
     * Set sandbox secret command.
     */
    constructor(sandboxIdResolver, secretClient, readStream = process.stdin) {
        this.sandboxIdResolver = sandboxIdResolver;
        this.secretClient = secretClient;
        this.readStream = readStream;
        this.command = 'set <secret-name>';
        this.describe = 'Set a sandbox secret';
    }
    /**
     * @inheritDoc
     */
    handler = async (args) => {
        const secretValue = await this.readSecretValue();
        const secretIdentifier = await this.secretClient.setSecret(await this.sandboxIdResolver.resolve(args.identifier), args.secretName, secretValue);
        printer.print(`Successfully created version ${secretIdentifier.version} of secret ${secretIdentifier.name}`);
    };
    /**
     * @inheritDoc
     */
    builder = (yargs) => {
        return yargs
            .positional('secret-name', {
            describe: 'Name of the secret to set',
            type: 'string',
            demandOption: true,
        })
            .check(async (argv) => {
            if (argv['secret-name']) {
                const secretNameRegex = /^[a-zA-Z0-9_.-]+$/;
                if (!argv['secret-name'].match(secretNameRegex)) {
                    throw new AmplifyUserError('InvalidCommandInputError', {
                        message: `Invalid secret name provided: ${argv['secret-name']}`,
                        resolution: 'Use a secret name that matches [a-zA-Z0-9_.-]+',
                    });
                }
            }
            return true;
        });
    };
    /**
     * Prompt (or) read secret value from stdin based on terminal interactive mode
     */
    readSecretValue = async () => {
        let secretValue = '';
        if (this.readStream.isTTY) {
            // This input is for interactive mode.
            secretValue = await AmplifyPrompter.secretValue();
        }
        else {
            // This allows to accept secret value from redirected input `|` and `>`.
            this.readStream.on('readable', () => {
                const chunk = this.readStream.read();
                if (chunk !== null) {
                    secretValue += chunk;
                }
            });
            // Wait for the end of the input.
            await once(this.readStream, 'end');
        }
        return secretValue;
    };
}
