// @aws-amplify/backend-cli/lib/commands/sandbox/sandbox-secret/sandbox_secret_remove_command.js
import { printer } from '@aws-amplify/cli-core';
import { AmplifyUserError } from '@aws-amplify/platform-core';
/**
 * Command to remove sandbox secret.
 */
export class SandboxSecretRemoveCommand {
    sandboxIdResolver;
    secretClient;
    /**
     * @inheritDoc
     */
    command;
    /**
     * @inheritDoc
     */
    describe;
    /**
     * Remove sandbox secret command.
     */
    constructor(sandboxIdResolver, secretClient) {
        this.sandboxIdResolver = sandboxIdResolver;
        this.secretClient = secretClient;
        this.command = 'remove [secret-name]';
        this.describe = 'Remove a sandbox secret';
    }
    /**
     * @inheritDoc
     */
    handler = async (args) => {
        const sandboxBackendIdentifier = await this.sandboxIdResolver.resolve(args.identifier);
        if (args.secretName) {
            await this.secretClient.removeSecret(sandboxBackendIdentifier, args.secretName);
            printer.print(`Successfully removed secret ${args.secretName}`);
        }
        else if (args.all) {
            const secrets = await this.secretClient.listSecrets(sandboxBackendIdentifier);
            const names = secrets.map((secret) => secret.name);
            await this.secretClient.removeSecrets(sandboxBackendIdentifier, names);
            printer.print('Successfully removed all secrets');
        }
    };
    /**
     * @inheritDoc
     */
    builder = (yargs) => {
        return yargs
            .option('all', {
            describe: 'Remove all secrets',
            type: 'boolean',
            conflicts: ['secret-name'],
        })
            .positional('secret-name', {
            describe: 'Name of the secret to remove',
            type: 'string',
            demandOption: false,
        })
            .check((argv) => {
            if (!argv.all && !argv['secret-name']) {
                throw new AmplifyUserError('InvalidCommandInputError', {
                    message: 'Either secret-name or all flag must be provided',
                    resolution: 'Provide either secret-name or all flag',
                });
            }
            return true;
        });
    };
}
