// @aws-amplify/graphql-conversation-transformer/lib/tools/process-tools.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.processTools = exports.isCustomQueryToolPredicate = exports.isModelOperationToolPredicate = void 0;
const graphql_transformer_core_1 = require("@aws-amplify/graphql-transformer-core");
const graphql_1 = require("graphql");
const graphql_transformer_common_1 = require("graphql-transformer-common");
const graphql_json_schema_type_1 = require("./graphql-json-schema-type");
const isModelOperationToolPredicate = (tool) => tool.modelName !== undefined && tool.modelOperation !== undefined && tool.queryName === undefined;
exports.isModelOperationToolPredicate = isModelOperationToolPredicate;
const isCustomQueryToolPredicate = (tool) => tool.queryName !== undefined && tool.modelName === undefined && tool.modelOperation === undefined;
exports.isCustomQueryToolPredicate = isCustomQueryToolPredicate;
const processTools = (toolDefinitions, ctx) => {
    if (!toolDefinitions || toolDefinitions.length === 0) {
        return undefined;
    }
    const queryType = ctx.output.getType('Query');
    if (!queryType.fields || queryType.fields.length === 0) {
        throw new graphql_transformer_core_1.InvalidDirectiveError('Tools must be queries - no queries found in the schema');
    }
    const tools = toolDefinitions.map((toolDefinition) => {
        var _a;
        const queryName = (0, exports.isModelOperationToolPredicate)(toolDefinition) ? modelListQueryName(toolDefinition, ctx) : toolDefinition.queryName;
        const queryField = (_a = queryType.fields) === null || _a === void 0 ? void 0 : _a.find((field) => field.name.value === queryName);
        if (!queryField) {
            throw new graphql_transformer_core_1.InvalidDirectiveError(`Tool "${toolDefinition.name}" defined in @conversation directive has no matching Query field definition`);
        }
        return createTool({ toolDefinition, queryField, ctx });
    });
    return tools;
};
exports.processTools = processTools;
const generateSelectionSet = (currentType, ctx, seenTypes = new Set(), fieldName = '') => {
    const enums = ctx.output.getTypeDefinitionsOfKind(graphql_1.Kind.ENUM_TYPE_DEFINITION);
    if ((0, graphql_transformer_common_1.isScalarOrEnum)(currentType, enums)) {
        return fieldName;
    }
    const typeName = (0, graphql_transformer_common_1.getBaseType)(currentType);
    if (seenTypes.has(typeName)) {
        return '';
    }
    const type = getObjectTypeFromName(typeName, ctx);
    seenTypes.add(type.name.value);
    if (!type.fields || type.fields.length === 0) {
        return '';
    }
    const fieldSelections = type.fields
        .map((field) => {
        if ((0, graphql_transformer_common_1.isScalarOrEnum)(field.type, enums)) {
            return field.name.value;
        }
        else {
            const nestedSelection = generateSelectionSet(field.type, ctx, new Set(seenTypes), field.name.value);
            return nestedSelection ? `${field.name.value} { ${nestedSelection} }` : '';
        }
    })
        .filter(Boolean);
    return fieldSelections.join(' ');
};
const getObjectTypeFromName = (name, ctx) => {
    const node = ctx.output.getObject(name);
    if (!node) {
        throw new Error(`Could not find type definition for ${name}`);
    }
    return node;
};
const createTool = (input) => {
    const { toolDefinition, queryField, ctx } = input;
    const { name: toolName, description } = toolDefinition;
    const { type: returnType, arguments: fieldArguments } = queryField;
    const { properties, required } = generateToolProperties(fieldArguments, ctx, toolDefinition);
    const selectionSet = generateSelectionSet(returnType, ctx).trim();
    const propertyTypes = generatePropertyTypes(fieldArguments, toolDefinition);
    const graphqlRequestInputDescriptor = {
        selectionSet,
        propertyTypes,
        queryName: queryField.name.value,
    };
    return {
        name: toolName,
        description,
        inputSchema: {
            json: {
                type: 'object',
                properties,
                required,
            },
        },
        graphqlRequestInputDescriptor,
    };
};
const generateToolProperties = (fieldArguments, ctx, toolDefinition) => {
    if (!fieldArguments || fieldArguments.length === 0) {
        return { properties: {}, required: [] };
    }
    if ((0, exports.isModelOperationToolPredicate)(toolDefinition)) {
        return { properties: {}, required: [] };
    }
    return fieldArguments.reduce((acc, fieldArgument) => {
        const fieldArgumentSchema = (0, graphql_json_schema_type_1.generateJSONSchemaFromTypeNode)(fieldArgument.type, ctx);
        acc.properties[fieldArgument.name.value] = fieldArgumentSchema;
        if (fieldArgument.type.kind === 'NonNullType') {
            acc.required.push(fieldArgument.name.value);
        }
        return acc;
    }, { properties: {}, required: [] });
};
const generatePropertyTypes = (fieldArguments, toolDefinition) => {
    if (!fieldArguments || fieldArguments.length === 0) {
        return {};
    }
    if ((0, exports.isModelOperationToolPredicate)(toolDefinition)) {
        return {};
    }
    return fieldArguments.reduce((acc, fieldArgument) => {
        const name = fieldArgument.name.value;
        const suffix = (0, graphql_transformer_common_1.isNonNullType)(fieldArgument.type) ? '!' : '';
        acc[name] = (0, graphql_transformer_common_1.getBaseType)(fieldArgument.type) + suffix;
        return acc;
    }, {});
};
const modelListQueryName = (modelTool, ctx) => {
    var _a, _b, _c;
    const { modelName } = modelTool;
    const model = getObjectTypeFromName(modelName, ctx);
    const modelDirective = (_a = model.directives) === null || _a === void 0 ? void 0 : _a.find((directive) => directive.name.value === 'model');
    const queriesArgument = (_b = modelDirective === null || modelDirective === void 0 ? void 0 : modelDirective.arguments) === null || _b === void 0 ? void 0 : _b.find((arg) => arg.name.value === 'queries');
    const queriesValue = queriesArgument === null || queriesArgument === void 0 ? void 0 : queriesArgument.value;
    const listField = (_c = queriesValue === null || queriesValue === void 0 ? void 0 : queriesValue.fields) === null || _c === void 0 ? void 0 : _c.find((field) => field.name.value === 'list');
    const listValue = listField === null || listField === void 0 ? void 0 : listField.value;
    const listQueryArgument = listValue === null || listValue === void 0 ? void 0 : listValue.value;
    return listQueryArgument !== null && listQueryArgument !== void 0 ? listQueryArgument : (0, graphql_transformer_core_1.getFieldNameFor)('list', modelName);
};
