// @aws-cdk/toolkit-lib/lib/util/serialize.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toYAML = toYAML;
exports.deserializeStructure = deserializeStructure;
exports.serializeStructure = serializeStructure;
exports.loadStructuredFile = loadStructuredFile;
exports.obscureTemplate = obscureTemplate;
exports.replacerBufferWithInfo = replacerBufferWithInfo;
const fs = require("fs/promises");
const bytes_1 = require("./bytes");
const yaml_cfn = require("./yaml-cfn");
/**
 * Stringify to YAML
 */
function toYAML(obj) {
    return yaml_cfn.serialize(obj);
}
/**
 * Parse either YAML or JSON
 */
function deserializeStructure(str) {
    return yaml_cfn.deserialize(str);
}
/**
 * Serialize to either YAML or JSON
 */
function serializeStructure(object, json) {
    if (json) {
        return JSON.stringify(object, undefined, 2);
    }
    else {
        return toYAML(object);
    }
}
/**
 * Load a YAML or JSON file from disk
 */
async function loadStructuredFile(fileName) {
    const contents = await fs.readFile(fileName, { encoding: 'utf-8' });
    return deserializeStructure(contents);
}
/**
 * Remove any template elements that we don't want to show users.
 */
function obscureTemplate(template = {}) {
    if (template.Rules) {
        // see https://github.com/aws/aws-cdk/issues/17942
        if (template.Rules.CheckBootstrapVersion) {
            if (Object.keys(template.Rules).length > 1) {
                delete template.Rules.CheckBootstrapVersion;
            }
            else {
                delete template.Rules;
            }
        }
    }
    return template;
}
/**
 * Detects a buffer that has been converted to a JSON-like object
 * In Node, `Buffer`s have `toJSON()` method that converts the buffer
 * into a JS object that can be JSON stringified.
 * Unfortunately this conversion happens before the replacer is called,
 * so normal means of detecting a `Buffer` objet don't work anymore.
 * @see https://github.com/nodejs/node-v0.x-archive/issues/5110
 */
function isJsonBuffer(value) {
    return typeof value === 'object'
        && 'type' in value
        && value.type === 'Buffer'
        && 'data' in value
        && Array.isArray(value.data);
}
/**
 * A JSON.stringify() replacer that converts Buffers into a string with information
 * Use this if you plan to print out JSON stringified objects that may contain a Buffer.
 * Without this, large buffers (think: Megabytes) will completely fill up the output
 * and even crash the system.
 */
function replacerBufferWithInfo(_key, value) {
    if (isJsonBuffer(value)) {
        return `<Buffer: ${(0, bytes_1.formatBytes)(value.data.length)}>`;
    }
    return value;
}
