// @aws-cdk/toolkit-lib/lib/util/string-manipulation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.padLeft = padLeft;
exports.padRight = padRight;
exports.formatTime = formatTime;
exports.lowerCaseFirstCharacter = lowerCaseFirstCharacter;
/**
 * Pad 's' on the left with 'char' until it is n characters wide
 */
function padLeft(n, x, char = ' ') {
    return char.repeat(Math.max(0, n - x.length)) + x;
}
/**
 * Pad 's' on the right with 'char' until it is n characters wide
 */
function padRight(n, x, char = ' ') {
    return x + char.repeat(Math.max(0, n - x.length));
}
/**
 * Formats time in milliseconds (which we get from 'Date.getTime()')
 * to a human-readable time; returns time in seconds rounded to 2
 * decimal places.
 */
function formatTime(num) {
    return roundPercentage(millisecondsToSeconds(num));
}
/**
 * Rounds a decimal number to two decimal points.
 * The function is useful for fractions that need to be outputted as percentages.
 */
function roundPercentage(num) {
    return Math.round(100 * num) / 100;
}
/**
 * Given a time in milliseconds, return an equivalent amount in seconds.
 */
function millisecondsToSeconds(num) {
    return num / 1000;
}
/**
 * This function lower cases the first character of the string provided.
 */
function lowerCaseFirstCharacter(str) {
    return str.length > 0 ? `${str[0].toLowerCase()}${str.slice(1)}` : str;
}
