// @aws-cdk/toolkit-lib/lib/api/cloud-assembly/stack-selector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpandStackSelection = exports.StackSelectionStrategy = void 0;
/**
 * Which stacks should be selected from a cloud assembly
 */
var StackSelectionStrategy;
(function (StackSelectionStrategy) {
    /**
     * Returns all stacks in the app regardless of patterns,
     * including stacks inside nested assemblies.
     */
    StackSelectionStrategy["ALL_STACKS"] = "all-stacks";
    /**
     * Returns all stacks in the main (top level) assembly only.
     */
    StackSelectionStrategy["MAIN_ASSEMBLY"] = "main-assembly";
    /**
     * If the assembly includes a single stack, returns it.
     * Otherwise throws an exception.
     */
    StackSelectionStrategy["ONLY_SINGLE"] = "only-single";
    /**
     * Return stacks matched by patterns.
     * If no stacks are found, execution is halted successfully.
     * Most likely you don't want to use this but `StackSelectionStrategy.MUST_MATCH_PATTERN`
     */
    StackSelectionStrategy["PATTERN_MATCH"] = "pattern-match";
    /**
     * Return stacks matched by patterns.
     * Throws an exception if the patterns don't match at least one stack in the assembly.
     */
    StackSelectionStrategy["PATTERN_MUST_MATCH"] = "pattern-must-match";
    /**
     * Returns if exactly one stack is matched by the pattern(s).
     * Throws an exception if no stack, or more than exactly one stack are matched.
     */
    StackSelectionStrategy["PATTERN_MUST_MATCH_SINGLE"] = "pattern-must-match-single";
})(StackSelectionStrategy || (exports.StackSelectionStrategy = StackSelectionStrategy = {}));
/**
 * When selecting stacks, what other stacks to include because of dependencies
 */
var ExpandStackSelection;
(function (ExpandStackSelection) {
    /**
     * Don't select any extra stacks
     */
    ExpandStackSelection["NONE"] = "none";
    /**
     * Include stacks that this stack depends on
     */
    ExpandStackSelection["UPSTREAM"] = "upstream";
    /**
     * Include stacks that depend on this stack
     */
    ExpandStackSelection["DOWNSTREAM"] = "downstream";
    /**
     * @TODO
     * Include both directions.
     * I.e. stacks that this stack depends on, and stacks that depend on this stack.
     */
    // FULL = 'full',
})(ExpandStackSelection || (exports.ExpandStackSelection = ExpandStackSelection = {}));
