// @aws-cdk/toolkit-lib/lib/api/notices/cached-data-source.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CachedDataSource = void 0;
const fs = require("fs-extra");
const toolkit_error_1 = require("../../toolkit/toolkit-error");
const TIME_TO_LIVE_SUCCESS = 60 * 60 * 1000; // 1 hour
const TIME_TO_LIVE_ERROR = 1 * 60 * 1000; // 1 minute
class CachedDataSource {
    ioHelper;
    fileName;
    dataSource;
    skipCache;
    constructor(ioHelper, fileName, dataSource, skipCache) {
        this.ioHelper = ioHelper;
        this.fileName = fileName;
        this.dataSource = dataSource;
        this.skipCache = skipCache;
    }
    async fetch() {
        const cachedData = await this.load();
        const data = cachedData.notices;
        const expiration = cachedData.expiration ?? 0;
        if (Date.now() > expiration || this.skipCache) {
            let updatedData = cachedData;
            try {
                updatedData = await this.fetchInner();
            }
            catch (e) {
                await this.ioHelper.defaults.debug(`Could not refresh notices: ${e}`);
                updatedData = {
                    expiration: Date.now() + TIME_TO_LIVE_ERROR,
                    notices: [],
                };
                throw toolkit_error_1.ToolkitError.withCause('Failed to load CDK notices. Please try again later.', e);
            }
            finally {
                await this.save(updatedData);
            }
            return updatedData.notices;
        }
        else {
            await this.ioHelper.defaults.debug(`Reading cached notices from ${this.fileName}`);
            return data;
        }
    }
    async fetchInner() {
        return {
            expiration: Date.now() + TIME_TO_LIVE_SUCCESS,
            notices: await this.dataSource.fetch(),
        };
    }
    async load() {
        const defaultValue = {
            expiration: 0,
            notices: [],
        };
        try {
            return fs.existsSync(this.fileName)
                ? await fs.readJSON(this.fileName)
                : defaultValue;
        }
        catch (e) {
            await this.ioHelper.defaults.debug(`Failed to load notices from cache: ${e}`);
            return defaultValue;
        }
    }
    async save(cached) {
        try {
            await fs.ensureFile(this.fileName);
            await fs.writeJSON(this.fileName, cached);
        }
        catch (e) {
            await this.ioHelper.defaults.debug(`Failed to store notices in the cache: ${e}`);
        }
    }
}
exports.CachedDataSource = CachedDataSource;
