// @aws-amplify/cli-core/lib/loggers/test-assets/failed_auth_structured_cdk_events.js
export const data = [
    {
        code: 'SYNTH_STARTED',
        action: 'amplify',
        message: 'Backend synthesis started',
    },
    {
        code: 'CDK_ASSEMBLY_I0000',
        message: 'Setting "CDK_DEFAULT_REGION" environment variable to [ \'us-west-2\' ]',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_ASSEMBLY_I0000',
        message: 'Setting "CDK_DEFAULT_ACCOUNT" environment variable to [ \'123456789012\' ]',
    },
    {
        code: 'CDK_ASSEMBLY_I0000',
        message: "outdir: [\n  '/Users/user/Workspaces/sample-project-app/.amplify/artifacts/cdk.out'\n]",
    },
    {
        code: 'CDK_ASSEMBLY_I0000',
        message: "env: [\n  {\n    CDK_DEFAULT_REGION: 'us-west-2',\n    CDK_DEFAULT_ACCOUNT: '123456789012',\n    CDK_OUTDIR: '/Users/user/Workspaces/sample-project-app/.amplify/artifacts/cdk.out',\n    CDK_CLI_ASM_VERSION: '41.0.0',\n    CDK_CLI_VERSION: '0.1.4'\n  }\n]",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "context: {\n  'amplify-backend-namespace': 'sample-project-app',\n  'amplify-backend-name': 'user',\n  'amplify-backend-type': 'sandbox',\n  'aws:cdk:enable-path-metadata': true,\n  'aws:cdk:enable-asset-metadata': true,\n  'aws:cdk:version-reporting': true,\n  'aws:cdk:bundling-stacks': [ '**' ]\n}",
    },
    {
        code: 'CDK_TOOLKIT_I1000',
        message: '  Synthesis time: 1.5s',
    },
    {
        code: 'CDK_TOOLKIT_I1901',
        message: 'Successfully synthesized to /Users/user/Workspaces/sample-project-app/.amplify/artifacts/cdk.out',
    },
    {
        code: 'SYNTH_FINISHED',
        action: 'amplify',
        level: 'result',
        message: 'Backend synthesized in 1.5 seconds',
    },
    {
        code: 'TS_STARTED',
        action: 'amplify',
        message: 'Backend type checks started',
    },
    {
        code: 'TS_FINISHED',
        action: 'amplify',
        level: 'result',
        message: 'Type checks completed in 5.04 seconds',
    },
    {
        code: 'CDK_TOOLKIT_I1000',
        message: '  Synthesis time: 0s',
    },
    {
        code: 'CDK_TOOLKIT_W0000',
        message: '⚠️ The --hotswap and --hotswap-fallback flags deliberately introduce CloudFormation drift to speed up deployments\n⚠️ They should only be used for development - never use them for your production Stacks!',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Checking for previously published assets',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c1721bcd774e27c514d3454b5be4f9bdd094c0161b57ddf053d618e3b0086a77.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/ca9699b9a2845c45e80cba1111f95ae2ac773af7d87163ae9c808ec547e38e00.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/a183ddccbd956316c38ef97177b8f088ef0826f62023323f5ae6053d348ccffc.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/9fcbe070ecd3023c5bf5b966fa9584757db9762eef123bad0820bd87591b2174.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/88d5b7081b0dc9c39fd0bffedf56f011b2a7cf1ce5cc227f47a1e94efee8067c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/cc01911d0269d4080ea57505dc445dfc315ef7ad85d3d9d4ea1357858bff451d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/4c6a2d29f01c6091bd1d9afe16e5849d456c96f17c3b215938c8067399532719.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/08f4d557693d96c1a4efba0f9dc91330e4b19772fd5477c156468843e3d9cb5e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"ca9699b9a2845c45e80cba1111f95ae2ac773af7d87163ae9c808ec547e38e00.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/ca9699b9a2845c45e80cba1111f95ae2ac773af7d87163ae9c808ec547e38e00.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/b371d68ddfbdd11b176756ce33b2c4e79d26fcc2fd4b7c0c9e75d06efacee90e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c1721bcd774e27c514d3454b5be4f9bdd094c0161b57ddf053d618e3b0086a77.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c1721bcd774e27c514d3454b5be4f9bdd094c0161b57ddf053d618e3b0086a77.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/47614d6b888488252a8a44b7911b69efcddb1eb2454ca31132d72d6444247204.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"88d5b7081b0dc9c39fd0bffedf56f011b2a7cf1ce5cc227f47a1e94efee8067c.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/88d5b7081b0dc9c39fd0bffedf56f011b2a7cf1ce5cc227f47a1e94efee8067c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"9fcbe070ecd3023c5bf5b966fa9584757db9762eef123bad0820bd87591b2174.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/9fcbe070ecd3023c5bf5b966fa9584757db9762eef123bad0820bd87591b2174.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"a183ddccbd956316c38ef97177b8f088ef0826f62023323f5ae6053d348ccffc.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/a183ddccbd956316c38ef97177b8f088ef0826f62023323f5ae6053d348ccffc.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"cc01911d0269d4080ea57505dc445dfc315ef7ad85d3d9d4ea1357858bff451d.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/cc01911d0269d4080ea57505dc445dfc315ef7ad85d3d9d4ea1357858bff451d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/f4a52b72209a9dfa197b5e7367a5c378c5bb86de6e29ddd9e48b49a3fe54b249.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/1f5fed297da9c32ae3af922bf3a38ccf23b956078887d16891ec06c20c64722c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/06db846fd14e6fc371f22b12b5545ba8e2dbfeda85d8c8d586c71c282166657b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/245ed8ac5185462a6923b7e45a318fec4c9ba0c45d1d720abd5b20317aa3dc6a.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"08f4d557693d96c1a4efba0f9dc91330e4b19772fd5477c156468843e3d9cb5e.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/08f4d557693d96c1a4efba0f9dc91330e4b19772fd5477c156468843e3d9cb5e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/474bf0776ec2164a13191d1a0a9e057154931e4918fea5086f49850d02a5371b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"4c6a2d29f01c6091bd1d9afe16e5849d456c96f17c3b215938c8067399532719.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/4c6a2d29f01c6091bd1d9afe16e5849d456c96f17c3b215938c8067399532719.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/29e4bdb03133f87f2b1acb547959ffd4cfd22804026994c5b61db33084127353.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"b371d68ddfbdd11b176756ce33b2c4e79d26fcc2fd4b7c0c9e75d06efacee90e.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/b371d68ddfbdd11b176756ce33b2c4e79d26fcc2fd4b7c0c9e75d06efacee90e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"1f5fed297da9c32ae3af922bf3a38ccf23b956078887d16891ec06c20c64722c.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/1f5fed297da9c32ae3af922bf3a38ccf23b956078887d16891ec06c20c64722c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/4f7907d1209a2c9953a0c053df402c634e359546d70c7cc5c2e8e21ea734880f.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/6f3a2a9d562ae3058cc1aed6741edbf5edcaef7b3df7b03f52478816a1c95ffb.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"47614d6b888488252a8a44b7911b69efcddb1eb2454ca31132d72d6444247204.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/47614d6b888488252a8a44b7911b69efcddb1eb2454ca31132d72d6444247204.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"06db846fd14e6fc371f22b12b5545ba8e2dbfeda85d8c8d586c71c282166657b.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/06db846fd14e6fc371f22b12b5545ba8e2dbfeda85d8c8d586c71c282166657b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/fe3c43ada4b9d681a5e2312663ef7a73386424d73b73e51f8e2e9d4b50f7c502.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/e0cff47fb007f0bbf2a4e43ca256d6aa7ec109821769fd79fa7c5e83f0e7f9fc.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"29e4bdb03133f87f2b1acb547959ffd4cfd22804026994c5b61db33084127353.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/29e4bdb03133f87f2b1acb547959ffd4cfd22804026994c5b61db33084127353.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/6479af76f5a4d8b61b53dd3cee8e18ae192b629ad15ced7b342690a9ebbdd8c2.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"245ed8ac5185462a6923b7e45a318fec4c9ba0c45d1d720abd5b20317aa3dc6a.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/245ed8ac5185462a6923b7e45a318fec4c9ba0c45d1d720abd5b20317aa3dc6a.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"474bf0776ec2164a13191d1a0a9e057154931e4918fea5086f49850d02a5371b.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/474bf0776ec2164a13191d1a0a9e057154931e4918fea5086f49850d02a5371b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"f4a52b72209a9dfa197b5e7367a5c378c5bb86de6e29ddd9e48b49a3fe54b249.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/f4a52b72209a9dfa197b5e7367a5c378c5bb86de6e29ddd9e48b49a3fe54b249.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/6ce8d75a422db2491407ccfc9807e5de43e86bed091909b805008adb8243b05d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/3cf32e28e3985b88b598dacbcc9a81cbb12cab62b6507082eb20a0702d3c197e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/fface358117b1edafc4dbe09854c750ff9d1cad2ab9aed9cff18b5df478d3d5d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"e0cff47fb007f0bbf2a4e43ca256d6aa7ec109821769fd79fa7c5e83f0e7f9fc.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/e0cff47fb007f0bbf2a4e43ca256d6aa7ec109821769fd79fa7c5e83f0e7f9fc.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"fe3c43ada4b9d681a5e2312663ef7a73386424d73b73e51f8e2e9d4b50f7c502.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/fe3c43ada4b9d681a5e2312663ef7a73386424d73b73e51f8e2e9d4b50f7c502.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"4f7907d1209a2c9953a0c053df402c634e359546d70c7cc5c2e8e21ea734880f.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/4f7907d1209a2c9953a0c053df402c634e359546d70c7cc5c2e8e21ea734880f.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/725b7ed3971927a1e99c853643b4ca575c3d612fff9b50d3f616c4feeb79a7d7.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/97e4049e52fffaf523b49966a1225677c0fd7c09e296e1832b5bb95086e5c3e4.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c89821d3a34d8f9a69feb7faded2de554cfee32627d772bc8d842a1a145408f5.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"6f3a2a9d562ae3058cc1aed6741edbf5edcaef7b3df7b03f52478816a1c95ffb.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/6f3a2a9d562ae3058cc1aed6741edbf5edcaef7b3df7b03f52478816a1c95ffb.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"6479af76f5a4d8b61b53dd3cee8e18ae192b629ad15ced7b342690a9ebbdd8c2.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/6479af76f5a4d8b61b53dd3cee8e18ae192b629ad15ced7b342690a9ebbdd8c2.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"fface358117b1edafc4dbe09854c750ff9d1cad2ab9aed9cff18b5df478d3d5d.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/fface358117b1edafc4dbe09854c750ff9d1cad2ab9aed9cff18b5df478d3d5d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/098c998f215c333f57348ca8912f66057ae6b574fe34013033dc54362ee2372c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/db826cc870f25dcd05aff645c3b384ba3a97115acc5373e580c14379eb0430b3.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/5b2dd26ac19f6a54684f2769d465463159cc27a441d7ac8b8efafdec1b2370fd.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"6ce8d75a422db2491407ccfc9807e5de43e86bed091909b805008adb8243b05d.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/6ce8d75a422db2491407ccfc9807e5de43e86bed091909b805008adb8243b05d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/49e7a1f2fb4c809b82bed6b0242a2a3bcfa2c5276ebd2539ff330fee36f8fe18.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"3cf32e28e3985b88b598dacbcc9a81cbb12cab62b6507082eb20a0702d3c197e.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/3cf32e28e3985b88b598dacbcc9a81cbb12cab62b6507082eb20a0702d3c197e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c3534669fc2486ce8786b1a3efd6c0848cd3c4835de33940c0f49c28c9d8abef.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"97e4049e52fffaf523b49966a1225677c0fd7c09e296e1832b5bb95086e5c3e4.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/97e4049e52fffaf523b49966a1225677c0fd7c09e296e1832b5bb95086e5c3e4.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/dcb70e4856bc9a423e02b8b9e9c2ab3f71479784572d4c04a02d1f5b61a9f9ac.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c89821d3a34d8f9a69feb7faded2de554cfee32627d772bc8d842a1a145408f5.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c89821d3a34d8f9a69feb7faded2de554cfee32627d772bc8d842a1a145408f5.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"098c998f215c333f57348ca8912f66057ae6b574fe34013033dc54362ee2372c.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/098c998f215c333f57348ca8912f66057ae6b574fe34013033dc54362ee2372c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/faa95a81ae7d7373f3e1f242268f904eb748d8d0fdd306e8a6fe515a1905a7d6.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/14700f3f8dd2f4997b0e6380f2714c17996184ef4a12d7990ce58b009105e158.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"5b2dd26ac19f6a54684f2769d465463159cc27a441d7ac8b8efafdec1b2370fd.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/5b2dd26ac19f6a54684f2769d465463159cc27a441d7ac8b8efafdec1b2370fd.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"725b7ed3971927a1e99c853643b4ca575c3d612fff9b50d3f616c4feeb79a7d7.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/725b7ed3971927a1e99c853643b4ca575c3d612fff9b50d3f616c4feeb79a7d7.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c3534669fc2486ce8786b1a3efd6c0848cd3c4835de33940c0f49c28c9d8abef.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c3534669fc2486ce8786b1a3efd6c0848cd3c4835de33940c0f49c28c9d8abef.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/907cb295088857ccb545b8aae83288bd9dfa82115bb0d20a0ae0d08dc8872f8a.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/2f050e6b9affea69843658b09b5d9073acc9cda9170a30ff043def78a33c6ad0.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"db826cc870f25dcd05aff645c3b384ba3a97115acc5373e580c14379eb0430b3.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/db826cc870f25dcd05aff645c3b384ba3a97115acc5373e580c14379eb0430b3.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"49e7a1f2fb4c809b82bed6b0242a2a3bcfa2c5276ebd2539ff330fee36f8fe18.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/49e7a1f2fb4c809b82bed6b0242a2a3bcfa2c5276ebd2539ff330fee36f8fe18.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/d00ac45dd4a60a3d170b74ff85c881d0e4359cbbc3de10dd76a4256db540dd2e.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"faa95a81ae7d7373f3e1f242268f904eb748d8d0fdd306e8a6fe515a1905a7d6.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/faa95a81ae7d7373f3e1f242268f904eb748d8d0fdd306e8a6fe515a1905a7d6.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/193c5f412086be216636e53fa8e74d99c431936435b456c4b0b7252704331c51.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"907cb295088857ccb545b8aae83288bd9dfa82115bb0d20a0ae0d08dc8872f8a.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/907cb295088857ccb545b8aae83288bd9dfa82115bb0d20a0ae0d08dc8872f8a.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/209a0048df688fe6a8997016c796a69b318d368ae30d416e1656d209d881a050.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"dcb70e4856bc9a423e02b8b9e9c2ab3f71479784572d4c04a02d1f5b61a9f9ac.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/dcb70e4856bc9a423e02b8b9e9c2ab3f71479784572d4c04a02d1f5b61a9f9ac.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"14700f3f8dd2f4997b0e6380f2714c17996184ef4a12d7990ce58b009105e158.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/14700f3f8dd2f4997b0e6380f2714c17996184ef4a12d7990ce58b009105e158.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/f7bafc4fb5028045158fda2645f666d5744ff73a7236cbd21784e1b94aaef720.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/83ef1d85f437ccf9bf6d749fd82f93b06ce961419558f4618e4da3765db7a706.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"2f050e6b9affea69843658b09b5d9073acc9cda9170a30ff043def78a33c6ad0.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/2f050e6b9affea69843658b09b5d9073acc9cda9170a30ff043def78a33c6ad0.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/76f3f913f2bffd4d3b833199e6d331903453796a752c920c60d3aa8456e8a975.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/37eb3e2a7ba2bef4bcdf5b23e0f6e728fc652c7d30d73e62878b18dae42ad678.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"d00ac45dd4a60a3d170b74ff85c881d0e4359cbbc3de10dd76a4256db540dd2e.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/d00ac45dd4a60a3d170b74ff85c881d0e4359cbbc3de10dd76a4256db540dd2e.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"f7bafc4fb5028045158fda2645f666d5744ff73a7236cbd21784e1b94aaef720.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/f7bafc4fb5028045158fda2645f666d5744ff73a7236cbd21784e1b94aaef720.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"83ef1d85f437ccf9bf6d749fd82f93b06ce961419558f4618e4da3765db7a706.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/83ef1d85f437ccf9bf6d749fd82f93b06ce961419558f4618e4da3765db7a706.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"193c5f412086be216636e53fa8e74d99c431936435b456c4b0b7252704331c51.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/193c5f412086be216636e53fa8e74d99c431936435b456c4b0b7252704331c51.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"209a0048df688fe6a8997016c796a69b318d368ae30d416e1656d209d881a050.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/209a0048df688fe6a8997016c796a69b318d368ae30d416e1656d209d881a050.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"37eb3e2a7ba2bef4bcdf5b23e0f6e728fc652c7d30d73e62878b18dae42ad678.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"76f3f913f2bffd4d3b833199e6d331903453796a752c920c60d3aa8456e8a975.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/76f3f913f2bffd4d3b833199e6d331903453796a752c920c60d3aa8456e8a975.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '47 total assets, 1 still need to be published',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-deploy-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-deploy-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] SSM.GetParameter({"Name":"/cdk-bootstrap/hnb659fds/version"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5210',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: start: Building 37eb3e2a7ba2bef4bcdf5b23e0f6e728fc652c7d30d73e62878b18dae42ad678',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: success: Built 37eb3e2a7ba2bef4bcdf5b23e0f6e728fc652c7d30d73e62878b18dae42ad678',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"CDKToolkit"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: start: Publishing 37eb3e2a7ba2bef4bcdf5b23e0f6e728fc652c7d30d73e62878b18dae42ad678:current_account-current_region',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/37eb3e2a7ba2bef4bcdf5b23e0f6e728fc652c7d30d73e62878b18dae42ad678.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.GetBucketLocation({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"37eb3e2a7ba2bef4bcdf5b23e0f6e728fc652c7d30d73e62878b18dae42ad678.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.GetBucketEncryption({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: upload: Upload s3://cdk-hnb659fds-assets-123456789012-us-west-2/37eb3e2a7ba2bef4bcdf5b23e0f6e728fc652c7d30d73e62878b18dae42ad678.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.PutObject({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Key":"37eb3e2a7ba2bef4bcdf5b23e0f6e728fc652c7d30d73e62878b18dae42ad678.json","Body":"<Buffer: 14 KiB>","ContentType":"application/json","ChecksumAlgorithm":"SHA256","ServerSideEncryption":"aws:kms"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5221',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: success: Published 37eb3e2a7ba2bef4bcdf5b23e0f6e728fc652c7d30d73e62878b18dae42ad678:current_account-current_region',
    },
    {
        code: 'CDK_TOOLKIT_I5100',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: deploying... [1/1]',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-deploy-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-deploy-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: checking if we can skip deploy',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: template has changed',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: deploying...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Waiting for stack CDKToolkit to finish creating or updating...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"CDKToolkit"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"CDKToolkit"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.GetCallerIdentity({}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataAmplifyTableManagerNestedStackA-1JMJ1D9EH891L"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataAmplifyTableManagerNestedStackA-1JMJ1D9EH891L","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataPostNestedStackPostNestedStackR-Q6BBZ46K7MMS"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataPostNestedStackPostNestedStackR-Q6BBZ46K7MMS","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataPersonNestedStackPersonNestedSt-72810Y0AHA1T"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataPersonNestedStackPersonNestedSt-72810Y0AHA1T","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataConnectionStackNestedStackConne-L130VQDWLV66"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataConnectionStackNestedStackConne-L130VQDWLV66","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '⚠️ The following non-hotswappable changes were found:\n    logicalID: amplifyAuthUserPool4BA7F805, type: AWS::Cognito::UserPool, rejected changes: EmailConfiguration, reason: This resource type is not supported for hotswap deployments',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Could not perform a hotswap deployment, as the stack amplify-sampleprojectapp-user-sandbox-0e1752a889 contains non-Asset changes',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Falling back to doing a full deployment',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Removing existing change set with name cdk-deploy-change-set if it exists',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DeleteChangeSet({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","ChangeSetName":"cdk-deploy-change-set"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Attempting to create ChangeSet with name cdk-deploy-change-set to update stack amplify-sampleprojectapp-user-sandbox-0e1752a889',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: creating CloudFormation changeset...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.CreateChangeSet({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","ChangeSetName":"cdk-deploy-change-set","ChangeSetType":"UPDATE","Description":"CDK Changeset for execution 68c0d34b-36e7-4663-a2f2-9200a5f44eef","ClientToken":"create68c0d34b-36e7-4663-a2f2-9200a5f44eef","ImportExistingResources":false,"Capabilities":["CAPABILITY_IAM","CAPABILITY_NAMED_IAM","CAPABILITY_AUTO_EXPAND"],"Parameters":[],"RoleARN":"arn:aws:iam::123456789012:role/cdk-hnb659fds-cfn-exec-role-123456789012-us-west-2","TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/37eb3e2a7ba2bef4bcdf5b23e0f6e728fc652c7d30d73e62878b18dae42ad678.json","Tags":[{"Key":"amplify:deployment-type","Value":"sandbox"},{"Key":"created-by","Value":"amplify"}]}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Initiated creation of changeset: arn:aws:cloudformation:us-west-2:123456789012:changeSet/cdk-deploy-change-set/02946133-d5f0-450a-9a9b-459e668a8986; waiting for it to finish creating...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Waiting for changeset cdk-deploy-change-set on stack amplify-sampleprojectapp-user-sandbox-0e1752a889 to finish creating...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeChangeSet({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","ChangeSetName":"cdk-deploy-change-set"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Changeset cdk-deploy-change-set on stack amplify-sampleprojectapp-user-sandbox-0e1752a889 is still creating',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeChangeSet({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","ChangeSetName":"cdk-deploy-change-set"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Initiating execution of changeset arn:aws:cloudformation:us-west-2:123456789012:changeSet/cdk-deploy-change-set/02946133-d5f0-450a-9a9b-459e668a8986 on stack amplify-sampleprojectapp-user-sandbox-0e1752a889',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.ExecuteChangeSet({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","ChangeSetName":"cdk-deploy-change-set","ClientRequestToken":"exec68c0d34b-36e7-4663-a2f2-9200a5f44eef"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Execution of changeset arn:aws:cloudformation:us-west-2:123456789012:changeSet/cdk-deploy-change-set/02946133-d5f0-450a-9a9b-459e668a8986 on stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has started; waiting for the update to complete...',
    },
    {
        code: 'CDK_TOOLKIT_I5501',
        message: 'Deploying amplify-sampleprojectapp-user-sandbox-0e1752a889',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Waiting for stack amplify-sampleprojectapp-user-sandbox-0e1752a889 to finish creating or updating...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has an ongoing operation in progress and is not stable (UPDATE_IN_PROGRESS (User Initiated))',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 0/5 | 11:31:09 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | amplify-sampleprojectapp-user-sandbox-0e1752a889 User Initiated',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: '462e5310-fe29-11ef-b168-06a2818fc099',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:09.619'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceStatusReason: 'User Initiated',
                ClientRequestToken: 'exec68c0d34b-36e7-4663-a2f2-9200a5f44eef',
            },
            progress: {
                total: 5,
                completed: 0,
                formatted: '0/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has an ongoing operation in progress and is not stable (UPDATE_IN_PROGRESS)',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 0/5 | 11:31:12 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | MyCustomResources.NestedStack/MyCustomResources.NestedStackResource (MyCustomResourcesBB610599)',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'MyCustomResourcesBB610599-UPDATE_IN_PROGRESS-2025-03-11T03:31:12.338Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'MyCustomResourcesBB610599',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:12.338'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec68c0d34b-36e7-4663-a2f2-9200a5f44eef',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'MyCustomResourcesBB610599',
                },
                constructPath: 'MyCustomResources.NestedStack/MyCustomResources.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 0,
                formatted: '0/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 0/5 | 11:31:12 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | auth.NestedStack/auth.NestedStackResource (auth179371D7)',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'auth179371D7-UPDATE_IN_PROGRESS-2025-03-11T03:31:12.380Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'auth179371D7',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:12.380'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/d00ac45dd4a60a3d170b74ff85c881d0e4359cbbc3de10dd76a4256db540dd2e.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec68c0d34b-36e7-4663-a2f2-9200a5f44eef',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'auth179371D7',
                },
                constructPath: 'auth.NestedStack/auth.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 0,
                formatted: '0/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV | 0/5 | 11:31:12 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV User Initiated',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                EventId: '4820d440-fe29-11ef-8435-0ac0e5399efb',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV',
                LogicalResourceId: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:12.895'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceStatusReason: 'User Initiated',
            },
            progress: {
                total: 5,
                completed: 0,
                formatted: '0/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 1/5 | 11:31:12 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | MyCustomResources.NestedStack/MyCustomResources.NestedStackResource (MyCustomResourcesBB610599)',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'MyCustomResourcesBB610599-UPDATE_COMPLETE-2025-03-11T03:31:12.930Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'MyCustomResourcesBB610599',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:12.930'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec68c0d34b-36e7-4663-a2f2-9200a5f44eef',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'MyCustomResourcesBB610599',
                },
                constructPath: 'MyCustomResources.NestedStack/MyCustomResources.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 1,
                formatted: '1/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV | 1/5 | 11:31:15 PM | UPDATE_IN_PROGRESS | AWS::Cognito::UserPool | auth/amplifyAuth/UserPool (amplifyAuthUserPool4BA7F805)',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                EventId: 'amplifyAuthUserPool4BA7F805-UPDATE_IN_PROGRESS-2025-03-11T03:31:15.774Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV',
                LogicalResourceId: 'amplifyAuthUserPool4BA7F805',
                PhysicalResourceId: 'us-west-2_j6BNFhRJd',
                ResourceType: 'AWS::Cognito::UserPool',
                Timestamp: new Date('2025-03-11T03:31:15.774'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceProperties: '{"UserPoolTags":{"amplify:deployment-type":"sandbox","amplify:friendly-name":"amplifyAuth","created-by":"amplify"},"Policies":{"PasswordPolicy":{"RequireNumbers":"true","MinimumLength":"8","RequireUppercase":"true","RequireLowercase":"true","RequireSymbols":"true"}},"VerificationMessageTemplate":{"EmailMessage":"The verification code to your new account is {####}","SmsMessage":"The verification code to your new account is {####}","EmailSubject":"Verify your new account","DefaultEmailOption":"CONFIRM_WITH_CODE"},"Schema":[{"Mutable":"true","Required":"true","Name":"email"}],"AdminCreateUserConfig":{"AllowAdminCreateUserOnly":"false"},"UsernameConfiguration":{"CaseSensitive":"false"},"SmsVerificationMessage":"The verification code to your new account is {####}","UserAttributeUpdateSettings":{"AttributesRequireVerificationBeforeUpdate":["email"]},"EmailConfiguration":{"EmailSendingAccount":"DEVELOPER","SourceArn":"arn:aws:ses:us-west-2:123456789012:identity/blah@blah.com","From":"blah@blah.com"},"EmailVerificationSubject":"Verify your new account","AutoVerifiedAttributes":["email"],"EmailVerificationMessage":"The verification code to your new account is {####}","UsernameAttributes":["email"],"AccountRecoverySetting":{"RecoveryMechanisms":[{"Priority":"1","Name":"verified_email"}]}}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyAuthUserPool4BA7F805',
                },
                constructPath: 'auth/amplifyAuth/UserPool',
            },
            progress: {
                total: 5,
                completed: 1,
                formatted: '1/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV | 1/5 | 11:31:17 PM | UPDATE_FAILED | AWS::Cognito::UserPool | auth/amplifyAuth/UserPool (amplifyAuthUserPool4BA7F805) Resource handler returned message: "Cognito received the following error from Amazon SES when attempting to send email: Email address is not verified. The following identities failed the check in region US-WEST-2: arn:aws:ses:us-west-2:123456789012:identity/blah@blah.com (Service: CognitoIdentityProvider, Status Code: 400, Request ID: 3bf35a6e-9667-4baf-8eab-19676643ac8d)" (RequestToken: 8c4cec48-3f40-7c73-a8b6-3afc0314b029, HandlerErrorCode: InvalidRequest)',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                EventId: 'amplifyAuthUserPool4BA7F805-UPDATE_FAILED-2025-03-11T03:31:17.316Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV',
                LogicalResourceId: 'amplifyAuthUserPool4BA7F805',
                PhysicalResourceId: 'us-west-2_j6BNFhRJd',
                ResourceType: 'AWS::Cognito::UserPool',
                Timestamp: new Date('2025-03-11T03:31:17.316'),
                ResourceStatus: 'UPDATE_FAILED',
                ResourceStatusReason: 'Resource handler returned message: "Cognito received the following error from Amazon SES when attempting to send email: Email address is not verified. The following identities failed the check in region US-WEST-2: arn:aws:ses:us-west-2:123456789012:identity/blah@blah.com (Service: CognitoIdentityProvider, Status Code: 400, Request ID: 3bf35a6e-9667-4baf-8eab-19676643ac8d)" (RequestToken: 8c4cec48-3f40-7c73-a8b6-3afc0314b029, HandlerErrorCode: InvalidRequest)',
                ResourceProperties: '{"UserPoolTags":{"amplify:deployment-type":"sandbox","amplify:friendly-name":"amplifyAuth","created-by":"amplify"},"Policies":{"PasswordPolicy":{"RequireNumbers":"true","MinimumLength":"8","RequireUppercase":"true","RequireLowercase":"true","RequireSymbols":"true"}},"VerificationMessageTemplate":{"EmailMessage":"The verification code to your new account is {####}","SmsMessage":"The verification code to your new account is {####}","EmailSubject":"Verify your new account","DefaultEmailOption":"CONFIRM_WITH_CODE"},"Schema":[{"Mutable":"true","Required":"true","Name":"email"}],"AdminCreateUserConfig":{"AllowAdminCreateUserOnly":"false"},"UsernameConfiguration":{"CaseSensitive":"false"},"SmsVerificationMessage":"The verification code to your new account is {####}","UserAttributeUpdateSettings":{"AttributesRequireVerificationBeforeUpdate":["email"]},"EmailConfiguration":{"EmailSendingAccount":"DEVELOPER","SourceArn":"arn:aws:ses:us-west-2:123456789012:identity/blah@blah.com","From":"blah@blah.com"},"EmailVerificationSubject":"Verify your new account","AutoVerifiedAttributes":["email"],"EmailVerificationMessage":"The verification code to your new account is {####}","UsernameAttributes":["email"],"AccountRecoverySetting":{"RecoveryMechanisms":[{"Priority":"1","Name":"verified_email"}]}}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyAuthUserPool4BA7F805',
                },
                constructPath: 'auth/amplifyAuth/UserPool',
            },
            progress: {
                total: 5,
                completed: 1,
                formatted: '1/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV | 1/5 | 11:31:17 PM | UPDATE_ROLLBACK_IN_PROGRESS | AWS::CloudFormation::Stack | amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV The following resource(s) failed to update: [amplifyAuthUserPool4BA7F805].',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                EventId: '4af48db0-fe29-11ef-8028-0294f79708b3',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV',
                LogicalResourceId: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:17.634'),
                ResourceStatus: 'UPDATE_ROLLBACK_IN_PROGRESS',
                ResourceStatusReason: 'The following resource(s) failed to update: [amplifyAuthUserPool4BA7F805]. ',
            },
            progress: {
                total: 5,
                completed: 1,
                formatted: '1/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has an ongoing operation in progress and is not stable (UPDATE_IN_PROGRESS)',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has an ongoing operation in progress and is not stable (UPDATE_ROLLBACK_IN_PROGRESS)',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 1/5 | 11:31:23 PM | UPDATE_FAILED | AWS::CloudFormation::Stack | auth.NestedStack/auth.NestedStackResource (auth179371D7) Embedded stack arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847 was not successfully updated. Currently in UPDATE_ROLLBACK_IN_PROGRESS with reason: The following resource(s) failed to update: [amplifyAuthUserPool4BA7F805].',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'auth179371D7-UPDATE_FAILED-2025-03-11T03:31:23.235Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'auth179371D7',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:23.235'),
                ResourceStatus: 'UPDATE_FAILED',
                ResourceStatusReason: 'Embedded stack arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847 was not successfully updated. Currently in UPDATE_ROLLBACK_IN_PROGRESS with reason: The following resource(s) failed to update: [amplifyAuthUserPool4BA7F805]. ',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/d00ac45dd4a60a3d170b74ff85c881d0e4359cbbc3de10dd76a4256db540dd2e.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec68c0d34b-36e7-4663-a2f2-9200a5f44eef',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'auth179371D7',
                },
                constructPath: 'auth.NestedStack/auth.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 1,
                formatted: '1/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 1/5 | 11:31:23 PM | UPDATE_ROLLBACK_IN_PROGRESS | AWS::CloudFormation::Stack | amplify-sampleprojectapp-user-sandbox-0e1752a889 The following resource(s) failed to update: [auth179371D7].',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: '4e7af550-fe29-11ef-a863-0afbc44afa35',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:23.538'),
                ResourceStatus: 'UPDATE_ROLLBACK_IN_PROGRESS',
                ResourceStatusReason: 'The following resource(s) failed to update: [auth179371D7]. ',
                ClientRequestToken: 'exec68c0d34b-36e7-4663-a2f2-9200a5f44eef',
            },
            progress: {
                total: 5,
                completed: 1,
                formatted: '1/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 1/5 | 11:31:25 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | auth.NestedStack/auth.NestedStackResource (auth179371D7)',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'auth179371D7-UPDATE_IN_PROGRESS-2025-03-11T03:31:25.602Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'auth179371D7',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:25.602'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/53b8945b363f121c80e07e812f42fa6e33a8f3fdbc9b75d4b3374959bb62f47d.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec68c0d34b-36e7-4663-a2f2-9200a5f44eef',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'auth179371D7',
                },
                constructPath: 'auth.NestedStack/auth.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 1,
                formatted: '1/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 1/5 | 11:31:25 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | MyCustomResources.NestedStack/MyCustomResources.NestedStackResource (MyCustomResourcesBB610599)',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'MyCustomResourcesBB610599-UPDATE_IN_PROGRESS-2025-03-11T03:31:25.625Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'MyCustomResourcesBB610599',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:25.625'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec68c0d34b-36e7-4663-a2f2-9200a5f44eef',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'MyCustomResourcesBB610599',
                },
                constructPath: 'MyCustomResources.NestedStack/MyCustomResources.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 1,
                formatted: '1/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 0/5 | 11:31:25 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | MyCustomResources.NestedStack/MyCustomResources.NestedStackResource (MyCustomResourcesBB610599)',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'MyCustomResourcesBB610599-UPDATE_COMPLETE-2025-03-11T03:31:25.903Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'MyCustomResourcesBB610599',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:25.903'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec68c0d34b-36e7-4663-a2f2-9200a5f44eef',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'MyCustomResourcesBB610599',
                },
                constructPath: 'MyCustomResources.NestedStack/MyCustomResources.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 0,
                formatted: '0/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has an ongoing operation in progress and is not stable (UPDATE_ROLLBACK_IN_PROGRESS)',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV | 0/5 | 11:31:29 PM | UPDATE_IN_PROGRESS | AWS::Cognito::UserPool | auth/amplifyAuth/UserPool (amplifyAuthUserPool4BA7F805)',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                EventId: 'amplifyAuthUserPool4BA7F805-UPDATE_IN_PROGRESS-2025-03-11T03:31:29.693Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV',
                LogicalResourceId: 'amplifyAuthUserPool4BA7F805',
                PhysicalResourceId: 'us-west-2_j6BNFhRJd',
                ResourceType: 'AWS::Cognito::UserPool',
                Timestamp: new Date('2025-03-11T03:31:29.693'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceProperties: '{"UserPoolTags":{"amplify:deployment-type":"sandbox","amplify:friendly-name":"amplifyAuth","created-by":"amplify"},"Policies":{"PasswordPolicy":{"RequireNumbers":"true","MinimumLength":"8","RequireUppercase":"true","RequireLowercase":"true","RequireSymbols":"true"}},"VerificationMessageTemplate":{"EmailMessage":"The verification code to your new account is {####}","SmsMessage":"The verification code to your new account is {####}","EmailSubject":"Verify your new account","DefaultEmailOption":"CONFIRM_WITH_CODE"},"Schema":[{"Mutable":"true","Required":"true","Name":"email"}],"AdminCreateUserConfig":{"AllowAdminCreateUserOnly":"false"},"UsernameConfiguration":{"CaseSensitive":"false"},"SmsVerificationMessage":"The verification code to your new account is {####}","UserAttributeUpdateSettings":{"AttributesRequireVerificationBeforeUpdate":["email"]},"EmailVerificationSubject":"Verify your new account","AutoVerifiedAttributes":["email"],"EmailVerificationMessage":"The verification code to your new account is {####}","UsernameAttributes":["email"],"AccountRecoverySetting":{"RecoveryMechanisms":[{"Priority":"1","Name":"verified_email"}]}}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyAuthUserPool4BA7F805',
                },
                constructPath: 'auth/amplifyAuth/UserPool',
            },
            progress: {
                total: 5,
                completed: 0,
                formatted: '0/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV | 1/5 | 11:31:31 PM | UPDATE_COMPLETE | AWS::Cognito::UserPool | auth/amplifyAuth/UserPool (amplifyAuthUserPool4BA7F805)',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                EventId: 'amplifyAuthUserPool4BA7F805-UPDATE_COMPLETE-2025-03-11T03:31:31.647Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV',
                LogicalResourceId: 'amplifyAuthUserPool4BA7F805',
                PhysicalResourceId: 'us-west-2_j6BNFhRJd',
                ResourceType: 'AWS::Cognito::UserPool',
                Timestamp: new Date('2025-03-11T03:31:31.647'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"UserPoolTags":{"amplify:deployment-type":"sandbox","amplify:friendly-name":"amplifyAuth","created-by":"amplify"},"Policies":{"PasswordPolicy":{"RequireNumbers":"true","MinimumLength":"8","RequireUppercase":"true","RequireLowercase":"true","RequireSymbols":"true"}},"VerificationMessageTemplate":{"EmailMessage":"The verification code to your new account is {####}","SmsMessage":"The verification code to your new account is {####}","EmailSubject":"Verify your new account","DefaultEmailOption":"CONFIRM_WITH_CODE"},"Schema":[{"Mutable":"true","Required":"true","Name":"email"}],"AdminCreateUserConfig":{"AllowAdminCreateUserOnly":"false"},"UsernameConfiguration":{"CaseSensitive":"false"},"SmsVerificationMessage":"The verification code to your new account is {####}","UserAttributeUpdateSettings":{"AttributesRequireVerificationBeforeUpdate":["email"]},"EmailVerificationSubject":"Verify your new account","AutoVerifiedAttributes":["email"],"EmailVerificationMessage":"The verification code to your new account is {####}","UsernameAttributes":["email"],"AccountRecoverySetting":{"RecoveryMechanisms":[{"Priority":"1","Name":"verified_email"}]}}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyAuthUserPool4BA7F805',
                },
                constructPath: 'auth/amplifyAuth/UserPool',
            },
            progress: {
                total: 5,
                completed: 1,
                formatted: '1/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV | 2/5 | 11:31:33 PM | UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS | AWS::CloudFormation::Stack | amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                EventId: '5462f3a0-fe29-11ef-b336-0214c1e7ba71',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV',
                LogicalResourceId: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:33.457'),
                ResourceStatus: 'UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS',
            },
            progress: {
                total: 5,
                completed: 2,
                formatted: '2/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has an ongoing operation in progress and is not stable (UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS)',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 3/5 | 11:31:36 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | auth.NestedStack/auth.NestedStackResource (auth179371D7)',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'auth179371D7-UPDATE_COMPLETE-2025-03-11T03:31:36.241Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'auth179371D7',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:36.241'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/53b8945b363f121c80e07e812f42fa6e33a8f3fdbc9b75d4b3374959bb62f47d.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec68c0d34b-36e7-4663-a2f2-9200a5f44eef',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'auth179371D7',
                },
                constructPath: 'auth.NestedStack/auth.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 3,
                formatted: '3/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 4/5 | 11:31:36 PM | UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS | AWS::CloudFormation::Stack | amplify-sampleprojectapp-user-sandbox-0e1752a889',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: '565c2b90-fe29-11ef-a2fa-026a2e7f24a7',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:36.769'),
                ResourceStatus: 'UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS',
                ClientRequestToken: 'exec68c0d34b-36e7-4663-a2f2-9200a5f44eef',
            },
            progress: {
                total: 5,
                completed: 4,
                formatted: '4/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 3/5 | 11:31:37 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | MyCustomResources.NestedStack/MyCustomResources.NestedStackResource (MyCustomResourcesBB610599)',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'MyCustomResourcesBB610599-UPDATE_COMPLETE-2025-03-11T03:31:37.807Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'MyCustomResourcesBB610599',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:37.807'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec68c0d34b-36e7-4663-a2f2-9200a5f44eef',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'MyCustomResourcesBB610599',
                },
                constructPath: 'MyCustomResources.NestedStack/MyCustomResources.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 3,
                formatted: '3/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV | 4/5 | 11:31:38 PM | UPDATE_ROLLBACK_COMPLETE | AWS::CloudFormation::Stack | amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                EventId: '57917740-fe29-11ef-af0c-06596f54ec5d',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV',
                LogicalResourceId: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:38.791'),
                ResourceStatus: 'UPDATE_ROLLBACK_COMPLETE',
            },
            progress: {
                total: 5,
                completed: 4,
                formatted: '4/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has an ongoing operation in progress and is not stable (UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS)',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has an ongoing operation in progress and is not stable (UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS)',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 3/5 | 11:31:47 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | auth.NestedStack/auth.NestedStackResource (auth179371D7)',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'auth179371D7-UPDATE_COMPLETE-2025-03-11T03:31:47.995Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'auth179371D7',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:47.995'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/53b8945b363f121c80e07e812f42fa6e33a8f3fdbc9b75d4b3374959bb62f47d.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec68c0d34b-36e7-4663-a2f2-9200a5f44eef',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'auth179371D7',
                },
                constructPath: 'auth.NestedStack/auth.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 3,
                formatted: '3/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 4/5 | 11:31:48 PM | UPDATE_ROLLBACK_COMPLETE | AWS::CloudFormation::Stack | amplify-sampleprojectapp-user-sandbox-0e1752a889',
        data: {
            deployment: 'aa821bf0-c30e-4f6b-83a8-39d89479127f',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: '5d27fa30-fe29-11ef-b5db-06aa5705cd77',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:31:48.167'),
                ResourceStatus: 'UPDATE_ROLLBACK_COMPLETE',
                ClientRequestToken: 'exec68c0d34b-36e7-4663-a2f2-9200a5f44eef',
            },
            progress: {
                total: 5,
                completed: 4,
                formatted: '4/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5503',
        message: 'Completed amplify-sampleprojectapp-user-sandbox-0e1752a889',
    },
];
