// @aws-amplify/cli-core/lib/loggers/test-assets/hotswap_lambda_data_resources_cdk_events.js
export const data = [
    {
        code: 'SYNTH_STARTED',
        action: 'amplify',
        message: 'Backend synthesis started',
    },
    {
        code: 'CDK_TOOLKIT_I1001',
        message: 'Starting Synthesis ...',
    },
    {
        code: 'CDK_ASSEMBLY_I0010',
        message: 'Setting "CDK_DEFAULT_REGION" environment variable to us-west-2',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Looking up default account ID from STS',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] STS.GetCallerIdentity({}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Default account ID: 123456789012',
    },
    {
        code: 'CDK_ASSEMBLY_I0010',
        message: 'Setting "CDK_DEFAULT_ACCOUNT" environment variable to 123456789012',
    },
    {
        code: 'CDK_ASSEMBLY_I0010',
        message: 'outdir: /Users/user/Workspaces/amplify-cdk-toolkit/sample-samsara-app/.amplify/artifacts/cdk.out',
    },
    {
        code: 'CDK_ASSEMBLY_I0010',
        message: "env: {\n  CDK_DEFAULT_REGION: 'us-west-2',\n  CDK_DEFAULT_ACCOUNT: '123456789012',\n  CDK_OUTDIR: '/Users/user/Workspaces/amplify-cdk-toolkit/sample-samsara-app/.amplify/artifacts/cdk.out',\n  CDK_CLI_ASM_VERSION: '43.0.0',\n  CDK_CLI_VERSION: '0.1.8'\n}",
    },
    {
        code: 'CDK_ASSEMBLY_I0000',
        message: "context: {\n  'amplify-backend-namespace': 'sample-samsara-app',\n  'amplify-backend-name': 'user',\n  'amplify-backend-type': 'sandbox',\n  'aws:cdk:enable-path-metadata': true,\n  'aws:cdk:enable-asset-metadata': true,\n  'aws:cdk:version-reporting': true,\n  'aws:cdk:bundling-stacks': [ '**' ]\n}",
    },
    {
        code: 'CDK_TOOLKIT_I1000',
        message: '  Synthesis time: 2.67s',
    },
    {
        code: 'CDK_TOOLKIT_I1901',
        message: 'Successfully synthesized to /Users/user/Workspaces/amplify-cdk-toolkit/sample-samsara-app/.amplify/artifacts/cdk.out',
    },
    {
        code: 'SYNTH_FINISHED',
        action: 'amplify',
        level: 'result',
        message: 'Backend synthesized in 2.68 seconds',
    },
    {
        code: 'TS_STARTED',
        action: 'amplify',
        message: 'Backend type checks started',
    },
    {
        code: 'TS_FINISHED',
        action: 'amplify',
        level: 'result',
        message: 'Type checks completed in 5.98 seconds',
    },
    {
        code: 'CDK_TOOLKIT_I1001',
        message: 'Starting Synthesis ...',
    },
    {
        code: 'CDK_TOOLKIT_I1000',
        message: '  Synthesis time: 0s',
    },
    {
        code: 'CDK_TOOLKIT_W5400',
        message: '⚠️ The --hotswap and --hotswap-fallback flags deliberately introduce CloudFormation drift to speed up deployments\n⚠️ They should only be used for development - never use them for your production Stacks!',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Checking for previously published assets',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/88d5b7081b0dc9c39fd0bffedf56f011b2a7cf1ce5cc227f47a1e94efee8067c.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/4c6a2d29f01c6091bd1d9afe16e5849d456c96f17c3b215938c8067399532719.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/9fcbe070ecd3023c5bf5b966fa9584757db9762eef123bad0820bd87591b2174.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/08f4d557693d96c1a4efba0f9dc91330e4b19772fd5477c156468843e3d9cb5e.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/1f7e4e867309fea8e686c6b2efd8d07ad8f06818fb62f26bb5d6ff80ffcf4bbb.zip',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/cc01911d0269d4080ea57505dc445dfc315ef7ad85d3d9d4ea1357858bff451d.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/a183ddccbd956316c38ef97177b8f088ef0826f62023323f5ae6053d348ccffc.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c1721bcd774e27c514d3454b5be4f9bdd094c0161b57ddf053d618e3b0086a77.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"08f4d557693d96c1a4efba0f9dc91330e4b19772fd5477c156468843e3d9cb5e.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/08f4d557693d96c1a4efba0f9dc91330e4b19772fd5477c156468843e3d9cb5e.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"88d5b7081b0dc9c39fd0bffedf56f011b2a7cf1ce5cc227f47a1e94efee8067c.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/88d5b7081b0dc9c39fd0bffedf56f011b2a7cf1ce5cc227f47a1e94efee8067c.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"cc01911d0269d4080ea57505dc445dfc315ef7ad85d3d9d4ea1357858bff451d.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/cc01911d0269d4080ea57505dc445dfc315ef7ad85d3d9d4ea1357858bff451d.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"a183ddccbd956316c38ef97177b8f088ef0826f62023323f5ae6053d348ccffc.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/a183ddccbd956316c38ef97177b8f088ef0826f62023323f5ae6053d348ccffc.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"4c6a2d29f01c6091bd1d9afe16e5849d456c96f17c3b215938c8067399532719.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/4c6a2d29f01c6091bd1d9afe16e5849d456c96f17c3b215938c8067399532719.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"1f7e4e867309fea8e686c6b2efd8d07ad8f06818fb62f26bb5d6ff80ffcf4bbb.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/1f7e4e867309fea8e686c6b2efd8d07ad8f06818fb62f26bb5d6ff80ffcf4bbb.zip',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c1721bcd774e27c514d3454b5be4f9bdd094c0161b57ddf053d618e3b0086a77.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c1721bcd774e27c514d3454b5be4f9bdd094c0161b57ddf053d618e3b0086a77.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"9fcbe070ecd3023c5bf5b966fa9584757db9762eef123bad0820bd87591b2174.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/9fcbe070ecd3023c5bf5b966fa9584757db9762eef123bad0820bd87591b2174.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/b371d68ddfbdd11b176756ce33b2c4e79d26fcc2fd4b7c0c9e75d06efacee90e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/47614d6b888488252a8a44b7911b69efcddb1eb2454ca31132d72d6444247204.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/f4a52b72209a9dfa197b5e7367a5c378c5bb86de6e29ddd9e48b49a3fe54b249.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/06db846fd14e6fc371f22b12b5545ba8e2dbfeda85d8c8d586c71c282166657b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/1f5fed297da9c32ae3af922bf3a38ccf23b956078887d16891ec06c20c64722c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/245ed8ac5185462a6923b7e45a318fec4c9ba0c45d1d720abd5b20317aa3dc6a.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/474bf0776ec2164a13191d1a0a9e057154931e4918fea5086f49850d02a5371b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/29e4bdb03133f87f2b1acb547959ffd4cfd22804026994c5b61db33084127353.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"29e4bdb03133f87f2b1acb547959ffd4cfd22804026994c5b61db33084127353.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/29e4bdb03133f87f2b1acb547959ffd4cfd22804026994c5b61db33084127353.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"b371d68ddfbdd11b176756ce33b2c4e79d26fcc2fd4b7c0c9e75d06efacee90e.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/b371d68ddfbdd11b176756ce33b2c4e79d26fcc2fd4b7c0c9e75d06efacee90e.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"06db846fd14e6fc371f22b12b5545ba8e2dbfeda85d8c8d586c71c282166657b.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/06db846fd14e6fc371f22b12b5545ba8e2dbfeda85d8c8d586c71c282166657b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/4f7907d1209a2c9953a0c053df402c634e359546d70c7cc5c2e8e21ea734880f.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/6f3a2a9d562ae3058cc1aed6741edbf5edcaef7b3df7b03f52478816a1c95ffb.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/fe3c43ada4b9d681a5e2312663ef7a73386424d73b73e51f8e2e9d4b50f7c502.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"474bf0776ec2164a13191d1a0a9e057154931e4918fea5086f49850d02a5371b.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/474bf0776ec2164a13191d1a0a9e057154931e4918fea5086f49850d02a5371b.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"1f5fed297da9c32ae3af922bf3a38ccf23b956078887d16891ec06c20c64722c.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/1f5fed297da9c32ae3af922bf3a38ccf23b956078887d16891ec06c20c64722c.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"f4a52b72209a9dfa197b5e7367a5c378c5bb86de6e29ddd9e48b49a3fe54b249.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/f4a52b72209a9dfa197b5e7367a5c378c5bb86de6e29ddd9e48b49a3fe54b249.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"47614d6b888488252a8a44b7911b69efcddb1eb2454ca31132d72d6444247204.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/47614d6b888488252a8a44b7911b69efcddb1eb2454ca31132d72d6444247204.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"245ed8ac5185462a6923b7e45a318fec4c9ba0c45d1d720abd5b20317aa3dc6a.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/245ed8ac5185462a6923b7e45a318fec4c9ba0c45d1d720abd5b20317aa3dc6a.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/e0cff47fb007f0bbf2a4e43ca256d6aa7ec109821769fd79fa7c5e83f0e7f9fc.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/6479af76f5a4d8b61b53dd3cee8e18ae192b629ad15ced7b342690a9ebbdd8c2.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/6ce8d75a422db2491407ccfc9807e5de43e86bed091909b805008adb8243b05d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/3cf32e28e3985b88b598dacbcc9a81cbb12cab62b6507082eb20a0702d3c197e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/fface358117b1edafc4dbe09854c750ff9d1cad2ab9aed9cff18b5df478d3d5d.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"4f7907d1209a2c9953a0c053df402c634e359546d70c7cc5c2e8e21ea734880f.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/4f7907d1209a2c9953a0c053df402c634e359546d70c7cc5c2e8e21ea734880f.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/725b7ed3971927a1e99c853643b4ca575c3d612fff9b50d3f616c4feeb79a7d7.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"3cf32e28e3985b88b598dacbcc9a81cbb12cab62b6507082eb20a0702d3c197e.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/3cf32e28e3985b88b598dacbcc9a81cbb12cab62b6507082eb20a0702d3c197e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/97e4049e52fffaf523b49966a1225677c0fd7c09e296e1832b5bb95086e5c3e4.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"e0cff47fb007f0bbf2a4e43ca256d6aa7ec109821769fd79fa7c5e83f0e7f9fc.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/e0cff47fb007f0bbf2a4e43ca256d6aa7ec109821769fd79fa7c5e83f0e7f9fc.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"fface358117b1edafc4dbe09854c750ff9d1cad2ab9aed9cff18b5df478d3d5d.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/fface358117b1edafc4dbe09854c750ff9d1cad2ab9aed9cff18b5df478d3d5d.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"6479af76f5a4d8b61b53dd3cee8e18ae192b629ad15ced7b342690a9ebbdd8c2.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/6479af76f5a4d8b61b53dd3cee8e18ae192b629ad15ced7b342690a9ebbdd8c2.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"6ce8d75a422db2491407ccfc9807e5de43e86bed091909b805008adb8243b05d.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/6ce8d75a422db2491407ccfc9807e5de43e86bed091909b805008adb8243b05d.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"6f3a2a9d562ae3058cc1aed6741edbf5edcaef7b3df7b03f52478816a1c95ffb.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/6f3a2a9d562ae3058cc1aed6741edbf5edcaef7b3df7b03f52478816a1c95ffb.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"fe3c43ada4b9d681a5e2312663ef7a73386424d73b73e51f8e2e9d4b50f7c502.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/fe3c43ada4b9d681a5e2312663ef7a73386424d73b73e51f8e2e9d4b50f7c502.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c89821d3a34d8f9a69feb7faded2de554cfee32627d772bc8d842a1a145408f5.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/098c998f215c333f57348ca8912f66057ae6b574fe34013033dc54362ee2372c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/5b2dd26ac19f6a54684f2769d465463159cc27a441d7ac8b8efafdec1b2370fd.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/db826cc870f25dcd05aff645c3b384ba3a97115acc5373e580c14379eb0430b3.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/49e7a1f2fb4c809b82bed6b0242a2a3bcfa2c5276ebd2539ff330fee36f8fe18.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c3534669fc2486ce8786b1a3efd6c0848cd3c4835de33940c0f49c28c9d8abef.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"725b7ed3971927a1e99c853643b4ca575c3d612fff9b50d3f616c4feeb79a7d7.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/725b7ed3971927a1e99c853643b4ca575c3d612fff9b50d3f616c4feeb79a7d7.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/dcb70e4856bc9a423e02b8b9e9c2ab3f71479784572d4c04a02d1f5b61a9f9ac.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"97e4049e52fffaf523b49966a1225677c0fd7c09e296e1832b5bb95086e5c3e4.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/97e4049e52fffaf523b49966a1225677c0fd7c09e296e1832b5bb95086e5c3e4.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/faa95a81ae7d7373f3e1f242268f904eb748d8d0fdd306e8a6fe515a1905a7d6.zip',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"db826cc870f25dcd05aff645c3b384ba3a97115acc5373e580c14379eb0430b3.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/db826cc870f25dcd05aff645c3b384ba3a97115acc5373e580c14379eb0430b3.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c89821d3a34d8f9a69feb7faded2de554cfee32627d772bc8d842a1a145408f5.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c89821d3a34d8f9a69feb7faded2de554cfee32627d772bc8d842a1a145408f5.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/12157b27d30ab71eb24ae65825f672ba5cc2c09dbb1703cd7adfcff3aeaca136.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/4fe0aba5e672b596d0f72505a9eec502f98d46906bb30fae2511fbdc1df4956f.zip',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c3534669fc2486ce8786b1a3efd6c0848cd3c4835de33940c0f49c28c9d8abef.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c3534669fc2486ce8786b1a3efd6c0848cd3c4835de33940c0f49c28c9d8abef.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"49e7a1f2fb4c809b82bed6b0242a2a3bcfa2c5276ebd2539ff330fee36f8fe18.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/49e7a1f2fb4c809b82bed6b0242a2a3bcfa2c5276ebd2539ff330fee36f8fe18.vtl',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"098c998f215c333f57348ca8912f66057ae6b574fe34013033dc54362ee2372c.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/098c998f215c333f57348ca8912f66057ae6b574fe34013033dc54362ee2372c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/2f050e6b9affea69843658b09b5d9073acc9cda9170a30ff043def78a33c6ad0.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/907cb295088857ccb545b8aae83288bd9dfa82115bb0d20a0ae0d08dc8872f8a.zip',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"dcb70e4856bc9a423e02b8b9e9c2ab3f71479784572d4c04a02d1f5b61a9f9ac.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/dcb70e4856bc9a423e02b8b9e9c2ab3f71479784572d4c04a02d1f5b61a9f9ac.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"5b2dd26ac19f6a54684f2769d465463159cc27a441d7ac8b8efafdec1b2370fd.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/5b2dd26ac19f6a54684f2769d465463159cc27a441d7ac8b8efafdec1b2370fd.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/d9872cd69fe68872fa841134bdd56f3f93f86090760bb649a3f2a4b4c34cc68f.json',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"faa95a81ae7d7373f3e1f242268f904eb748d8d0fdd306e8a6fe515a1905a7d6.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/faa95a81ae7d7373f3e1f242268f904eb748d8d0fdd306e8a6fe515a1905a7d6.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/aeee25d4b9e20f541bebb704a03982cd2f766d6f316c874393d02cfa6accbe2d.json',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"4fe0aba5e672b596d0f72505a9eec502f98d46906bb30fae2511fbdc1df4956f.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/4fe0aba5e672b596d0f72505a9eec502f98d46906bb30fae2511fbdc1df4956f.zip',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"2f050e6b9affea69843658b09b5d9073acc9cda9170a30ff043def78a33c6ad0.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/2f050e6b9affea69843658b09b5d9073acc9cda9170a30ff043def78a33c6ad0.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/822858b4a7630f1af10bc11c608fccbe16b194fc4064da128abb4e1caaf3709d.json',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"12157b27d30ab71eb24ae65825f672ba5cc2c09dbb1703cd7adfcff3aeaca136.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/12157b27d30ab71eb24ae65825f672ba5cc2c09dbb1703cd7adfcff3aeaca136.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/0805d28c048e2f7229795dd93757c7715bc5e720fdb92d3699d973f22df53dcb.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/20c585e2b8651025809e472021cb72b8518d98debfae6d9b2a55bbe4147dd34d.json',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"907cb295088857ccb545b8aae83288bd9dfa82115bb0d20a0ae0d08dc8872f8a.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/907cb295088857ccb545b8aae83288bd9dfa82115bb0d20a0ae0d08dc8872f8a.zip',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/2433702cd2f34d8c493b92f40674cee7f2cc9544534a44a7e7f665bf7e3a5de5.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/e7d2383507cabbe208eb6d8f9a9109362a23973b7179f726f6580b6e2c2be4fb.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/135a3bdaa54a2c2bf8336b5ad01e0d15301ea2c807ea9b259fa3a631212bf369.json',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"d9872cd69fe68872fa841134bdd56f3f93f86090760bb649a3f2a4b4c34cc68f.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/d9872cd69fe68872fa841134bdd56f3f93f86090760bb649a3f2a4b4c34cc68f.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/a508c871807468496d9b0c94bded0b80979e5ccfdddb756a6453cef49e0d0262.json',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"aeee25d4b9e20f541bebb704a03982cd2f766d6f316c874393d02cfa6accbe2d.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/aeee25d4b9e20f541bebb704a03982cd2f766d6f316c874393d02cfa6accbe2d.json',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"0805d28c048e2f7229795dd93757c7715bc5e720fdb92d3699d973f22df53dcb.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/0805d28c048e2f7229795dd93757c7715bc5e720fdb92d3699d973f22df53dcb.json',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"20c585e2b8651025809e472021cb72b8518d98debfae6d9b2a55bbe4147dd34d.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/20c585e2b8651025809e472021cb72b8518d98debfae6d9b2a55bbe4147dd34d.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/3903194462b0b595c18667e5a3b281fbd78a51a1e18974f997465ca079e889cb.json',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"822858b4a7630f1af10bc11c608fccbe16b194fc4064da128abb4e1caaf3709d.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/822858b4a7630f1af10bc11c608fccbe16b194fc4064da128abb4e1caaf3709d.json',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"2433702cd2f34d8c493b92f40674cee7f2cc9544534a44a7e7f665bf7e3a5de5.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/2433702cd2f34d8c493b92f40674cee7f2cc9544534a44a7e7f665bf7e3a5de5.json',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"e7d2383507cabbe208eb6d8f9a9109362a23973b7179f726f6580b6e2c2be4fb.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/e7d2383507cabbe208eb6d8f9a9109362a23973b7179f726f6580b6e2c2be4fb.json',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"135a3bdaa54a2c2bf8336b5ad01e0d15301ea2c807ea9b259fa3a631212bf369.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/135a3bdaa54a2c2bf8336b5ad01e0d15301ea2c807ea9b259fa3a631212bf369.json',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"a508c871807468496d9b0c94bded0b80979e5ccfdddb756a6453cef49e0d0262.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/a508c871807468496d9b0c94bded0b80979e5ccfdddb756a6453cef49e0d0262.json',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"3903194462b0b595c18667e5a3b281fbd78a51a1e18974f997465ca079e889cb.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/3903194462b0b595c18667e5a3b281fbd78a51a1e18974f997465ca079e889cb.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '50 total assets, 0 still need to be published',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Reading existing template for stack amplify-project-user-sandbox-0e1752a889.',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-lookup-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-lookup-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] SSM.GetParameter({"Name":"/cdk-bootstrap/hnb659fds/version"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-project-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-project-user-sandbox-0e1752a889","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5100',
        message: 'amplify-project-user-sandbox-0e1752a889: deploying... [1/1]',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-deploy-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-deploy-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-project-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: checking if we can skip deploy',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-project-user-sandbox-0e1752a889","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: template has changed',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-project-user-sandbox-0e1752a889: deploying...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Waiting for stack CDKToolkit to finish creating or updating...',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"CDKToolkit"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"CDKToolkit"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5400',
        message: 'Starting hotswap-deployment ...',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] STS.GetCallerIdentity({}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-project-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-project-user-sandbox-0e1752a889","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-project-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-project-user-sandbox-0e1752a889-auth179371D7-1LGC25T4CT602"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-project-user-sandbox-0e1752a889-auth179371D7-1LGC25T4CT602","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-project-user-sandbox-0e1752a889-data7552DF31-1V92HBC0Y8ZFP"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-project-user-sandbox-0e1752a889-data7552DF31-1V92HBC0Y8ZFP","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-project-user-sandbox-0e1752a889-data7552DF31-1V92HBC0Y8ZFP"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-project-user-sandbox-0-amplifyDataAmplifyTableManagerNestedStackA-1PTF9EU11VI02"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-project-user-sandbox-0-amplifyDataAmplifyTableManagerNestedStackA-1PTF9EU11VI02","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-project-user-sandbox-0-amplifyDataPostNestedStackPostNestedStackR-28DZXRDGX948"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-project-user-sandbox-0-amplifyDataPostNestedStackPostNestedStackR-28DZXRDGX948","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-project-user-sandbox-0-amplifyDataPersonNestedStackPersonNestedSt-VYUOJCN4YF0P"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-project-user-sandbox-0-amplifyDataPersonNestedStackPersonNestedSt-VYUOJCN4YF0P","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-project-user-sandbox-0-amplifyDataConnectionStackNestedStackConne-1EY2A6R5ES4ZZ"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-project-user-sandbox-0-amplifyDataConnectionStackNestedStackConne-1EY2A6R5ES4ZZ","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-project-user-sandbox-0e1752a889-function1351588B-NLR076HPHXG7"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-project-user-sandbox-0e1752a889-function1351588B-NLR076HPHXG7","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-project-user-sandbox-0e1752a889-storage0EC3F24A-1S3QC4E0FRJSK"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-project-user-sandbox-0e1752a889-storage0EC3F24A-1S3QC4E0FRJSK","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-project-user-sandbox-0e1752a889-MyCustomResourcesBB610599-1HJAUQ1UMP7MB"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-project-user-sandbox-0e1752a889-MyCustomResourcesBB610599-1HJAUQ1UMP7MB","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-project-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-project-user-sandbox-0e1752a889-auth179371D7-1LGC25T4CT602"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-project-user-sandbox-0e1752a889-auth179371D7-1LGC25T4CT602"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-project-user-sandbox-0e1752a889-auth179371D7-1LGC25T4CT602"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-project-user-sandbox-0e1752a889-auth179371D7-1LGC25T4CT602"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-project-user-sandbox-0e1752a889-data7552DF31-1V92HBC0Y8ZFP"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-project-user-sandbox-0e1752a889-function1351588B-NLR076HPHXG7"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5401',
        message: 'Hotswap plan created',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: ' hotswapping resources:',
    },
    {
        code: 'CDK_TOOLKIT_I5402',
        message: " AWS::AppSync::GraphQLSchema 'qf3pkl3rvjdlbccnohxt5gkdyyGraphQLSchema'",
    },
    {
        code: 'CDK_TOOLKIT_I5402',
        message: " Contents of AWS::S3::Bucket 'amplify-project--amplifydataamplifycodege-clxvnkfko9ki'",
    },
    {
        code: 'CDK_TOOLKIT_I5402',
        message: " Contents of AWS::S3::Bucket 'amplify-project--modelintrospectionschema-cwdah99w4vhc'",
    },
    {
        code: 'CDK_TOOLKIT_I5402',
        message: " AWS::Lambda::Function 'amplify-project-p-sayhelloslambda734D68D9-33dq3RGkPNTR'",
    },
    {
        code: 'CDK_TOOLKIT_I5402',
        message: " AWS::Lambda::Function 'amplify-project-user-firstlambda23886240-onnWijqFcbCB'",
    },
    {
        code: 'CDK_TOOLKIT_I5402',
        message: " AWS::Lambda::Function 'amplify-project-prav-secondlambdaBB16058D-qr2Pe32Olu4U'",
    },
    {
        code: 'CDK_TOOLKIT_I5402',
        message: " AWS::Lambda::Function 'amplify-project-user-thirdlambda3EC3C10B-J4fEZYAKBBf1'",
    },
    {
        code: 'CDK_TOOLKIT_I5402',
        message: " AWS::Lambda::Function 'amplify-project-prav-fourthlambda8DE196FD-OlYUvlKcriu6'",
    },
    {
        code: 'CDK_TOOLKIT_I5402',
        message: " AWS::Lambda::Function 'amplify-project-user-fifthlambda2306F097-jV2mMa7c6Ltg'",
    },
    {
        code: 'CDK_TOOLKIT_I5402',
        message: " AWS::Lambda::Function 'amplify-project-user-sixthlambda68A95826-MVDPZVKKJypK'",
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] S3.GetObject({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Key":"e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.UpdateFunctionCode({"FunctionName":"amplify-project-user-thirdlambda3EC3C10B-J4fEZYAKBBf1","S3Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","S3Key":"7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.UpdateFunctionCode({"FunctionName":"amplify-project-prav-secondlambdaBB16058D-qr2Pe32Olu4U","S3Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","S3Key":"7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.UpdateFunctionCode({"FunctionName":"amplify-project-user-sixthlambda68A95826-MVDPZVKKJypK","S3Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","S3Key":"7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.UpdateFunctionCode({"FunctionName":"amplify-project-user-firstlambda23886240-onnWijqFcbCB","S3Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","S3Key":"7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.UpdateFunctionCode({"FunctionName":"amplify-project-p-sayhelloslambda734D68D9-33dq3RGkPNTR","S3Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","S3Key":"7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.UpdateFunctionCode({"FunctionName":"amplify-project-prav-fourthlambda8DE196FD-OlYUvlKcriu6","S3Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","S3Key":"7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-prav-secondlambdaBB16058D-qr2Pe32Olu4U"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-user-thirdlambda3EC3C10B-J4fEZYAKBBf1"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-user-sixthlambda68A95826-MVDPZVKKJypK"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-user-firstlambda23886240-onnWijqFcbCB"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.UpdateFunctionCode({"FunctionName":"amplify-project-user-fifthlambda2306F097-jV2mMa7c6Ltg","S3Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","S3Key":"7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-p-sayhelloslambda734D68D9-33dq3RGkPNTR"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-prav-fourthlambda8DE196FD-OlYUvlKcriu6"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-user-fifthlambda2306F097-jV2mMa7c6Ltg"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] AppSync.StartSchemaCreation({"apiId":"qf3pkl3rvjdlbccnohxt5gkdyy","definition":"type Post @aws_cognito_user_pools @aws_iam {\\n  title: String!\\n  content: String!\\n  authorName: String\\n  authorDoB: AWSDate\\n  author: Person\\n  id: ID!\\n  createdAt: AWSDateTime!\\n  updatedAt: AWSDateTime!\\n}\\n\\ntype Person @aws_cognito_user_pools @aws_iam {\\n  name: String!\\n  dateOfBirth: AWSDate!\\n  authoredPosts(filter: ModelPostFilterInput, sortDirection: ModelSortDirection, limit: Int, nextToken: String): ModelPostConnection\\n  createdAt: AWSDateTime!\\n  updatedAt: AWSDateTime!\\n}\\n\\ninput AmplifyAIConversationTurnErrorInput {\\n  errorType: String!\\n  message: String!\\n}\\n\\ninput ModelStringInput {\\n  ne: String\\n  eq: String\\n  le: String\\n  lt: String\\n  ge: String\\n  gt: String\\n  contains: String\\n  notContains: String\\n  between: [String]\\n  beginsWith: String\\n  attributeExists: Boolean\\n  attributeType: ModelAttributeTypes\\n  size: ModelSizeInput\\n}\\n\\ninput ModelIntInput {\\n  ne: Int\\n  eq: Int\\n  le: Int\\n  lt: Int\\n  ge: Int\\n  gt: Int\\n  between: [Int]\\n  attributeExists: Boolean\\n  attributeType: ModelAttributeTypes\\n}\\n\\ninput ModelFloatInput {\\n  ne: Float\\n  eq: Float\\n  le: Float\\n  lt: Float\\n  ge: Float\\n  gt: Float\\n  between: [Float]\\n  attributeExists: Boolean\\n  attributeType: ModelAttributeTypes\\n}\\n\\ninput ModelBooleanInput {\\n  ne: Boolean\\n  eq: Boolean\\n  attributeExists: Boolean\\n  attributeType: ModelAttributeTypes\\n}\\n\\ninput ModelIDInput {\\n  ne: ID\\n  eq: ID\\n  le: ID\\n  lt: ID\\n  ge: ID\\n  gt: ID\\n  contains: ID\\n  notContains: ID\\n  between: [ID]\\n  beginsWith: ID\\n  attributeExists: Boolean\\n  attributeType: ModelAttributeTypes\\n  size: ModelSizeInput\\n}\\n\\ninput ModelSubscriptionStringInput {\\n  ne: String\\n  eq: String\\n  le: String\\n  lt: String\\n  ge: String\\n  gt: String\\n  contains: String\\n  notContains: String\\n  between: [String]\\n  beginsWith: String\\n  in: [String]\\n  notIn: [String]\\n}\\n\\ninput ModelSubscriptionIntInput {\\n  ne: Int\\n  eq: Int\\n  le: Int\\n  lt: Int\\n  ge: Int\\n  gt: Int\\n  between: [Int]\\n  in: [Int]\\n  notIn: [Int]\\n}\\n\\ninput ModelSubscriptionFloatInput {\\n  ne: Float\\n  eq: Float\\n  le: Float\\n  lt: Float\\n  ge: Float\\n  gt: Float\\n  between: [Float]\\n  in: [Float]\\n  notIn: [Float]\\n}\\n\\ninput ModelSubscriptionBooleanInput {\\n  ne: Boolean\\n  eq: Boolean\\n}\\n\\ninput ModelSubscriptionIDInput {\\n  ne: ID\\n  eq: ID\\n  le: ID\\n  lt: ID\\n  ge: ID\\n  gt: ID\\n  contains: ID\\n  notContains: ID\\n  between: [ID]\\n  beginsWith: ID\\n  in: [ID]\\n  notIn: [ID]\\n}\\n\\nenum ModelAttributeTypes {\\n  binary\\n  binarySet\\n  bool\\n  list\\n  map\\n  number\\n  numberSet\\n  string\\n  stringSet\\n  _null\\n}\\n\\ninput ModelSizeInput {\\n  ne: Int\\n  eq: Int\\n  le: Int\\n  lt: Int\\n  ge: Int\\n  gt: Int\\n  between: [Int]\\n}\\n\\nenum ModelSortDirection {\\n  ASC\\n  DESC\\n}\\n\\ntype ModelPostConnection @aws_cognito_user_pools @aws_iam {\\n  items: [Post]!\\n  nextToken: String\\n}\\n\\ninput ModelPostFilterInput {\\n  title: ModelStringInput\\n  content: ModelStringInput\\n  authorName: ModelStringInput\\n  authorDoB: ModelStringInput\\n  id: ModelIDInput\\n  createdAt: ModelStringInput\\n  updatedAt: ModelStringInput\\n  and: [ModelPostFilterInput]\\n  or: [ModelPostFilterInput]\\n  not: ModelPostFilterInput\\n}\\n\\ntype Query {\\n  getPost(id: ID!): Post @aws_cognito_user_pools @aws_iam\\n  listPosts(filter: ModelPostFilterInput, limit: Int, nextToken: String): ModelPostConnection @aws_cognito_user_pools @aws_iam\\n  listPostByAuthorName(authorName: String!, sortDirection: ModelSortDirection, filter: ModelPostFilterInput, limit: Int, nextToken: String): ModelPostConnection @aws_cognito_user_pools @aws_iam\\n  getPerson(name: String!, dateOfBirth: AWSDate!): Person @aws_cognito_user_pools @aws_iam\\n  listPeople(name: String, dateOfBirth: ModelStringKeyConditionInput, filter: ModelPersonFilterInput, limit: Int, nextToken: String, sortDirection: ModelSortDirection): ModelPersonConnection @aws_cognito_user_pools @aws_iam\\n}\\n\\ninput ModelPostConditionInput {\\n  title: ModelStringInput\\n  content: ModelStringInput\\n  authorName: ModelStringInput\\n  authorDoB: ModelStringInput\\n  and: [ModelPostConditionInput]\\n  or: [ModelPostConditionInput]\\n  not: ModelPostConditionInput\\n  createdAt: ModelStringInput\\n  updatedAt: ModelStringInput\\n}\\n\\ninput CreatePostInput {\\n  title: String!\\n  content: String!\\n  authorName: String\\n  authorDoB: AWSDate\\n  id: ID\\n}\\n\\ninput UpdatePostInput {\\n  title: String\\n  content: String\\n  authorName: String\\n  authorDoB: AWSDate\\n  id: ID!\\n}\\n\\ninput DeletePostInput {\\n  id: ID!\\n}\\n\\ntype Mutation {\\n  createPost(input: CreatePostInput!, condition: ModelPostConditionInput): Post @aws_cognito_user_pools @aws_iam\\n  updatePost(input: UpdatePostInput!, condition: ModelPostConditionInput): Post @aws_cognito_user_pools @aws_iam\\n  deletePost(input: DeletePostInput!, condition: ModelPostConditionInput): Post @aws_cognito_user_pools @aws_iam\\n  createPerson(input: CreatePersonInput!, condition: ModelPersonConditionInput): Person @aws_cognito_user_pools @aws_iam\\n  updatePerson(input: UpdatePersonInput!, condition: ModelPersonConditionInput): Person @aws_cognito_user_pools @aws_iam\\n  deletePerson(input: DeletePersonInput!, condition: ModelPersonConditionInput): Person @aws_cognito_user_pools @aws_iam\\n}\\n\\ninput ModelSubscriptionPostFilterInput {\\n  title: ModelSubscriptionStringInput\\n  content: ModelSubscriptionStringInput\\n  authorName: ModelSubscriptionStringInput\\n  authorDoB: ModelSubscriptionStringInput\\n  id: ModelSubscriptionIDInput\\n  createdAt: ModelSubscriptionStringInput\\n  updatedAt: ModelSubscriptionStringInput\\n  and: [ModelSubscriptionPostFilterInput]\\n  or: [ModelSubscriptionPostFilterInput]\\n}\\n\\ntype Subscription {\\n  onCreatePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"createPost\\"]) @aws_cognito_user_pools @aws_iam\\n  onUpdatePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"updatePost\\"]) @aws_cognito_user_pools @aws_iam\\n  onDeletePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"deletePost\\"]) @aws_cognito_user_pools @aws_iam\\n  onCreatePerson(filter: ModelSubscriptionPersonFilterInput): Person @aws_subscribe(mutations: [\\"createPerson\\"]) @aws_cognito_user_pools @aws_iam\\n  onUpdatePerson(filter: ModelSubscriptionPersonFilterInput): Person @aws_subscribe(mutations: [\\"updatePerson\\"]) @aws_cognito_user_pools @aws_iam\\n  onDeletePerson(filter: ModelSubscriptionPersonFilterInput): Person @aws_subscribe(mutations: [\\"deletePerson\\"]) @aws_cognito_user_pools @aws_iam\\n}\\n\\ntype ModelPersonConnection @aws_cognito_user_pools @aws_iam {\\n  items: [Person]!\\n  nextToken: String\\n}\\n\\ninput ModelPersonFilterInput {\\n  name: ModelStringInput\\n  dateOfBirth: ModelStringInput\\n  id: ModelIDInput\\n  createdAt: ModelStringInput\\n  updatedAt: ModelStringInput\\n  and: [ModelPersonFilterInput]\\n  or: [ModelPersonFilterInput]\\n  not: ModelPersonFilterInput\\n}\\n\\ninput ModelPersonConditionInput {\\n  and: [ModelPersonConditionInput]\\n  or: [ModelPersonConditionInput]\\n  not: ModelPersonConditionInput\\n  createdAt: ModelStringInput\\n  updatedAt: ModelStringInput\\n}\\n\\ninput CreatePersonInput {\\n  name: String!\\n  dateOfBirth: AWSDate!\\n}\\n\\ninput UpdatePersonInput {\\n  name: String!\\n  dateOfBirth: AWSDate!\\n}\\n\\ninput DeletePersonInput {\\n  name: String!\\n  dateOfBirth: AWSDate!\\n}\\n\\ninput ModelSubscriptionPersonFilterInput {\\n  name: ModelSubscriptionStringInput\\n  dateOfBirth: ModelSubscriptionStringInput\\n  id: ModelSubscriptionIDInput\\n  createdAt: ModelSubscriptionStringInput\\n  updatedAt: ModelSubscriptionStringInput\\n  and: [ModelSubscriptionPersonFilterInput]\\n  or: [ModelSubscriptionPersonFilterInput]\\n}\\n\\ninput ModelStringKeyConditionInput {\\n  eq: String\\n  le: String\\n  lt: String\\n  ge: String\\n  gt: String\\n  between: [String]\\n  beginsWith: String\\n}\\n"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-user-thirdlambda3EC3C10B-J4fEZYAKBBf1"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5403',
        message: " AWS::Lambda::Function 'amplify-project-user-thirdlambda3EC3C10B-J4fEZYAKBBf1' hotswapped!",
        data: {
            cause: {
                logicalId: 'thirdlambda3EC3C10B',
                oldValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['thirdlambdaServiceRoleEA7CB9CB', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'third',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['thirdlambdaServiceRoleEA7CB9CB'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/third-lambda/Resource',
                        'aws:asset:path': 'asset.1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                newValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['thirdlambdaServiceRoleEA7CB9CB', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'third',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['thirdlambdaServiceRoleEA7CB9CB'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/third-lambda/Resource',
                        'aws:asset:path': 'asset.7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                propertyUpdates: {
                    Code: {
                        oldValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        newValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        isDifferent: true,
                        changeImpact: 'WILL_UPDATE',
                    },
                },
                metadata: {
                    entry: {
                        type: 'aws:cdk:logicalId',
                        data: 'thirdlambda3EC3C10B',
                    },
                    constructPath: 'function/third-lambda',
                },
            },
            resources: [
                {
                    logicalId: 'thirdlambda3EC3C10B',
                    resourceType: 'AWS::Lambda::Function',
                    physicalName: 'amplify-project-user-thirdlambda3EC3C10B-J4fEZYAKBBf1',
                    metadata: {
                        entry: {
                            type: 'aws:cdk:logicalId',
                            data: 'thirdlambda3EC3C10B',
                        },
                        constructPath: 'function/third-lambda',
                    },
                },
            ],
        },
    },
    {
        code: 'CDK_TOOLKIT_I5402',
        message: " AWS::Lambda::Function 'amplify-project-pra-seventhlambda73310E16-P3YECxPNfISl'",
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-prav-secondlambdaBB16058D-qr2Pe32Olu4U"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5403',
        message: " AWS::Lambda::Function 'amplify-project-prav-secondlambdaBB16058D-qr2Pe32Olu4U' hotswapped!",
        data: {
            cause: {
                logicalId: 'secondlambdaBB16058D',
                oldValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['secondlambdaServiceRoleD98B591C', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'second',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['secondlambdaServiceRoleD98B591C'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/second-lambda/Resource',
                        'aws:asset:path': 'asset.1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                newValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['secondlambdaServiceRoleD98B591C', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'second',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['secondlambdaServiceRoleD98B591C'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/second-lambda/Resource',
                        'aws:asset:path': 'asset.7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                propertyUpdates: {
                    Code: {
                        oldValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        newValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        isDifferent: true,
                        changeImpact: 'WILL_UPDATE',
                    },
                },
                metadata: {
                    entry: {
                        type: 'aws:cdk:logicalId',
                        data: 'secondlambdaBB16058D',
                    },
                    constructPath: 'function/second-lambda',
                },
            },
            resources: [
                {
                    logicalId: 'secondlambdaBB16058D',
                    resourceType: 'AWS::Lambda::Function',
                    physicalName: 'amplify-project-prav-secondlambdaBB16058D-qr2Pe32Olu4U',
                    metadata: {
                        entry: {
                            type: 'aws:cdk:logicalId',
                            data: 'secondlambdaBB16058D',
                        },
                        constructPath: 'function/second-lambda',
                    },
                },
            ],
        },
    },
    {
        code: 'CDK_TOOLKIT_I5402',
        message: " AWS::Lambda::Function 'amplify-project-prav-eighthlambda175C4185-OMTffC3yD8sr'",
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-user-sixthlambda68A95826-MVDPZVKKJypK"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5403',
        message: " AWS::Lambda::Function 'amplify-project-user-sixthlambda68A95826-MVDPZVKKJypK' hotswapped!",
        data: {
            cause: {
                logicalId: 'sixthlambda68A95826',
                oldValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['sixthlambdaServiceRole1B7BA68C', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'sixth',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['sixthlambdaServiceRole1B7BA68C'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/sixth-lambda/Resource',
                        'aws:asset:path': 'asset.1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                newValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['sixthlambdaServiceRole1B7BA68C', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'sixth',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['sixthlambdaServiceRole1B7BA68C'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/sixth-lambda/Resource',
                        'aws:asset:path': 'asset.7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                propertyUpdates: {
                    Code: {
                        oldValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        newValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        isDifferent: true,
                        changeImpact: 'WILL_UPDATE',
                    },
                },
                metadata: {
                    entry: {
                        type: 'aws:cdk:logicalId',
                        data: 'sixthlambda68A95826',
                    },
                    constructPath: 'function/sixth-lambda',
                },
            },
            resources: [
                {
                    logicalId: 'sixthlambda68A95826',
                    resourceType: 'AWS::Lambda::Function',
                    physicalName: 'amplify-project-user-sixthlambda68A95826-MVDPZVKKJypK',
                    metadata: {
                        entry: {
                            type: 'aws:cdk:logicalId',
                            data: 'sixthlambda68A95826',
                        },
                        constructPath: 'function/sixth-lambda',
                    },
                },
            ],
        },
    },
    {
        code: 'CDK_TOOLKIT_I5402',
        message: " AWS::Lambda::Function 'amplify-project-user-ninthlambdaD4B9138E-bkBYtRQ8aoTp'",
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-p-sayhelloslambda734D68D9-33dq3RGkPNTR"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5403',
        message: " AWS::Lambda::Function 'amplify-project-p-sayhelloslambda734D68D9-33dq3RGkPNTR' hotswapped!",
        data: {
            cause: {
                logicalId: 'sayhelloslambda734D68D9',
                oldValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        Environment: {
                            Variables: {
                                existing: 'someValue',
                                NEW_ENV_VAR: 'someValue',
                                AMPLIFY_SSM_ENV_CONFIG: '{"TEST_NAME_BUCKET_NAME":{"path":"/amplify/resource_reference/project/user-sandbox-0e1752a889/TEST_NAME_BUCKET_NAME"}}',
                                TEST_NAME_BUCKET_NAME: '<value will be resolved during runtime>',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['sayhelloslambdaServiceRole0A182822', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'say-hellos',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['sayhelloslambdaServiceRole0A182822'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/say-hellos-lambda/Resource',
                        'aws:asset:path': 'asset.1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                newValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        Environment: {
                            Variables: {
                                existing: 'someValue',
                                NEW_ENV_VAR: 'someValue',
                                AMPLIFY_SSM_ENV_CONFIG: '{"TEST_NAME_BUCKET_NAME":{"path":"/amplify/resource_reference/project/user-sandbox-0e1752a889/TEST_NAME_BUCKET_NAME"}}',
                                TEST_NAME_BUCKET_NAME: '<value will be resolved during runtime>',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['sayhelloslambdaServiceRole0A182822', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'say-hellos',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['sayhelloslambdaServiceRole0A182822'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/say-hellos-lambda/Resource',
                        'aws:asset:path': 'asset.7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                propertyUpdates: {
                    Code: {
                        oldValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        newValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        isDifferent: true,
                        changeImpact: 'WILL_UPDATE',
                    },
                },
                metadata: {
                    entry: {
                        type: 'aws:cdk:logicalId',
                        data: 'sayhelloslambda734D68D9',
                    },
                    constructPath: 'function/say-hellos-lambda',
                },
            },
            resources: [
                {
                    logicalId: 'sayhelloslambda734D68D9',
                    resourceType: 'AWS::Lambda::Function',
                    physicalName: 'amplify-project-p-sayhelloslambda734D68D9-33dq3RGkPNTR',
                    metadata: {
                        entry: {
                            type: 'aws:cdk:logicalId',
                            data: 'sayhelloslambda734D68D9',
                        },
                        constructPath: 'function/say-hellos-lambda',
                    },
                },
            ],
        },
    },
    {
        code: 'CDK_TOOLKIT_I5402',
        message: " AWS::Lambda::Function 'amplify-project-user-tenthlambda31BF89D1-18CnDSHhPkXF'",
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-user-firstlambda23886240-onnWijqFcbCB"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5403',
        message: " AWS::Lambda::Function 'amplify-project-user-firstlambda23886240-onnWijqFcbCB' hotswapped!",
        data: {
            cause: {
                logicalId: 'firstlambda23886240',
                oldValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['firstlambdaServiceRole4BC75CC6', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'first',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['firstlambdaServiceRole4BC75CC6'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/first-lambda/Resource',
                        'aws:asset:path': 'asset.1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                newValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['firstlambdaServiceRole4BC75CC6', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'first',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['firstlambdaServiceRole4BC75CC6'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/first-lambda/Resource',
                        'aws:asset:path': 'asset.7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                propertyUpdates: {
                    Code: {
                        oldValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        newValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        isDifferent: true,
                        changeImpact: 'WILL_UPDATE',
                    },
                },
                metadata: {
                    entry: {
                        type: 'aws:cdk:logicalId',
                        data: 'firstlambda23886240',
                    },
                    constructPath: 'function/first-lambda',
                },
            },
            resources: [
                {
                    logicalId: 'firstlambda23886240',
                    resourceType: 'AWS::Lambda::Function',
                    physicalName: 'amplify-project-user-firstlambda23886240-onnWijqFcbCB',
                    metadata: {
                        entry: {
                            type: 'aws:cdk:logicalId',
                            data: 'firstlambda23886240',
                        },
                        constructPath: 'function/first-lambda',
                    },
                },
            ],
        },
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-user-fifthlambda2306F097-jV2mMa7c6Ltg"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5403',
        message: " AWS::Lambda::Function 'amplify-project-user-fifthlambda2306F097-jV2mMa7c6Ltg' hotswapped!",
        data: {
            cause: {
                logicalId: 'fifthlambda2306F097',
                oldValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['fifthlambdaServiceRoleABD52544', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'fifth',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['fifthlambdaServiceRoleABD52544'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/fifth-lambda/Resource',
                        'aws:asset:path': 'asset.1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                newValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['fifthlambdaServiceRoleABD52544', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'fifth',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['fifthlambdaServiceRoleABD52544'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/fifth-lambda/Resource',
                        'aws:asset:path': 'asset.7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                propertyUpdates: {
                    Code: {
                        oldValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        newValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        isDifferent: true,
                        changeImpact: 'WILL_UPDATE',
                    },
                },
                metadata: {
                    entry: {
                        type: 'aws:cdk:logicalId',
                        data: 'fifthlambda2306F097',
                    },
                    constructPath: 'function/fifth-lambda',
                },
            },
            resources: [
                {
                    logicalId: 'fifthlambda2306F097',
                    resourceType: 'AWS::Lambda::Function',
                    physicalName: 'amplify-project-user-fifthlambda2306F097-jV2mMa7c6Ltg',
                    metadata: {
                        entry: {
                            type: 'aws:cdk:logicalId',
                            data: 'fifthlambda2306F097',
                        },
                        constructPath: 'function/fifth-lambda',
                    },
                },
            ],
        },
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] AppSync.GetSchemaCreationStatus({"apiId":"qf3pkl3rvjdlbccnohxt5gkdyy"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5403',
        message: " AWS::AppSync::GraphQLSchema 'qf3pkl3rvjdlbccnohxt5gkdyyGraphQLSchema' hotswapped!",
        data: {
            cause: {
                logicalId: 'amplifyDataGraphQLAPITransformerSchemaFF50A789',
                oldValue: {
                    Type: 'AWS::AppSync::GraphQLSchema',
                    Properties: {
                        ApiId: {
                            'Fn::GetAtt': ['amplifyDataGraphQLAPI42A6FA33', 'ApiId'],
                        },
                        DefinitionS3Location: {
                            'Fn::Sub': 's3://cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}/62596653ff7c40d0eaf05047310298abd3716047a4767a1df8d394323fa9d415.graphql',
                        },
                    },
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/data/amplifyData/GraphQLAPI/TransformerSchema',
                    },
                },
                newValue: {
                    Type: 'AWS::AppSync::GraphQLSchema',
                    Properties: {
                        ApiId: {
                            'Fn::GetAtt': ['amplifyDataGraphQLAPI42A6FA33', 'ApiId'],
                        },
                        DefinitionS3Location: {
                            'Fn::Sub': 's3://cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}/e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql',
                        },
                    },
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/data/amplifyData/GraphQLAPI/TransformerSchema',
                    },
                },
                propertyUpdates: {
                    DefinitionS3Location: {
                        oldValue: {
                            'Fn::Sub': 's3://cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}/62596653ff7c40d0eaf05047310298abd3716047a4767a1df8d394323fa9d415.graphql',
                        },
                        newValue: {
                            'Fn::Sub': 's3://cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}/e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql',
                        },
                        isDifferent: true,
                        changeImpact: 'WILL_UPDATE',
                    },
                },
                metadata: {
                    entry: {
                        type: 'aws:cdk:logicalId',
                        data: 'amplifyDataGraphQLAPITransformerSchemaFF50A789',
                    },
                    constructPath: 'data/amplifyData/GraphQLAPI/TransformerSchema',
                },
            },
            resources: [
                {
                    logicalId: 'amplifyDataGraphQLAPITransformerSchemaFF50A789',
                    resourceType: 'AWS::AppSync::GraphQLSchema',
                    physicalName: 'qf3pkl3rvjdlbccnohxt5gkdyyGraphQLSchema',
                    metadata: {
                        entry: {
                            type: 'aws:cdk:logicalId',
                            data: 'amplifyDataGraphQLAPITransformerSchemaFF50A789',
                        },
                        constructPath: 'data/amplifyData/GraphQLAPI/TransformerSchema',
                    },
                },
            ],
        },
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-prav-fourthlambda8DE196FD-OlYUvlKcriu6"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5403',
        message: " AWS::Lambda::Function 'amplify-project-prav-fourthlambda8DE196FD-OlYUvlKcriu6' hotswapped!",
        data: {
            cause: {
                logicalId: 'fourthlambda8DE196FD',
                oldValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['fourthlambdaServiceRole4AC9C036', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'fourth',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['fourthlambdaServiceRole4AC9C036'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/fourth-lambda/Resource',
                        'aws:asset:path': 'asset.1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                newValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['fourthlambdaServiceRole4AC9C036', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'fourth',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['fourthlambdaServiceRole4AC9C036'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/fourth-lambda/Resource',
                        'aws:asset:path': 'asset.7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                propertyUpdates: {
                    Code: {
                        oldValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        newValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        isDifferent: true,
                        changeImpact: 'WILL_UPDATE',
                    },
                },
                metadata: {
                    entry: {
                        type: 'aws:cdk:logicalId',
                        data: 'fourthlambda8DE196FD',
                    },
                    constructPath: 'function/fourth-lambda',
                },
            },
            resources: [
                {
                    logicalId: 'fourthlambda8DE196FD',
                    resourceType: 'AWS::Lambda::Function',
                    physicalName: 'amplify-project-prav-fourthlambda8DE196FD-OlYUvlKcriu6',
                    metadata: {
                        entry: {
                            type: 'aws:cdk:logicalId',
                            data: 'fourthlambda8DE196FD',
                        },
                        constructPath: 'function/fourth-lambda',
                    },
                },
            ],
        },
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.UpdateFunctionCode({"FunctionName":"amplify-project-user-ninthlambdaD4B9138E-bkBYtRQ8aoTp","S3Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","S3Key":"7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.UpdateFunctionCode({"FunctionName":"amplify-project-prav-eighthlambda175C4185-OMTffC3yD8sr","S3Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","S3Key":"7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-user-ninthlambdaD4B9138E-bkBYtRQ8aoTp"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.UpdateFunctionCode({"FunctionName":"amplify-project-pra-seventhlambda73310E16-P3YECxPNfISl","S3Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","S3Key":"7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-prav-eighthlambda175C4185-OMTffC3yD8sr"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.UpdateFunctionCode({"FunctionName":"amplify-project-user-tenthlambda31BF89D1-18CnDSHhPkXF","S3Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","S3Key":"7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-pra-seventhlambda73310E16-P3YECxPNfISl"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-user-tenthlambda31BF89D1-18CnDSHhPkXF"}) -> OK',
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-user-ninthlambdaD4B9138E-bkBYtRQ8aoTp"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5403',
        message: " AWS::Lambda::Function 'amplify-project-user-ninthlambdaD4B9138E-bkBYtRQ8aoTp' hotswapped!",
        data: {
            cause: {
                logicalId: 'ninthlambdaD4B9138E',
                oldValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['ninthlambdaServiceRoleD3E9B493', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'ninth',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['ninthlambdaServiceRoleD3E9B493'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/ninth-lambda/Resource',
                        'aws:asset:path': 'asset.1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                newValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['ninthlambdaServiceRoleD3E9B493', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'ninth',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['ninthlambdaServiceRoleD3E9B493'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/ninth-lambda/Resource',
                        'aws:asset:path': 'asset.7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                propertyUpdates: {
                    Code: {
                        oldValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        newValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        isDifferent: true,
                        changeImpact: 'WILL_UPDATE',
                    },
                },
                metadata: {
                    entry: {
                        type: 'aws:cdk:logicalId',
                        data: 'ninthlambdaD4B9138E',
                    },
                    constructPath: 'function/ninth-lambda',
                },
            },
            resources: [
                {
                    logicalId: 'ninthlambdaD4B9138E',
                    resourceType: 'AWS::Lambda::Function',
                    physicalName: 'amplify-project-user-ninthlambdaD4B9138E-bkBYtRQ8aoTp',
                    metadata: {
                        entry: {
                            type: 'aws:cdk:logicalId',
                            data: 'ninthlambdaD4B9138E',
                        },
                        constructPath: 'function/ninth-lambda',
                    },
                },
            ],
        },
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-prav-eighthlambda175C4185-OMTffC3yD8sr"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5403',
        message: " AWS::Lambda::Function 'amplify-project-prav-eighthlambda175C4185-OMTffC3yD8sr' hotswapped!",
        data: {
            cause: {
                logicalId: 'eighthlambda175C4185',
                oldValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['eighthlambdaServiceRole3875CF41', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'eighth',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['eighthlambdaServiceRole3875CF41'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/eighth-lambda/Resource',
                        'aws:asset:path': 'asset.1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                newValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['eighthlambdaServiceRole3875CF41', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'eighth',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['eighthlambdaServiceRole3875CF41'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/eighth-lambda/Resource',
                        'aws:asset:path': 'asset.7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                propertyUpdates: {
                    Code: {
                        oldValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        newValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        isDifferent: true,
                        changeImpact: 'WILL_UPDATE',
                    },
                },
                metadata: {
                    entry: {
                        type: 'aws:cdk:logicalId',
                        data: 'eighthlambda175C4185',
                    },
                    constructPath: 'function/eighth-lambda',
                },
            },
            resources: [
                {
                    logicalId: 'eighthlambda175C4185',
                    resourceType: 'AWS::Lambda::Function',
                    physicalName: 'amplify-project-prav-eighthlambda175C4185-OMTffC3yD8sr',
                    metadata: {
                        entry: {
                            type: 'aws:cdk:logicalId',
                            data: 'eighthlambda175C4185',
                        },
                        constructPath: 'function/eighth-lambda',
                    },
                },
            ],
        },
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-pra-seventhlambda73310E16-P3YECxPNfISl"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5403',
        message: " AWS::Lambda::Function 'amplify-project-pra-seventhlambda73310E16-P3YECxPNfISl' hotswapped!",
        data: {
            cause: {
                logicalId: 'seventhlambda73310E16',
                oldValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['seventhlambdaServiceRole33B6BD17', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'seventh',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['seventhlambdaServiceRole33B6BD17'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/seventh-lambda/Resource',
                        'aws:asset:path': 'asset.1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                newValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['seventhlambdaServiceRole33B6BD17', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'seventh',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['seventhlambdaServiceRole33B6BD17'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/seventh-lambda/Resource',
                        'aws:asset:path': 'asset.7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                propertyUpdates: {
                    Code: {
                        oldValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        newValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        isDifferent: true,
                        changeImpact: 'WILL_UPDATE',
                    },
                },
                metadata: {
                    entry: {
                        type: 'aws:cdk:logicalId',
                        data: 'seventhlambda73310E16',
                    },
                    constructPath: 'function/seventh-lambda',
                },
            },
            resources: [
                {
                    logicalId: 'seventhlambda73310E16',
                    resourceType: 'AWS::Lambda::Function',
                    physicalName: 'amplify-project-pra-seventhlambda73310E16-P3YECxPNfISl',
                    metadata: {
                        entry: {
                            type: 'aws:cdk:logicalId',
                            data: 'seventhlambda73310E16',
                        },
                        constructPath: 'function/seventh-lambda',
                    },
                },
            ],
        },
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.GetFunction({"FunctionName":"amplify-project-user-tenthlambda31BF89D1-18CnDSHhPkXF"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5403',
        message: " AWS::Lambda::Function 'amplify-project-user-tenthlambda31BF89D1-18CnDSHhPkXF' hotswapped!",
        data: {
            cause: {
                logicalId: 'tenthlambda31BF89D1',
                oldValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['tenthlambdaServiceRoleB53516F6', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'tenth',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['tenthlambdaServiceRoleB53516F6'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/tenth-lambda/Resource',
                        'aws:asset:path': 'asset.1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                newValue: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Architectures: ['x86_64'],
                        Code: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        Environment: {
                            Variables: {
                                AMPLIFY_SSM_ENV_CONFIG: '{}',
                            },
                        },
                        EphemeralStorage: {
                            Size: 512,
                        },
                        Handler: 'index.handler',
                        MemorySize: 512,
                        Role: {
                            'Fn::GetAtt': ['tenthlambdaServiceRoleB53516F6', 'Arn'],
                        },
                        Runtime: 'nodejs20.x',
                        Tags: [
                            {
                                Key: 'amplify:deployment-type',
                                Value: 'sandbox',
                            },
                            {
                                Key: 'amplify:friendly-name',
                                Value: 'tenth',
                            },
                            {
                                Key: 'created-by',
                                Value: 'amplify',
                            },
                        ],
                        Timeout: 3,
                    },
                    DependsOn: ['tenthlambdaServiceRoleB53516F6'],
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/function/tenth-lambda/Resource',
                        'aws:asset:path': 'asset.7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74',
                        'aws:asset:is-bundled': true,
                        'aws:asset:property': 'Code',
                    },
                },
                propertyUpdates: {
                    Code: {
                        oldValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '1b6ab78969563aeed1415e5617dfcab31b5c7c0994c9b5892ac0e0100905379e.zip',
                        },
                        newValue: {
                            S3Bucket: {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                            S3Key: '7df5c7fefd9a8037d8cc3cbdb812cecc315293fadc81132033e5707e76d9cf74.zip',
                        },
                        isDifferent: true,
                        changeImpact: 'WILL_UPDATE',
                    },
                },
                metadata: {
                    entry: {
                        type: 'aws:cdk:logicalId',
                        data: 'tenthlambda31BF89D1',
                    },
                    constructPath: 'function/tenth-lambda',
                },
            },
            resources: [
                {
                    logicalId: 'tenthlambda31BF89D1',
                    resourceType: 'AWS::Lambda::Function',
                    physicalName: 'amplify-project-user-tenthlambda31BF89D1-18CnDSHhPkXF',
                    metadata: {
                        entry: {
                            type: 'aws:cdk:logicalId',
                            data: 'tenthlambda31BF89D1',
                        },
                        constructPath: 'function/tenth-lambda',
                    },
                },
            ],
        },
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.Invoke({"FunctionName":"arn:aws:lambda:us-west-2:123456789012:function:amplify-project--CustomCDKBucketDeploymen-SiZ2pdZyFqe4","Payload":"***SensitiveInformation***"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5403',
        message: " Contents of AWS::S3::Bucket 'amplify-project--amplifydataamplifycodege-clxvnkfko9ki' hotswapped!",
        data: {
            cause: {
                logicalId: 'amplifyDataAmplifyCodegenAssetsAmplifyCodegenAssetsDeploymentCustomResource1536MiB21775929',
                oldValue: {
                    Type: 'Custom::CDKBucketDeployment',
                    Properties: {
                        ServiceToken: {
                            'Fn::GetAtt': [
                                'CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C1536MiBC5D8AB21',
                                'Arn',
                            ],
                        },
                        SourceBucketNames: [
                            {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                        ],
                        SourceObjectKeys: [
                            '81666066097613bde4eabd6431a8a786f04d690433708e24941fd16cbe2b14ee.zip',
                        ],
                        SourceMarkers: [{}],
                        DestinationBucketName: {
                            Ref: 'amplifyDataAmplifyCodegenAssetsAmplifyCodegenAssetsBucket9CCB4ACA',
                        },
                        Prune: true,
                        OutputObjectKeys: true,
                        DestinationBucketArn: {
                            'Fn::GetAtt': [
                                'amplifyDataAmplifyCodegenAssetsAmplifyCodegenAssetsBucket9CCB4ACA',
                                'Arn',
                            ],
                        },
                    },
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/data/amplifyData/AmplifyCodegenAssets/AmplifyCodegenAssetsDeployment/CustomResource-1536MiB/Default',
                    },
                },
                newValue: {
                    Type: 'Custom::CDKBucketDeployment',
                    Properties: {
                        ServiceToken: {
                            'Fn::GetAtt': [
                                'CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C1536MiBC5D8AB21',
                                'Arn',
                            ],
                        },
                        SourceBucketNames: [
                            {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                        ],
                        SourceObjectKeys: [
                            '2f050e6b9affea69843658b09b5d9073acc9cda9170a30ff043def78a33c6ad0.zip',
                        ],
                        SourceMarkers: [{}],
                        DestinationBucketName: {
                            Ref: 'amplifyDataAmplifyCodegenAssetsAmplifyCodegenAssetsBucket9CCB4ACA',
                        },
                        Prune: true,
                        OutputObjectKeys: true,
                        DestinationBucketArn: {
                            'Fn::GetAtt': [
                                'amplifyDataAmplifyCodegenAssetsAmplifyCodegenAssetsBucket9CCB4ACA',
                                'Arn',
                            ],
                        },
                    },
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/data/amplifyData/AmplifyCodegenAssets/AmplifyCodegenAssetsDeployment/CustomResource-1536MiB/Default',
                    },
                },
                propertyUpdates: {
                    SourceObjectKeys: {
                        oldValue: [
                            '81666066097613bde4eabd6431a8a786f04d690433708e24941fd16cbe2b14ee.zip',
                        ],
                        newValue: [
                            '2f050e6b9affea69843658b09b5d9073acc9cda9170a30ff043def78a33c6ad0.zip',
                        ],
                        isDifferent: true,
                        changeImpact: 'NO_CHANGE',
                    },
                },
                metadata: {
                    entry: {
                        type: 'aws:cdk:logicalId',
                        data: 'amplifyDataAmplifyCodegenAssetsAmplifyCodegenAssetsDeploymentCustomResource1536MiB21775929',
                    },
                    constructPath: 'data/amplifyData/AmplifyCodegenAssets/AmplifyCodegenAssetsDeployment/CustomResource-1536MiB/Default',
                },
            },
            resources: [
                {
                    logicalId: 'amplifyDataAmplifyCodegenAssetsAmplifyCodegenAssetsDeploymentCustomResource1536MiB21775929',
                    physicalName: 'amplify-project--amplifydataamplifycodege-clxvnkfko9ki',
                    resourceType: 'Custom::CDKBucketDeployment',
                    description: "Contents of AWS::S3::Bucket 'amplify-project--amplifydataamplifycodege-clxvnkfko9ki'",
                    metadata: {
                        entry: {
                            type: 'aws:cdk:logicalId',
                            data: 'amplifyDataAmplifyCodegenAssetsAmplifyCodegenAssetsDeploymentCustomResource1536MiB21775929',
                        },
                        constructPath: 'data/amplifyData/AmplifyCodegenAssets/AmplifyCodegenAssetsDeployment/CustomResource-1536MiB/Default',
                    },
                },
            ],
        },
    },
    {
        code: 'CDK_SDK_I0100',
        message: '[sdk info] Lambda.Invoke({"FunctionName":"arn:aws:lambda:us-west-2:123456789012:function:amplify-project--CustomCDKBucketDeploymen-SiZ2pdZyFqe4","Payload":"***SensitiveInformation***"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5403',
        message: " Contents of AWS::S3::Bucket 'amplify-project--modelintrospectionschema-cwdah99w4vhc' hotswapped!",
        data: {
            cause: {
                logicalId: 'modelIntrospectionSchemaBucketDeploymentCustomResource1536MiB104B97EC',
                oldValue: {
                    Type: 'Custom::CDKBucketDeployment',
                    Properties: {
                        ServiceToken: {
                            'Fn::GetAtt': [
                                'CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C1536MiBC5D8AB21',
                                'Arn',
                            ],
                        },
                        SourceBucketNames: [
                            {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                        ],
                        SourceObjectKeys: [
                            '7ab2d5b7a80b81b1dbda7e758e6ac8f06699e8186171dd2a6be1bf0785b84916.zip',
                        ],
                        SourceMarkers: [{}],
                        DestinationBucketName: {
                            Ref: 'modelIntrospectionSchemaBucketF566B665',
                        },
                        Prune: true,
                        OutputObjectKeys: true,
                    },
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/data/modelIntrospectionSchemaBucketDeployment/CustomResource-1536MiB/Default',
                    },
                },
                newValue: {
                    Type: 'Custom::CDKBucketDeployment',
                    Properties: {
                        ServiceToken: {
                            'Fn::GetAtt': [
                                'CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C1536MiBC5D8AB21',
                                'Arn',
                            ],
                        },
                        SourceBucketNames: [
                            {
                                'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
                            },
                        ],
                        SourceObjectKeys: [
                            '907cb295088857ccb545b8aae83288bd9dfa82115bb0d20a0ae0d08dc8872f8a.zip',
                        ],
                        SourceMarkers: [{}],
                        DestinationBucketName: {
                            Ref: 'modelIntrospectionSchemaBucketF566B665',
                        },
                        Prune: true,
                        OutputObjectKeys: true,
                    },
                    UpdateReplacePolicy: 'Delete',
                    DeletionPolicy: 'Delete',
                    Metadata: {
                        'aws:cdk:path': 'amplify-project-user-sandbox-0e1752a889/data/modelIntrospectionSchemaBucketDeployment/CustomResource-1536MiB/Default',
                    },
                },
                propertyUpdates: {
                    SourceObjectKeys: {
                        oldValue: [
                            '7ab2d5b7a80b81b1dbda7e758e6ac8f06699e8186171dd2a6be1bf0785b84916.zip',
                        ],
                        newValue: [
                            '907cb295088857ccb545b8aae83288bd9dfa82115bb0d20a0ae0d08dc8872f8a.zip',
                        ],
                        isDifferent: true,
                        changeImpact: 'NO_CHANGE',
                    },
                },
                metadata: {
                    entry: {
                        type: 'aws:cdk:logicalId',
                        data: 'modelIntrospectionSchemaBucketDeploymentCustomResource1536MiB104B97EC',
                    },
                    constructPath: 'data/modelIntrospectionSchemaBucketDeployment/CustomResource-1536MiB/Default',
                },
            },
            resources: [
                {
                    logicalId: 'modelIntrospectionSchemaBucketDeploymentCustomResource1536MiB104B97EC',
                    physicalName: 'amplify-project--modelintrospectionschema-cwdah99w4vhc',
                    resourceType: 'Custom::CDKBucketDeployment',
                    description: "Contents of AWS::S3::Bucket 'amplify-project--modelintrospectionschema-cwdah99w4vhc'",
                    metadata: {
                        entry: {
                            type: 'aws:cdk:logicalId',
                            data: 'modelIntrospectionSchemaBucketDeploymentCustomResource1536MiB104B97EC',
                        },
                        constructPath: 'data/modelIntrospectionSchemaBucketDeployment/CustomResource-1536MiB/Default',
                    },
                },
            ],
        },
    },
    {
        code: 'CDK_TOOLKIT_I5410',
        message: '  hotswap-deployment time: 9.18s',
    },
    {
        code: 'CDK_TOOLKIT_I5900',
        message: '  amplify-project-user-sandbox-0e1752a889',
    },
    {
        code: 'CDK_TOOLKIT_I5000',
        message: '  Deployment time: 9.67s',
    },
];
