// @aws-amplify/cli-core/lib/loggers/test-assets/noop_cdk_events.js
export const data = [
    {
        code: 'SYNTH_STARTED',
        action: 'amplify',
        message: 'Backend synthesis started',
    },
    {
        code: 'CDK_ASSEMBLY_I0000',
        message: 'Setting "CDK_DEFAULT_REGION" environment variable to [ \'us-west-2\' ]',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_ASSEMBLY_I0000',
        message: 'Setting "CDK_DEFAULT_ACCOUNT" environment variable to [ \'123456789012\' ]',
    },
    {
        code: 'CDK_ASSEMBLY_I0000',
        message: "outdir: [\n  '/Users/user/Workspaces/sample-project-app/.amplify/artifacts/cdk.out'\n]",
    },
    {
        code: 'CDK_ASSEMBLY_I0000',
        message: "env: [\n  {\n    CDK_DEFAULT_REGION: 'us-west-2',\n    CDK_DEFAULT_ACCOUNT: '123456789012',\n    CDK_OUTDIR: '/Users/user/Workspaces/sample-project-app/.amplify/artifacts/cdk.out',\n    CDK_CLI_ASM_VERSION: '41.0.0',\n    CDK_CLI_VERSION: '0.1.4'\n  }\n]",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "context: {\n  'amplify-backend-namespace': 'sample-project-app',\n  'amplify-backend-name': 'user',\n  'amplify-backend-type': 'sandbox',\n  'aws:cdk:enable-path-metadata': true,\n  'aws:cdk:enable-asset-metadata': true,\n  'aws:cdk:version-reporting': true,\n  'aws:cdk:bundling-stacks': [ '**' ]\n}",
    },
    {
        code: 'CDK_TOOLKIT_I1000',
        message: '  Synthesis time: 1.9s',
    },
    {
        code: 'CDK_TOOLKIT_I1901',
        message: 'Successfully synthesized to /Users/user/Workspaces/sample-project-app/.amplify/artifacts/cdk.out',
    },
    {
        code: 'SYNTH_FINISHED',
        action: 'amplify',
        level: 'result',
        message: 'Backend synthesized in 1.9 seconds',
    },
    {
        code: 'TS_STARTED',
        action: 'amplify',
        message: 'Backend type checks started',
    },
    {
        code: 'TS_FINISHED',
        action: 'amplify',
        level: 'result',
        message: 'Type checks completed in 4.79 seconds',
    },
    {
        code: 'CDK_TOOLKIT_I1000',
        message: '  Synthesis time: 0s',
    },
    {
        code: 'CDK_TOOLKIT_W0000',
        message: '⚠️ The --hotswap and --hotswap-fallback flags deliberately introduce CloudFormation drift to speed up deployments\n⚠️ They should only be used for development - never use them for your production Stacks!',
    },
    {
        code: 'CDK_TOOLKIT_I5210',
        message: 'Checking for previously published assets',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/4c6a2d29f01c6091bd1d9afe16e5849d456c96f17c3b215938c8067399532719.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/ca9699b9a2845c45e80cba1111f95ae2ac773af7d87163ae9c808ec547e38e00.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/a183ddccbd956316c38ef97177b8f088ef0826f62023323f5ae6053d348ccffc.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/9fcbe070ecd3023c5bf5b966fa9584757db9762eef123bad0820bd87591b2174.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c1721bcd774e27c514d3454b5be4f9bdd094c0161b57ddf053d618e3b0086a77.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/88d5b7081b0dc9c39fd0bffedf56f011b2a7cf1ce5cc227f47a1e94efee8067c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/08f4d557693d96c1a4efba0f9dc91330e4b19772fd5477c156468843e3d9cb5e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/cc01911d0269d4080ea57505dc445dfc315ef7ad85d3d9d4ea1357858bff451d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"4c6a2d29f01c6091bd1d9afe16e5849d456c96f17c3b215938c8067399532719.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/4c6a2d29f01c6091bd1d9afe16e5849d456c96f17c3b215938c8067399532719.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"ca9699b9a2845c45e80cba1111f95ae2ac773af7d87163ae9c808ec547e38e00.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/ca9699b9a2845c45e80cba1111f95ae2ac773af7d87163ae9c808ec547e38e00.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/47614d6b888488252a8a44b7911b69efcddb1eb2454ca31132d72d6444247204.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/b371d68ddfbdd11b176756ce33b2c4e79d26fcc2fd4b7c0c9e75d06efacee90e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"9fcbe070ecd3023c5bf5b966fa9584757db9762eef123bad0820bd87591b2174.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/9fcbe070ecd3023c5bf5b966fa9584757db9762eef123bad0820bd87591b2174.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"a183ddccbd956316c38ef97177b8f088ef0826f62023323f5ae6053d348ccffc.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/a183ddccbd956316c38ef97177b8f088ef0826f62023323f5ae6053d348ccffc.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c1721bcd774e27c514d3454b5be4f9bdd094c0161b57ddf053d618e3b0086a77.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c1721bcd774e27c514d3454b5be4f9bdd094c0161b57ddf053d618e3b0086a77.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"08f4d557693d96c1a4efba0f9dc91330e4b19772fd5477c156468843e3d9cb5e.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/08f4d557693d96c1a4efba0f9dc91330e4b19772fd5477c156468843e3d9cb5e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"88d5b7081b0dc9c39fd0bffedf56f011b2a7cf1ce5cc227f47a1e94efee8067c.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/88d5b7081b0dc9c39fd0bffedf56f011b2a7cf1ce5cc227f47a1e94efee8067c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/06db846fd14e6fc371f22b12b5545ba8e2dbfeda85d8c8d586c71c282166657b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/f4a52b72209a9dfa197b5e7367a5c378c5bb86de6e29ddd9e48b49a3fe54b249.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/1f5fed297da9c32ae3af922bf3a38ccf23b956078887d16891ec06c20c64722c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/245ed8ac5185462a6923b7e45a318fec4c9ba0c45d1d720abd5b20317aa3dc6a.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/474bf0776ec2164a13191d1a0a9e057154931e4918fea5086f49850d02a5371b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"b371d68ddfbdd11b176756ce33b2c4e79d26fcc2fd4b7c0c9e75d06efacee90e.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/b371d68ddfbdd11b176756ce33b2c4e79d26fcc2fd4b7c0c9e75d06efacee90e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/29e4bdb03133f87f2b1acb547959ffd4cfd22804026994c5b61db33084127353.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"cc01911d0269d4080ea57505dc445dfc315ef7ad85d3d9d4ea1357858bff451d.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/cc01911d0269d4080ea57505dc445dfc315ef7ad85d3d9d4ea1357858bff451d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/4f7907d1209a2c9953a0c053df402c634e359546d70c7cc5c2e8e21ea734880f.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"47614d6b888488252a8a44b7911b69efcddb1eb2454ca31132d72d6444247204.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/47614d6b888488252a8a44b7911b69efcddb1eb2454ca31132d72d6444247204.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/6f3a2a9d562ae3058cc1aed6741edbf5edcaef7b3df7b03f52478816a1c95ffb.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"06db846fd14e6fc371f22b12b5545ba8e2dbfeda85d8c8d586c71c282166657b.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/06db846fd14e6fc371f22b12b5545ba8e2dbfeda85d8c8d586c71c282166657b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"1f5fed297da9c32ae3af922bf3a38ccf23b956078887d16891ec06c20c64722c.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/1f5fed297da9c32ae3af922bf3a38ccf23b956078887d16891ec06c20c64722c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/fe3c43ada4b9d681a5e2312663ef7a73386424d73b73e51f8e2e9d4b50f7c502.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"245ed8ac5185462a6923b7e45a318fec4c9ba0c45d1d720abd5b20317aa3dc6a.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/245ed8ac5185462a6923b7e45a318fec4c9ba0c45d1d720abd5b20317aa3dc6a.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/e0cff47fb007f0bbf2a4e43ca256d6aa7ec109821769fd79fa7c5e83f0e7f9fc.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/6479af76f5a4d8b61b53dd3cee8e18ae192b629ad15ced7b342690a9ebbdd8c2.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"f4a52b72209a9dfa197b5e7367a5c378c5bb86de6e29ddd9e48b49a3fe54b249.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/f4a52b72209a9dfa197b5e7367a5c378c5bb86de6e29ddd9e48b49a3fe54b249.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/6ce8d75a422db2491407ccfc9807e5de43e86bed091909b805008adb8243b05d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"474bf0776ec2164a13191d1a0a9e057154931e4918fea5086f49850d02a5371b.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/474bf0776ec2164a13191d1a0a9e057154931e4918fea5086f49850d02a5371b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/3cf32e28e3985b88b598dacbcc9a81cbb12cab62b6507082eb20a0702d3c197e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"6f3a2a9d562ae3058cc1aed6741edbf5edcaef7b3df7b03f52478816a1c95ffb.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/6f3a2a9d562ae3058cc1aed6741edbf5edcaef7b3df7b03f52478816a1c95ffb.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/fface358117b1edafc4dbe09854c750ff9d1cad2ab9aed9cff18b5df478d3d5d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"4f7907d1209a2c9953a0c053df402c634e359546d70c7cc5c2e8e21ea734880f.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/4f7907d1209a2c9953a0c053df402c634e359546d70c7cc5c2e8e21ea734880f.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/725b7ed3971927a1e99c853643b4ca575c3d612fff9b50d3f616c4feeb79a7d7.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"29e4bdb03133f87f2b1acb547959ffd4cfd22804026994c5b61db33084127353.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/29e4bdb03133f87f2b1acb547959ffd4cfd22804026994c5b61db33084127353.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/97e4049e52fffaf523b49966a1225677c0fd7c09e296e1832b5bb95086e5c3e4.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"fe3c43ada4b9d681a5e2312663ef7a73386424d73b73e51f8e2e9d4b50f7c502.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/fe3c43ada4b9d681a5e2312663ef7a73386424d73b73e51f8e2e9d4b50f7c502.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"e0cff47fb007f0bbf2a4e43ca256d6aa7ec109821769fd79fa7c5e83f0e7f9fc.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/e0cff47fb007f0bbf2a4e43ca256d6aa7ec109821769fd79fa7c5e83f0e7f9fc.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"3cf32e28e3985b88b598dacbcc9a81cbb12cab62b6507082eb20a0702d3c197e.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/3cf32e28e3985b88b598dacbcc9a81cbb12cab62b6507082eb20a0702d3c197e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c89821d3a34d8f9a69feb7faded2de554cfee32627d772bc8d842a1a145408f5.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/098c998f215c333f57348ca8912f66057ae6b574fe34013033dc54362ee2372c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/5b2dd26ac19f6a54684f2769d465463159cc27a441d7ac8b8efafdec1b2370fd.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"6479af76f5a4d8b61b53dd3cee8e18ae192b629ad15ced7b342690a9ebbdd8c2.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/6479af76f5a4d8b61b53dd3cee8e18ae192b629ad15ced7b342690a9ebbdd8c2.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"725b7ed3971927a1e99c853643b4ca575c3d612fff9b50d3f616c4feeb79a7d7.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/725b7ed3971927a1e99c853643b4ca575c3d612fff9b50d3f616c4feeb79a7d7.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"6ce8d75a422db2491407ccfc9807e5de43e86bed091909b805008adb8243b05d.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/6ce8d75a422db2491407ccfc9807e5de43e86bed091909b805008adb8243b05d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"fface358117b1edafc4dbe09854c750ff9d1cad2ab9aed9cff18b5df478d3d5d.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/fface358117b1edafc4dbe09854c750ff9d1cad2ab9aed9cff18b5df478d3d5d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/db826cc870f25dcd05aff645c3b384ba3a97115acc5373e580c14379eb0430b3.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c3534669fc2486ce8786b1a3efd6c0848cd3c4835de33940c0f49c28c9d8abef.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/49e7a1f2fb4c809b82bed6b0242a2a3bcfa2c5276ebd2539ff330fee36f8fe18.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/dcb70e4856bc9a423e02b8b9e9c2ab3f71479784572d4c04a02d1f5b61a9f9ac.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"97e4049e52fffaf523b49966a1225677c0fd7c09e296e1832b5bb95086e5c3e4.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/97e4049e52fffaf523b49966a1225677c0fd7c09e296e1832b5bb95086e5c3e4.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/faa95a81ae7d7373f3e1f242268f904eb748d8d0fdd306e8a6fe515a1905a7d6.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c3534669fc2486ce8786b1a3efd6c0848cd3c4835de33940c0f49c28c9d8abef.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c3534669fc2486ce8786b1a3efd6c0848cd3c4835de33940c0f49c28c9d8abef.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c89821d3a34d8f9a69feb7faded2de554cfee32627d772bc8d842a1a145408f5.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c89821d3a34d8f9a69feb7faded2de554cfee32627d772bc8d842a1a145408f5.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"098c998f215c333f57348ca8912f66057ae6b574fe34013033dc54362ee2372c.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/098c998f215c333f57348ca8912f66057ae6b574fe34013033dc54362ee2372c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"5b2dd26ac19f6a54684f2769d465463159cc27a441d7ac8b8efafdec1b2370fd.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/5b2dd26ac19f6a54684f2769d465463159cc27a441d7ac8b8efafdec1b2370fd.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"db826cc870f25dcd05aff645c3b384ba3a97115acc5373e580c14379eb0430b3.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/db826cc870f25dcd05aff645c3b384ba3a97115acc5373e580c14379eb0430b3.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/14700f3f8dd2f4997b0e6380f2714c17996184ef4a12d7990ce58b009105e158.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"dcb70e4856bc9a423e02b8b9e9c2ab3f71479784572d4c04a02d1f5b61a9f9ac.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/dcb70e4856bc9a423e02b8b9e9c2ab3f71479784572d4c04a02d1f5b61a9f9ac.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"49e7a1f2fb4c809b82bed6b0242a2a3bcfa2c5276ebd2539ff330fee36f8fe18.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/49e7a1f2fb4c809b82bed6b0242a2a3bcfa2c5276ebd2539ff330fee36f8fe18.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"faa95a81ae7d7373f3e1f242268f904eb748d8d0fdd306e8a6fe515a1905a7d6.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/faa95a81ae7d7373f3e1f242268f904eb748d8d0fdd306e8a6fe515a1905a7d6.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/2f050e6b9affea69843658b09b5d9073acc9cda9170a30ff043def78a33c6ad0.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/907cb295088857ccb545b8aae83288bd9dfa82115bb0d20a0ae0d08dc8872f8a.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/aca5c01abb5bae3c9a0387478d7962aa5580a6ed97a5d12549501d7f1874b091.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/53b8945b363f121c80e07e812f42fa6e33a8f3fdbc9b75d4b3374959bb62f47d.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/193c5f412086be216636e53fa8e74d99c431936435b456c4b0b7252704331c51.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"14700f3f8dd2f4997b0e6380f2714c17996184ef4a12d7990ce58b009105e158.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/14700f3f8dd2f4997b0e6380f2714c17996184ef4a12d7990ce58b009105e158.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/209a0048df688fe6a8997016c796a69b318d368ae30d416e1656d209d881a050.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/f7bafc4fb5028045158fda2645f666d5744ff73a7236cbd21784e1b94aaef720.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"2f050e6b9affea69843658b09b5d9073acc9cda9170a30ff043def78a33c6ad0.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/2f050e6b9affea69843658b09b5d9073acc9cda9170a30ff043def78a33c6ad0.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"907cb295088857ccb545b8aae83288bd9dfa82115bb0d20a0ae0d08dc8872f8a.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/907cb295088857ccb545b8aae83288bd9dfa82115bb0d20a0ae0d08dc8872f8a.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"53b8945b363f121c80e07e812f42fa6e33a8f3fdbc9b75d4b3374959bb62f47d.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/53b8945b363f121c80e07e812f42fa6e33a8f3fdbc9b75d4b3374959bb62f47d.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/83ef1d85f437ccf9bf6d749fd82f93b06ce961419558f4618e4da3765db7a706.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/76f3f913f2bffd4d3b833199e6d331903453796a752c920c60d3aa8456e8a975.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/5065dd0842feb96b950d214ec1e2f9faabd3ad868f6d75d5d999b56e9327c4d6.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/a217e7c300711883159bfa09d4a6a99f762c9d59542ae5bb802ebac85abdb443.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"aca5c01abb5bae3c9a0387478d7962aa5580a6ed97a5d12549501d7f1874b091.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/aca5c01abb5bae3c9a0387478d7962aa5580a6ed97a5d12549501d7f1874b091.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"193c5f412086be216636e53fa8e74d99c431936435b456c4b0b7252704331c51.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/193c5f412086be216636e53fa8e74d99c431936435b456c4b0b7252704331c51.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/93d72ae98d8a9ba960cf939344d1444d6cde50e0a07de22c5d60cd6af42a9c00.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"209a0048df688fe6a8997016c796a69b318d368ae30d416e1656d209d881a050.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/209a0048df688fe6a8997016c796a69b318d368ae30d416e1656d209d881a050.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"76f3f913f2bffd4d3b833199e6d331903453796a752c920c60d3aa8456e8a975.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/76f3f913f2bffd4d3b833199e6d331903453796a752c920c60d3aa8456e8a975.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"5065dd0842feb96b950d214ec1e2f9faabd3ad868f6d75d5d999b56e9327c4d6.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/5065dd0842feb96b950d214ec1e2f9faabd3ad868f6d75d5d999b56e9327c4d6.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"f7bafc4fb5028045158fda2645f666d5744ff73a7236cbd21784e1b94aaef720.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/f7bafc4fb5028045158fda2645f666d5744ff73a7236cbd21784e1b94aaef720.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"a217e7c300711883159bfa09d4a6a99f762c9d59542ae5bb802ebac85abdb443.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/a217e7c300711883159bfa09d4a6a99f762c9d59542ae5bb802ebac85abdb443.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"93d72ae98d8a9ba960cf939344d1444d6cde50e0a07de22c5d60cd6af42a9c00.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/93d72ae98d8a9ba960cf939344d1444d6cde50e0a07de22c5d60cd6af42a9c00.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"83ef1d85f437ccf9bf6d749fd82f93b06ce961419558f4618e4da3765db7a706.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/83ef1d85f437ccf9bf6d749fd82f93b06ce961419558f4618e4da3765db7a706.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '50 total assets, 0 still need to be published',
    },
    {
        code: 'CDK_TOOLKIT_I5100',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: deploying... [1/1]',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-deploy-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-deploy-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] SSM.GetParameter({"Name":"/cdk-bootstrap/hnb659fds/version"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: checking if we can skip deploy',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: template has changed',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: deploying...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Waiting for stack CDKToolkit to finish creating or updating...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"CDKToolkit"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"CDKToolkit"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.GetCallerIdentity({}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-LDBNE1ZGD2C9"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-LDBNE1ZGD2C9","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1D9PUISB02OQG"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1D9PUISB02OQG","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1D9PUISB02OQG"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataAmplifyTableManagerNestedStackA-S36HIH9FEA0A"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataAmplifyTableManagerNestedStackA-S36HIH9FEA0A","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataPostNestedStackPostNestedStackR-59XZFM7X3ZNK"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataPostNestedStackPostNestedStackR-59XZFM7X3ZNK","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataPersonNestedStackPersonNestedSt-2U08ANIC9RIH"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataPersonNestedStackPersonNestedSt-2U08ANIC9RIH","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataConnectionStackNestedStackConne-1FA5HEOVF8WQF"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataConnectionStackNestedStackConne-1FA5HEOVF8WQF","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-function1351588B-1IVJW8S0RIZDH"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-function1351588B-1IVJW8S0RIZDH","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-storage0EC3F24A-4MGALTQV2LOL"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-storage0EC3F24A-4MGALTQV2LOL","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-11ZJAEIIDS4CU"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-11ZJAEIIDS4CU","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I5900',
        message: '  amplify-sampleprojectapp-user-sandbox-0e1752a889 (no changes)',
        data: {
            type: 'did-deploy-stack',
            noOp: true,
            stackArn: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/fd628630-fdc2-11ef-aea8-0aed8d33b907',
            outputs: {
                webClientId: '2fu223tmuj6usafa73tavsv3uj',
                socialProviders: '',
                usernameAttributes: '["email"]',
                bucketName: 'amplify-sampleprojectapp-pr-testnamebucketb4152ad5-j4pp16uhbrgf',
                signupAttributes: '["email"]',
                oauthClientId: '2fu223tmuj6usafa73tavsv3uj',
                allowUnauthenticatedIdentities: 'true',
                oauthCognitoDomain: '',
                oauthScope: '["profile","phone","email","openid","aws.cognito.signin.user.admin"]',
                oauthRedirectSignOut: '',
                awsAppsyncAdditionalAuthenticationTypes: 'AMAZON_COGNITO_USER_POOLS',
                mfaTypes: '[]',
                mfaConfiguration: 'OFF',
                storageRegion: 'us-west-2',
                oauthResponseType: 'code',
                awsAppsyncAuthenticationType: 'AWS_IAM',
                customOutputs: '{"custom":{"some":"keys2"},"version":"1.3"}',
                oauthRedirectSignIn: 'https://example.com',
                passwordPolicyMinLength: '8',
                awsAppsyncApiEndpoint: 'https://32p5frr6ejcm7ddxhrrxherspm.appsync-api.us-west-2.amazonaws.com/graphql',
                awsAppsyncApiId: 'becwkazf5zev3btd74ssfxl5ou',
                authRegion: 'us-west-2',
                buckets: '["{\\"name\\":\\"testName\\",\\"bucketName\\":\\"amplify-sampleprojectapp-pr-testnamebucketb4152ad5-j4pp16uhbrgf\\",\\"storageRegion\\":\\"us-west-2\\",\\"paths\\":{\\"public/*\\":{}}}"]',
                amplifyApiModelSchemaS3Uri: 's3://amplify-sampleprojectapp--amplifydataamplifycodege-s8jr4py2kgyv/model-schema.graphql',
                groups: '[]',
                definedFunctions: '["amplify-sampleprojectapp-pr-sayhellolambdaEFA46D92-lfDpc7SzfAjk"]',
                awsAppsyncRegion: 'us-west-2',
                deploymentType: 'sandbox',
                passwordPolicyRequirements: '["REQUIRES_NUMBERS","REQUIRES_LOWERCASE","REQUIRES_UPPERCASE","REQUIRES_SYMBOLS"]',
                region: 'us-west-2',
                userPoolId: 'us-west-2_ouu7G5ZED',
                identityPoolId: 'us-west-2:57620453-de0b-4241-b9f0-9c71fc41dd7d',
                verificationMechanisms: '["email"]',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5900',
        message: '  amplify-sampleprojectapp-user-sandbox-0e1752a889 (no changes)',
    },
    {
        code: 'CDK_TOOLKIT_I5000',
        message: '  Deployment time: 14.05s',
        data: {
            duration: 14053,
        },
    },
    {
        code: 'CDK_TOOLKIT_I5000',
        message: '  Deployment time: 14.05s',
    },
];
