// @aws-amplify/cli-core/lib/loggers/test-assets/update_amplify_outputs_structured_cdk_events.js
export const data = [
    {
        code: 'SYNTH_STARTED',
        action: 'amplify',
        message: 'Backend synthesis started',
    },
    {
        code: 'CDK_ASSEMBLY_I0000',
        message: 'Setting "CDK_DEFAULT_REGION" environment variable to [ \'us-west-2\' ]',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_ASSEMBLY_I0000',
        message: 'Setting "CDK_DEFAULT_ACCOUNT" environment variable to [ \'123456789012\' ]',
    },
    {
        code: 'CDK_ASSEMBLY_I0000',
        message: "outdir: [\n  '/Users/user/Workspaces/sample-project-app/.amplify/artifacts/cdk.out'\n]",
    },
    {
        code: 'CDK_ASSEMBLY_I0000',
        message: "env: [\n  {\n    CDK_DEFAULT_REGION: 'us-west-2',\n    CDK_DEFAULT_ACCOUNT: '123456789012',\n    CDK_OUTDIR: '/Users/user/Workspaces/sample-project-app/.amplify/artifacts/cdk.out',\n    CDK_CLI_ASM_VERSION: '41.0.0',\n    CDK_CLI_VERSION: '0.1.4'\n  }\n]",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "context: {\n  'amplify-backend-namespace': 'sample-project-app',\n  'amplify-backend-name': 'user',\n  'amplify-backend-type': 'sandbox',\n  'aws:cdk:enable-path-metadata': true,\n  'aws:cdk:enable-asset-metadata': true,\n  'aws:cdk:version-reporting': true,\n  'aws:cdk:bundling-stacks': [ '**' ]\n}",
    },
    {
        code: 'CDK_TOOLKIT_I1000',
        message: '  Synthesis time: 1.54s',
    },
    {
        code: 'CDK_TOOLKIT_I1901',
        message: 'Successfully synthesized to /Users/user/Workspaces/sample-project-app/.amplify/artifacts/cdk.out',
    },
    {
        code: 'SYNTH_FINISHED',
        action: 'amplify',
        level: 'result',
        message: 'Backend synthesized in 1.55 seconds',
    },
    {
        code: 'TS_STARTED',
        action: 'amplify',
        message: 'Backend type checks started',
    },
    {
        code: 'TS_FINISHED',
        action: 'amplify',
        level: 'result',
        message: 'Type checks completed in 4.9 seconds',
    },
    {
        code: 'CDK_TOOLKIT_I1000',
        message: '  Synthesis time: 0s',
    },
    {
        code: 'CDK_TOOLKIT_W0000',
        message: '⚠️ The --hotswap and --hotswap-fallback flags deliberately introduce CloudFormation drift to speed up deployments\n⚠️ They should only be used for development - never use them for your production Stacks!',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Checking for previously published assets',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/08f4d557693d96c1a4efba0f9dc91330e4b19772fd5477c156468843e3d9cb5e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/ca9699b9a2845c45e80cba1111f95ae2ac773af7d87163ae9c808ec547e38e00.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/88d5b7081b0dc9c39fd0bffedf56f011b2a7cf1ce5cc227f47a1e94efee8067c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c1721bcd774e27c514d3454b5be4f9bdd094c0161b57ddf053d618e3b0086a77.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/9fcbe070ecd3023c5bf5b966fa9584757db9762eef123bad0820bd87591b2174.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/cc01911d0269d4080ea57505dc445dfc315ef7ad85d3d9d4ea1357858bff451d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/a183ddccbd956316c38ef97177b8f088ef0826f62023323f5ae6053d348ccffc.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/4c6a2d29f01c6091bd1d9afe16e5849d456c96f17c3b215938c8067399532719.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"88d5b7081b0dc9c39fd0bffedf56f011b2a7cf1ce5cc227f47a1e94efee8067c.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/88d5b7081b0dc9c39fd0bffedf56f011b2a7cf1ce5cc227f47a1e94efee8067c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"ca9699b9a2845c45e80cba1111f95ae2ac773af7d87163ae9c808ec547e38e00.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/ca9699b9a2845c45e80cba1111f95ae2ac773af7d87163ae9c808ec547e38e00.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"9fcbe070ecd3023c5bf5b966fa9584757db9762eef123bad0820bd87591b2174.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/9fcbe070ecd3023c5bf5b966fa9584757db9762eef123bad0820bd87591b2174.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"cc01911d0269d4080ea57505dc445dfc315ef7ad85d3d9d4ea1357858bff451d.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/cc01911d0269d4080ea57505dc445dfc315ef7ad85d3d9d4ea1357858bff451d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c1721bcd774e27c514d3454b5be4f9bdd094c0161b57ddf053d618e3b0086a77.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c1721bcd774e27c514d3454b5be4f9bdd094c0161b57ddf053d618e3b0086a77.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"08f4d557693d96c1a4efba0f9dc91330e4b19772fd5477c156468843e3d9cb5e.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/08f4d557693d96c1a4efba0f9dc91330e4b19772fd5477c156468843e3d9cb5e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"4c6a2d29f01c6091bd1d9afe16e5849d456c96f17c3b215938c8067399532719.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/4c6a2d29f01c6091bd1d9afe16e5849d456c96f17c3b215938c8067399532719.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"a183ddccbd956316c38ef97177b8f088ef0826f62023323f5ae6053d348ccffc.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/a183ddccbd956316c38ef97177b8f088ef0826f62023323f5ae6053d348ccffc.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/b371d68ddfbdd11b176756ce33b2c4e79d26fcc2fd4b7c0c9e75d06efacee90e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/47614d6b888488252a8a44b7911b69efcddb1eb2454ca31132d72d6444247204.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/f4a52b72209a9dfa197b5e7367a5c378c5bb86de6e29ddd9e48b49a3fe54b249.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/06db846fd14e6fc371f22b12b5545ba8e2dbfeda85d8c8d586c71c282166657b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/1f5fed297da9c32ae3af922bf3a38ccf23b956078887d16891ec06c20c64722c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/245ed8ac5185462a6923b7e45a318fec4c9ba0c45d1d720abd5b20317aa3dc6a.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/29e4bdb03133f87f2b1acb547959ffd4cfd22804026994c5b61db33084127353.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/474bf0776ec2164a13191d1a0a9e057154931e4918fea5086f49850d02a5371b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"474bf0776ec2164a13191d1a0a9e057154931e4918fea5086f49850d02a5371b.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/474bf0776ec2164a13191d1a0a9e057154931e4918fea5086f49850d02a5371b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"1f5fed297da9c32ae3af922bf3a38ccf23b956078887d16891ec06c20c64722c.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/1f5fed297da9c32ae3af922bf3a38ccf23b956078887d16891ec06c20c64722c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"06db846fd14e6fc371f22b12b5545ba8e2dbfeda85d8c8d586c71c282166657b.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/06db846fd14e6fc371f22b12b5545ba8e2dbfeda85d8c8d586c71c282166657b.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"f4a52b72209a9dfa197b5e7367a5c378c5bb86de6e29ddd9e48b49a3fe54b249.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/f4a52b72209a9dfa197b5e7367a5c378c5bb86de6e29ddd9e48b49a3fe54b249.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"29e4bdb03133f87f2b1acb547959ffd4cfd22804026994c5b61db33084127353.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/29e4bdb03133f87f2b1acb547959ffd4cfd22804026994c5b61db33084127353.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"b371d68ddfbdd11b176756ce33b2c4e79d26fcc2fd4b7c0c9e75d06efacee90e.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/b371d68ddfbdd11b176756ce33b2c4e79d26fcc2fd4b7c0c9e75d06efacee90e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"47614d6b888488252a8a44b7911b69efcddb1eb2454ca31132d72d6444247204.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/47614d6b888488252a8a44b7911b69efcddb1eb2454ca31132d72d6444247204.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/6f3a2a9d562ae3058cc1aed6741edbf5edcaef7b3df7b03f52478816a1c95ffb.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/4f7907d1209a2c9953a0c053df402c634e359546d70c7cc5c2e8e21ea734880f.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/fe3c43ada4b9d681a5e2312663ef7a73386424d73b73e51f8e2e9d4b50f7c502.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/e0cff47fb007f0bbf2a4e43ca256d6aa7ec109821769fd79fa7c5e83f0e7f9fc.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"245ed8ac5185462a6923b7e45a318fec4c9ba0c45d1d720abd5b20317aa3dc6a.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/245ed8ac5185462a6923b7e45a318fec4c9ba0c45d1d720abd5b20317aa3dc6a.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/6479af76f5a4d8b61b53dd3cee8e18ae192b629ad15ced7b342690a9ebbdd8c2.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/6ce8d75a422db2491407ccfc9807e5de43e86bed091909b805008adb8243b05d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/3cf32e28e3985b88b598dacbcc9a81cbb12cab62b6507082eb20a0702d3c197e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/fface358117b1edafc4dbe09854c750ff9d1cad2ab9aed9cff18b5df478d3d5d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"4f7907d1209a2c9953a0c053df402c634e359546d70c7cc5c2e8e21ea734880f.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/4f7907d1209a2c9953a0c053df402c634e359546d70c7cc5c2e8e21ea734880f.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"6f3a2a9d562ae3058cc1aed6741edbf5edcaef7b3df7b03f52478816a1c95ffb.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/6f3a2a9d562ae3058cc1aed6741edbf5edcaef7b3df7b03f52478816a1c95ffb.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/725b7ed3971927a1e99c853643b4ca575c3d612fff9b50d3f616c4feeb79a7d7.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/97e4049e52fffaf523b49966a1225677c0fd7c09e296e1832b5bb95086e5c3e4.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"6ce8d75a422db2491407ccfc9807e5de43e86bed091909b805008adb8243b05d.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/6ce8d75a422db2491407ccfc9807e5de43e86bed091909b805008adb8243b05d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"e0cff47fb007f0bbf2a4e43ca256d6aa7ec109821769fd79fa7c5e83f0e7f9fc.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/e0cff47fb007f0bbf2a4e43ca256d6aa7ec109821769fd79fa7c5e83f0e7f9fc.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"fe3c43ada4b9d681a5e2312663ef7a73386424d73b73e51f8e2e9d4b50f7c502.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/fe3c43ada4b9d681a5e2312663ef7a73386424d73b73e51f8e2e9d4b50f7c502.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"6479af76f5a4d8b61b53dd3cee8e18ae192b629ad15ced7b342690a9ebbdd8c2.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/6479af76f5a4d8b61b53dd3cee8e18ae192b629ad15ced7b342690a9ebbdd8c2.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c89821d3a34d8f9a69feb7faded2de554cfee32627d772bc8d842a1a145408f5.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/098c998f215c333f57348ca8912f66057ae6b574fe34013033dc54362ee2372c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/5b2dd26ac19f6a54684f2769d465463159cc27a441d7ac8b8efafdec1b2370fd.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"fface358117b1edafc4dbe09854c750ff9d1cad2ab9aed9cff18b5df478d3d5d.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/fface358117b1edafc4dbe09854c750ff9d1cad2ab9aed9cff18b5df478d3d5d.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"3cf32e28e3985b88b598dacbcc9a81cbb12cab62b6507082eb20a0702d3c197e.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/3cf32e28e3985b88b598dacbcc9a81cbb12cab62b6507082eb20a0702d3c197e.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/db826cc870f25dcd05aff645c3b384ba3a97115acc5373e580c14379eb0430b3.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c3534669fc2486ce8786b1a3efd6c0848cd3c4835de33940c0f49c28c9d8abef.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/49e7a1f2fb4c809b82bed6b0242a2a3bcfa2c5276ebd2539ff330fee36f8fe18.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"725b7ed3971927a1e99c853643b4ca575c3d612fff9b50d3f616c4feeb79a7d7.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/725b7ed3971927a1e99c853643b4ca575c3d612fff9b50d3f616c4feeb79a7d7.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/dcb70e4856bc9a423e02b8b9e9c2ab3f71479784572d4c04a02d1f5b61a9f9ac.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"97e4049e52fffaf523b49966a1225677c0fd7c09e296e1832b5bb95086e5c3e4.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/97e4049e52fffaf523b49966a1225677c0fd7c09e296e1832b5bb95086e5c3e4.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/faa95a81ae7d7373f3e1f242268f904eb748d8d0fdd306e8a6fe515a1905a7d6.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"098c998f215c333f57348ca8912f66057ae6b574fe34013033dc54362ee2372c.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/098c998f215c333f57348ca8912f66057ae6b574fe34013033dc54362ee2372c.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"db826cc870f25dcd05aff645c3b384ba3a97115acc5373e580c14379eb0430b3.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/db826cc870f25dcd05aff645c3b384ba3a97115acc5373e580c14379eb0430b3.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c89821d3a34d8f9a69feb7faded2de554cfee32627d772bc8d842a1a145408f5.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c89821d3a34d8f9a69feb7faded2de554cfee32627d772bc8d842a1a145408f5.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/14700f3f8dd2f4997b0e6380f2714c17996184ef4a12d7990ce58b009105e158.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/2f050e6b9affea69843658b09b5d9073acc9cda9170a30ff043def78a33c6ad0.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c3534669fc2486ce8786b1a3efd6c0848cd3c4835de33940c0f49c28c9d8abef.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c3534669fc2486ce8786b1a3efd6c0848cd3c4835de33940c0f49c28c9d8abef.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"49e7a1f2fb4c809b82bed6b0242a2a3bcfa2c5276ebd2539ff330fee36f8fe18.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/49e7a1f2fb4c809b82bed6b0242a2a3bcfa2c5276ebd2539ff330fee36f8fe18.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"5b2dd26ac19f6a54684f2769d465463159cc27a441d7ac8b8efafdec1b2370fd.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/5b2dd26ac19f6a54684f2769d465463159cc27a441d7ac8b8efafdec1b2370fd.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/907cb295088857ccb545b8aae83288bd9dfa82115bb0d20a0ae0d08dc8872f8a.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/53b8945b363f121c80e07e812f42fa6e33a8f3fdbc9b75d4b3374959bb62f47d.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"14700f3f8dd2f4997b0e6380f2714c17996184ef4a12d7990ce58b009105e158.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/14700f3f8dd2f4997b0e6380f2714c17996184ef4a12d7990ce58b009105e158.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/e4431ba6094a85d53a88932fa10581de7e18e61c2310024b9907e52df11e26e6.graphql',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"faa95a81ae7d7373f3e1f242268f904eb748d8d0fdd306e8a6fe515a1905a7d6.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/faa95a81ae7d7373f3e1f242268f904eb748d8d0fdd306e8a6fe515a1905a7d6.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"dcb70e4856bc9a423e02b8b9e9c2ab3f71479784572d4c04a02d1f5b61a9f9ac.vtl","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/dcb70e4856bc9a423e02b8b9e9c2ab3f71479784572d4c04a02d1f5b61a9f9ac.vtl',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"2f050e6b9affea69843658b09b5d9073acc9cda9170a30ff043def78a33c6ad0.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/2f050e6b9affea69843658b09b5d9073acc9cda9170a30ff043def78a33c6ad0.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"907cb295088857ccb545b8aae83288bd9dfa82115bb0d20a0ae0d08dc8872f8a.zip","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/907cb295088857ccb545b8aae83288bd9dfa82115bb0d20a0ae0d08dc8872f8a.zip',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"53b8945b363f121c80e07e812f42fa6e33a8f3fdbc9b75d4b3374959bb62f47d.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/53b8945b363f121c80e07e812f42fa6e33a8f3fdbc9b75d4b3374959bb62f47d.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/76f3f913f2bffd4d3b833199e6d331903453796a752c920c60d3aa8456e8a975.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/83ef1d85f437ccf9bf6d749fd82f93b06ce961419558f4618e4da3765db7a706.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/193c5f412086be216636e53fa8e74d99c431936435b456c4b0b7252704331c51.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/209a0048df688fe6a8997016c796a69b318d368ae30d416e1656d209d881a050.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/e1beb14132c9a02797667cfecbe86315e0c25f051290acd8d3f29333d8e01af0.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/f7bafc4fb5028045158fda2645f666d5744ff73a7236cbd21784e1b94aaef720.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"83ef1d85f437ccf9bf6d749fd82f93b06ce961419558f4618e4da3765db7a706.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/83ef1d85f437ccf9bf6d749fd82f93b06ce961419558f4618e4da3765db7a706.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"76f3f913f2bffd4d3b833199e6d331903453796a752c920c60d3aa8456e8a975.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/76f3f913f2bffd4d3b833199e6d331903453796a752c920c60d3aa8456e8a975.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"e1beb14132c9a02797667cfecbe86315e0c25f051290acd8d3f29333d8e01af0.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"193c5f412086be216636e53fa8e74d99c431936435b456c4b0b7252704331c51.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/193c5f412086be216636e53fa8e74d99c431936435b456c4b0b7252704331c51.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"f7bafc4fb5028045158fda2645f666d5744ff73a7236cbd21784e1b94aaef720.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/f7bafc4fb5028045158fda2645f666d5744ff73a7236cbd21784e1b94aaef720.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"209a0048df688fe6a8997016c796a69b318d368ae30d416e1656d209d881a050.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: found: Found s3://cdk-hnb659fds-assets-123456789012-us-west-2/209a0048df688fe6a8997016c796a69b318d368ae30d416e1656d209d881a050.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '47 total assets, 1 still need to be published',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-deploy-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-deploy-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] SSM.GetParameter({"Name":"/cdk-bootstrap/hnb659fds/version"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5210',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: start: Building e1beb14132c9a02797667cfecbe86315e0c25f051290acd8d3f29333d8e01af0',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: success: Built e1beb14132c9a02797667cfecbe86315e0c25f051290acd8d3f29333d8e01af0',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"CDKToolkit"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: start: Publishing e1beb14132c9a02797667cfecbe86315e0c25f051290acd8d3f29333d8e01af0:current_account-current_region',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-file-publishing-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: check: Check s3://cdk-hnb659fds-assets-123456789012-us-west-2/e1beb14132c9a02797667cfecbe86315e0c25f051290acd8d3f29333d8e01af0.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.GetBucketLocation({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.ListObjectsV2({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Prefix":"e1beb14132c9a02797667cfecbe86315e0c25f051290acd8d3f29333d8e01af0.json","MaxKeys":1}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.GetBucketEncryption({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: upload: Upload s3://cdk-hnb659fds-assets-123456789012-us-west-2/e1beb14132c9a02797667cfecbe86315e0c25f051290acd8d3f29333d8e01af0.json',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] S3.PutObject({"Bucket":"cdk-hnb659fds-assets-123456789012-us-west-2","Key":"e1beb14132c9a02797667cfecbe86315e0c25f051290acd8d3f29333d8e01af0.json","Body":"<Buffer: 14 KiB>","ContentType":"application/json","ChecksumAlgorithm":"SHA256","ServerSideEncryption":"aws:kms"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5221',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: success: Published e1beb14132c9a02797667cfecbe86315e0c25f051290acd8d3f29333d8e01af0:current_account-current_region',
    },
    {
        code: 'CDK_TOOLKIT_I5100',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: deploying... [1/1]',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: "Assuming role 'arn:aws:iam::123456789012:role/cdk-hnb659fds-deploy-role-123456789012-us-west-2'.",
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.AssumeRole({"RoleArn":"arn:aws:iam::123456789012:role/cdk-hnb659fds-deploy-role-123456789012-us-west-2","RoleSessionName":"aws-cdk-user"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: checking if we can skip deploy',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: template has changed',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: deploying...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Waiting for stack CDKToolkit to finish creating or updating...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"CDKToolkit"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"CDKToolkit"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] STS.GetCallerIdentity({}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.ListStackResources({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataAmplifyTableManagerNestedStackA-1JMJ1D9EH891L"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataAmplifyTableManagerNestedStackA-1JMJ1D9EH891L","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataPostNestedStackPostNestedStackR-Q6BBZ46K7MMS"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataPostNestedStackPostNestedStackR-Q6BBZ46K7MMS","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataPersonNestedStackPersonNestedSt-72810Y0AHA1T"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataPersonNestedStackPersonNestedSt-72810Y0AHA1T","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataConnectionStackNestedStackConne-L130VQDWLV66"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0-amplifyDataConnectionStackNestedStackConne-L130VQDWLV66","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.GetTemplate({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR","TemplateStage":"Original"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Retrieved account ID 123456789012 from disk cache',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '⚠️ The following non-hotswappable changes were found:\n    logicalID: customOutputs, type: Stack Output, reason: output was changed',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Could not perform a hotswap deployment, as the stack amplify-sampleprojectapp-user-sandbox-0e1752a889 contains non-Asset changes',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Falling back to doing a full deployment',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Removing existing change set with name cdk-deploy-change-set if it exists',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DeleteChangeSet({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","ChangeSetName":"cdk-deploy-change-set"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Attempting to create ChangeSet with name cdk-deploy-change-set to update stack amplify-sampleprojectapp-user-sandbox-0e1752a889',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889: creating CloudFormation changeset...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.CreateChangeSet({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","ChangeSetName":"cdk-deploy-change-set","ChangeSetType":"UPDATE","Description":"CDK Changeset for execution 35dd3144-7c45-43b2-bbc0-5c85440a20c6","ClientToken":"create35dd3144-7c45-43b2-bbc0-5c85440a20c6","ImportExistingResources":false,"Capabilities":["CAPABILITY_IAM","CAPABILITY_NAMED_IAM","CAPABILITY_AUTO_EXPAND"],"Parameters":[],"RoleARN":"arn:aws:iam::123456789012:role/cdk-hnb659fds-cfn-exec-role-123456789012-us-west-2","TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/e1beb14132c9a02797667cfecbe86315e0c25f051290acd8d3f29333d8e01af0.json","Tags":[{"Key":"amplify:deployment-type","Value":"sandbox"},{"Key":"created-by","Value":"amplify"}]}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Initiated creation of changeset: arn:aws:cloudformation:us-west-2:123456789012:changeSet/cdk-deploy-change-set/0712ea73-02e0-4980-bc89-bb7996dfceb1; waiting for it to finish creating...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Waiting for changeset cdk-deploy-change-set on stack amplify-sampleprojectapp-user-sandbox-0e1752a889 to finish creating...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeChangeSet({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","ChangeSetName":"cdk-deploy-change-set"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Changeset cdk-deploy-change-set on stack amplify-sampleprojectapp-user-sandbox-0e1752a889 is still creating',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeChangeSet({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","ChangeSetName":"cdk-deploy-change-set"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Initiating execution of changeset arn:aws:cloudformation:us-west-2:123456789012:changeSet/cdk-deploy-change-set/0712ea73-02e0-4980-bc89-bb7996dfceb1 on stack amplify-sampleprojectapp-user-sandbox-0e1752a889',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.ExecuteChangeSet({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889","ChangeSetName":"cdk-deploy-change-set","ClientRequestToken":"exec35dd3144-7c45-43b2-bbc0-5c85440a20c6"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Execution of changeset arn:aws:cloudformation:us-west-2:123456789012:changeSet/cdk-deploy-change-set/0712ea73-02e0-4980-bc89-bb7996dfceb1 on stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has started; waiting for the update to complete...',
    },
    {
        code: 'CDK_TOOLKIT_I5501',
        message: 'Deploying amplify-sampleprojectapp-user-sandbox-0e1752a889',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Waiting for stack amplify-sampleprojectapp-user-sandbox-0e1752a889 to finish creating or updating...',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has an ongoing operation in progress and is not stable (UPDATE_IN_PROGRESS (User Initiated))',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 0/5 | 11:26:02 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | amplify-sampleprojectapp-user-sandbox-0e1752a889 User Initiated',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: '8f2861b0-fe28-11ef-af7a-029c4b5bb2cf',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:02.531'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceStatusReason: 'User Initiated',
                ClientRequestToken: 'exec35dd3144-7c45-43b2-bbc0-5c85440a20c6',
            },
            progress: {
                total: 5,
                completed: 0,
                formatted: '0/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has an ongoing operation in progress and is not stable (UPDATE_IN_PROGRESS)',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 0/5 | 11:26:05 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | auth.NestedStack/auth.NestedStackResource (auth179371D7)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'auth179371D7-UPDATE_IN_PROGRESS-2025-03-11T03:26:05.358Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'auth179371D7',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:05.358'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/53b8945b363f121c80e07e812f42fa6e33a8f3fdbc9b75d4b3374959bb62f47d.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec35dd3144-7c45-43b2-bbc0-5c85440a20c6',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'auth179371D7',
                },
                constructPath: 'auth.NestedStack/auth.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 0,
                formatted: '0/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 0/5 | 11:26:05 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | MyCustomResources.NestedStack/MyCustomResources.NestedStackResource (MyCustomResourcesBB610599)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'MyCustomResourcesBB610599-UPDATE_IN_PROGRESS-2025-03-11T03:26:05.379Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'MyCustomResourcesBB610599',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:05.379'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec35dd3144-7c45-43b2-bbc0-5c85440a20c6',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'MyCustomResourcesBB610599',
                },
                constructPath: 'MyCustomResources.NestedStack/MyCustomResources.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 0,
                formatted: '0/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 1/5 | 11:26:05 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | auth.NestedStack/auth.NestedStackResource (auth179371D7)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'auth179371D7-UPDATE_COMPLETE-2025-03-11T03:26:05.962Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'auth179371D7',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:05.962'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/53b8945b363f121c80e07e812f42fa6e33a8f3fdbc9b75d4b3374959bb62f47d.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec35dd3144-7c45-43b2-bbc0-5c85440a20c6',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'auth179371D7',
                },
                constructPath: 'auth.NestedStack/auth.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 1,
                formatted: '1/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 2/5 | 11:26:06 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | MyCustomResources.NestedStack/MyCustomResources.NestedStackResource (MyCustomResourcesBB610599)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'MyCustomResourcesBB610599-UPDATE_COMPLETE-2025-03-11T03:26:06.112Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'MyCustomResourcesBB610599',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:06.112'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec35dd3144-7c45-43b2-bbc0-5c85440a20c6',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'MyCustomResourcesBB610599',
                },
                constructPath: 'MyCustomResources.NestedStack/MyCustomResources.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 2,
                formatted: '2/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 2/5 | 11:26:06 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | data.NestedStack/data.NestedStackResource (data7552DF31)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'data7552DF31-UPDATE_IN_PROGRESS-2025-03-11T03:26:06.670Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'data7552DF31',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:06.670'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/76f3f913f2bffd4d3b833199e6d331903453796a752c920c60d3aa8456e8a975.json","Parameters":{"referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthIdentityPool5160F57DRef":"us-west-2:6568ef94-3300-4149-bc98-f2266342f0bb","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthUserPool0A15674BRef":"us-west-2_j6BNFhRJd","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthunauthenticatedUserRoleC92E0CFERef":"amplify-sampleprojectapp--amplifyAuthunauthenticate-Rieux6W03N47","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthauthenticatedUserRole11751600Ref":"amplify-sampleprojectapp--amplifyAuthauthenticatedU-y26Xh6dN3Ku9"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec35dd3144-7c45-43b2-bbc0-5c85440a20c6',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'data7552DF31',
                },
                constructPath: 'data.NestedStack/data.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 2,
                formatted: '2/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 2/5 | 11:26:07 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ User Initiated',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: '91fc6940-fe28-11ef-85cb-06d294525831',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:07.299'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceStatusReason: 'User Initiated',
            },
            progress: {
                total: 5,
                completed: 2,
                formatted: '2/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has an ongoing operation in progress and is not stable (UPDATE_IN_PROGRESS)',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataPostNestedStackPostNestedStackR-Q6BBZ46K7MMS/8f707920-fe26-11ef-b737-06b888091933"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataPersonNestedStackPersonNestedSt-72810Y0AHA1T/8f510a40-fe26-11ef-978a-06d488f07943"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataAmplifyTableManagerNestedStackA-1JMJ1D9EH891L/60c880e0-fe26-11ef-ab14-0a8a0b41c98f"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 2/5 | 11:26:11 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | data/amplifyData/AmplifyTableManager.NestedStack/AmplifyTableManager.NestedStackResource (amplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource86290833)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: 'amplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource86290833-UPDATE_IN_PROGRESS-2025-03-11T03:26:11.423Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource86290833',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataAmplifyTableManagerNestedStackA-1JMJ1D9EH891L/60c880e0-fe26-11ef-ab14-0a8a0b41c98f',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:11.423'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/193c5f412086be216636e53fa8e74d99c431936435b456c4b0b7252704331c51.json","Parameters":{"referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPIB1F14B28ApiId":"viu3cvmcnvdx7dufpcbaimflau"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplifyData","Key":"amplify:friendly-name"},{"Value":"amplify","Key":"created-by"}]}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource86290833',
                },
                constructPath: 'data/amplifyData/AmplifyTableManager.NestedStack/AmplifyTableManager.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 2,
                formatted: '2/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 3/5 | 11:26:12 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | data/amplifyData/AmplifyTableManager.NestedStack/AmplifyTableManager.NestedStackResource (amplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource86290833)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: 'amplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource86290833-UPDATE_COMPLETE-2025-03-11T03:26:12.151Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource86290833',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataAmplifyTableManagerNestedStackA-1JMJ1D9EH891L/60c880e0-fe26-11ef-ab14-0a8a0b41c98f',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:12.151'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/193c5f412086be216636e53fa8e74d99c431936435b456c4b0b7252704331c51.json","Parameters":{"referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPIB1F14B28ApiId":"viu3cvmcnvdx7dufpcbaimflau"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplifyData","Key":"amplify:friendly-name"},{"Value":"amplify","Key":"created-by"}]}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource86290833',
                },
                constructPath: 'data/amplifyData/AmplifyTableManager.NestedStack/AmplifyTableManager.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 3,
                formatted: '3/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 3/5 | 11:26:12 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | data/amplifyData/Person.NestedStack/Person.NestedStackResource (amplifyDataPersonNestedStackPersonNestedStackResource28D60818)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: 'amplifyDataPersonNestedStackPersonNestedStackResource28D60818-UPDATE_IN_PROGRESS-2025-03-11T03:26:12.944Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplifyDataPersonNestedStackPersonNestedStackResource28D60818',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataPersonNestedStackPersonNestedSt-72810Y0AHA1T/8f510a40-fe26-11ef-978a-06d488f07943',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:12.944'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/f7bafc4fb5028045158fda2645f666d5744ff73a7236cbd21784e1b94aaef720.json","Parameters":{"DynamoDBModelTableReadIOPS":"5","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource9097D1B4Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataAmplifyTableManagerT93BF9650":"arn:aws:lambda:us-west-2:123456789012:function:amplify-sampleprojectapp--TableManagerCustomProvid-dJlbgplvtxTp","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthIdentityPool5160F57DRef":"us-west-2:6568ef94-3300-4149-bc98-f2266342f0bb","DynamoDBEnableServerSideEncryption":"true","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPIB1F14B28ApiId":"viu3cvmcnvdx7dufpcbaimflau","DynamoDBEnablePointInTimeRecovery":"false","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthunauthenticatedUserRoleC92E0CFERef":"amplify-sampleprojectapp--amplifyAuthunauthenticate-Rieux6W03N47","DynamoDBBillingMode":"PAY_PER_REQUEST","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPINONEDS97A6DFB0Name":"NONE_DS","DynamoDBModelTableWriteIOPS":"5","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthauthenticatedUserRole11751600Ref":"amplify-sampleprojectapp--amplifyAuthauthenticatedU-y26Xh6dN3Ku9"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplifyData","Key":"amplify:friendly-name"},{"Value":"amplify","Key":"created-by"}]}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyDataPersonNestedStackPersonNestedStackResource28D60818',
                },
                constructPath: 'data/amplifyData/Person.NestedStack/Person.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 3,
                formatted: '3/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 3/5 | 11:26:13 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | data/amplifyData/Post.NestedStack/Post.NestedStackResource (amplifyDataPostNestedStackPostNestedStackResourceB65AFCD3)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: 'amplifyDataPostNestedStackPostNestedStackResourceB65AFCD3-UPDATE_IN_PROGRESS-2025-03-11T03:26:13.025Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplifyDataPostNestedStackPostNestedStackResourceB65AFCD3',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataPostNestedStackPostNestedStackR-Q6BBZ46K7MMS/8f707920-fe26-11ef-b737-06b888091933',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:13.025'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/209a0048df688fe6a8997016c796a69b318d368ae30d416e1656d209d881a050.json","Parameters":{"DynamoDBModelTableReadIOPS":"5","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource9097D1B4Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataAmplifyTableManagerT93BF9650":"arn:aws:lambda:us-west-2:123456789012:function:amplify-sampleprojectapp--TableManagerCustomProvid-dJlbgplvtxTp","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthIdentityPool5160F57DRef":"us-west-2:6568ef94-3300-4149-bc98-f2266342f0bb","DynamoDBEnableServerSideEncryption":"true","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPIB1F14B28ApiId":"viu3cvmcnvdx7dufpcbaimflau","DynamoDBEnablePointInTimeRecovery":"false","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthunauthenticatedUserRoleC92E0CFERef":"amplify-sampleprojectapp--amplifyAuthunauthenticate-Rieux6W03N47","DynamoDBBillingMode":"PAY_PER_REQUEST","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPINONEDS97A6DFB0Name":"NONE_DS","DynamoDBModelTableWriteIOPS":"5","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthauthenticatedUserRole11751600Ref":"amplify-sampleprojectapp--amplifyAuthauthenticatedU-y26Xh6dN3Ku9"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplifyData","Key":"amplify:friendly-name"},{"Value":"amplify","Key":"created-by"}]}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyDataPostNestedStackPostNestedStackResourceB65AFCD3',
                },
                constructPath: 'data/amplifyData/Post.NestedStack/Post.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 3,
                formatted: '3/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has an ongoing operation in progress and is not stable (UPDATE_COMPLETE_CLEANUP_IN_PROGRESS)',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has an ongoing operation in progress and is not stable (UPDATE_COMPLETE_CLEANUP_IN_PROGRESS)',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataPostNestedStackPostNestedStackR-Q6BBZ46K7MMS/8f707920-fe26-11ef-b737-06b888091933"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataPersonNestedStackPersonNestedSt-72810Y0AHA1T/8f510a40-fe26-11ef-978a-06d488f07943"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataAmplifyTableManagerNestedStackA-1JMJ1D9EH891L/60c880e0-fe26-11ef-ab14-0a8a0b41c98f"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataConnectionStackNestedStackConne-L130VQDWLV66/b460a660-fe26-11ef-a711-06b25169542f"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 4/5 | 11:26:13 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | data/amplifyData/Person.NestedStack/Person.NestedStackResource (amplifyDataPersonNestedStackPersonNestedStackResource28D60818)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: 'amplifyDataPersonNestedStackPersonNestedStackResource28D60818-UPDATE_COMPLETE-2025-03-11T03:26:13.723Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplifyDataPersonNestedStackPersonNestedStackResource28D60818',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataPersonNestedStackPersonNestedSt-72810Y0AHA1T/8f510a40-fe26-11ef-978a-06d488f07943',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:13.723'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/f7bafc4fb5028045158fda2645f666d5744ff73a7236cbd21784e1b94aaef720.json","Parameters":{"DynamoDBModelTableReadIOPS":"5","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource9097D1B4Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataAmplifyTableManagerT93BF9650":"arn:aws:lambda:us-west-2:123456789012:function:amplify-sampleprojectapp--TableManagerCustomProvid-dJlbgplvtxTp","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthIdentityPool5160F57DRef":"us-west-2:6568ef94-3300-4149-bc98-f2266342f0bb","DynamoDBEnableServerSideEncryption":"true","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPIB1F14B28ApiId":"viu3cvmcnvdx7dufpcbaimflau","DynamoDBEnablePointInTimeRecovery":"false","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthunauthenticatedUserRoleC92E0CFERef":"amplify-sampleprojectapp--amplifyAuthunauthenticate-Rieux6W03N47","DynamoDBBillingMode":"PAY_PER_REQUEST","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPINONEDS97A6DFB0Name":"NONE_DS","DynamoDBModelTableWriteIOPS":"5","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthauthenticatedUserRole11751600Ref":"amplify-sampleprojectapp--amplifyAuthauthenticatedU-y26Xh6dN3Ku9"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplifyData","Key":"amplify:friendly-name"},{"Value":"amplify","Key":"created-by"}]}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyDataPersonNestedStackPersonNestedStackResource28D60818',
                },
                constructPath: 'data/amplifyData/Person.NestedStack/Person.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 4,
                formatted: '4/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 5/5 | 11:26:13 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | data/amplifyData/Post.NestedStack/Post.NestedStackResource (amplifyDataPostNestedStackPostNestedStackResourceB65AFCD3)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: 'amplifyDataPostNestedStackPostNestedStackResourceB65AFCD3-UPDATE_COMPLETE-2025-03-11T03:26:13.855Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplifyDataPostNestedStackPostNestedStackResourceB65AFCD3',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataPostNestedStackPostNestedStackR-Q6BBZ46K7MMS/8f707920-fe26-11ef-b737-06b888091933',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:13.855'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/209a0048df688fe6a8997016c796a69b318d368ae30d416e1656d209d881a050.json","Parameters":{"DynamoDBModelTableReadIOPS":"5","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource9097D1B4Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataAmplifyTableManagerT93BF9650":"arn:aws:lambda:us-west-2:123456789012:function:amplify-sampleprojectapp--TableManagerCustomProvid-dJlbgplvtxTp","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthIdentityPool5160F57DRef":"us-west-2:6568ef94-3300-4149-bc98-f2266342f0bb","DynamoDBEnableServerSideEncryption":"true","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPIB1F14B28ApiId":"viu3cvmcnvdx7dufpcbaimflau","DynamoDBEnablePointInTimeRecovery":"false","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthunauthenticatedUserRoleC92E0CFERef":"amplify-sampleprojectapp--amplifyAuthunauthenticate-Rieux6W03N47","DynamoDBBillingMode":"PAY_PER_REQUEST","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPINONEDS97A6DFB0Name":"NONE_DS","DynamoDBModelTableWriteIOPS":"5","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthauthenticatedUserRole11751600Ref":"amplify-sampleprojectapp--amplifyAuthauthenticatedU-y26Xh6dN3Ku9"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplifyData","Key":"amplify:friendly-name"},{"Value":"amplify","Key":"created-by"}]}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyDataPostNestedStackPostNestedStackResourceB65AFCD3',
                },
                constructPath: 'data/amplifyData/Post.NestedStack/Post.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 5,
                formatted: '5/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 5/5 | 11:26:14 PM | UPDATE_IN_PROGRESS | AWS::CloudFormation::Stack | data/amplifyData/ConnectionStack.NestedStack/ConnectionStack.NestedStackResource (amplifyDataConnectionStackNestedStackConnectionStackNestedStackResourceAB0F312B)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: 'amplifyDataConnectionStackNestedStackConnectionStackNestedStackResourceAB0F312B-UPDATE_IN_PROGRESS-2025-03-11T03:26:14.548Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplifyDataConnectionStackNestedStackConnectionStackNestedStackResourceAB0F312B',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataConnectionStackNestedStackConne-L130VQDWLV66/b460a660-fe26-11ef-a711-06b25169542f',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:14.548'),
                ResourceStatus: 'UPDATE_IN_PROGRESS',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/83ef1d85f437ccf9bf6d749fd82f93b06ce961419558f4618e4da3765db7a706.json","Parameters":{"referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPostNestedStackPostNestedStackResource01FEDE53Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPostPostTable365E2F9CTableArn":"arn:aws:dynamodb:us-west-2:123456789012:table/Post-viu3cvmcnvdx7dufpcbaimflau-NONE","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPersonNestedStackPersonNestedStackResource35C2DCE5Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPersonPersonTableF02A70DBTableArn":"arn:aws:dynamodb:us-west-2:123456789012:table/Person-viu3cvmcnvdx7dufpcbaimflau-NONE","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthIdentityPool5160F57DRef":"us-west-2:6568ef94-3300-4149-bc98-f2266342f0bb","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPersonNestedStackPersonNestedStackResource35C2DCE5Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPersonPersonDataSourceDA3812F2Name":"PersonTable","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPIB1F14B28ApiId":"viu3cvmcnvdx7dufpcbaimflau","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthunauthenticatedUserRoleC92E0CFERef":"amplify-sampleprojectapp--amplifyAuthunauthenticate-Rieux6W03N47","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPINONEDS97A6DFB0Name":"NONE_DS","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthauthenticatedUserRole11751600Ref":"amplify-sampleprojectapp--amplifyAuthauthenticatedU-y26Xh6dN3Ku9","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPostNestedStackPostNestedStackResource01FEDE53Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPostPostDataSourceED97F46CName":"PostTable"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplifyData","Key":"amplify:friendly-name"},{"Value":"amplify","Key":"created-by"}]}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyDataConnectionStackNestedStackConnectionStackNestedStackResourceAB0F312B',
                },
                constructPath: 'data/amplifyData/ConnectionStack.NestedStack/ConnectionStack.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 5,
                formatted: '5/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 6/5 | 11:26:15 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | data/amplifyData/ConnectionStack.NestedStack/ConnectionStack.NestedStackResource (amplifyDataConnectionStackNestedStackConnectionStackNestedStackResourceAB0F312B)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: 'amplifyDataConnectionStackNestedStackConnectionStackNestedStackResourceAB0F312B-UPDATE_COMPLETE-2025-03-11T03:26:15.223Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplifyDataConnectionStackNestedStackConnectionStackNestedStackResourceAB0F312B',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataConnectionStackNestedStackConne-L130VQDWLV66/b460a660-fe26-11ef-a711-06b25169542f',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:15.223'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/83ef1d85f437ccf9bf6d749fd82f93b06ce961419558f4618e4da3765db7a706.json","Parameters":{"referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPostNestedStackPostNestedStackResource01FEDE53Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPostPostTable365E2F9CTableArn":"arn:aws:dynamodb:us-west-2:123456789012:table/Post-viu3cvmcnvdx7dufpcbaimflau-NONE","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPersonNestedStackPersonNestedStackResource35C2DCE5Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPersonPersonTableF02A70DBTableArn":"arn:aws:dynamodb:us-west-2:123456789012:table/Person-viu3cvmcnvdx7dufpcbaimflau-NONE","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthIdentityPool5160F57DRef":"us-west-2:6568ef94-3300-4149-bc98-f2266342f0bb","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPersonNestedStackPersonNestedStackResource35C2DCE5Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPersonPersonDataSourceDA3812F2Name":"PersonTable","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPIB1F14B28ApiId":"viu3cvmcnvdx7dufpcbaimflau","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthunauthenticatedUserRoleC92E0CFERef":"amplify-sampleprojectapp--amplifyAuthunauthenticate-Rieux6W03N47","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPINONEDS97A6DFB0Name":"NONE_DS","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthauthenticatedUserRole11751600Ref":"amplify-sampleprojectapp--amplifyAuthauthenticatedU-y26Xh6dN3Ku9","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPostNestedStackPostNestedStackResource01FEDE53Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPostPostDataSourceED97F46CName":"PostTable"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplifyData","Key":"amplify:friendly-name"},{"Value":"amplify","Key":"created-by"}]}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyDataConnectionStackNestedStackConnectionStackNestedStackResourceAB0F312B',
                },
                constructPath: 'data/amplifyData/ConnectionStack.NestedStack/ConnectionStack.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 6,
                formatted: '6/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 7/5 | 11:26:16 PM | UPDATE_COMPLETE_CLEANUP_IN_PROGRESS | AWS::CloudFormation::Stack | amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: '979f9660-fe28-11ef-86f3-020b31df2b65',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:16.764'),
                ResourceStatus: 'UPDATE_COMPLETE_CLEANUP_IN_PROGRESS',
            },
            progress: {
                total: 5,
                completed: 7,
                formatted: '7/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 8/5 | 11:26:17 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | data.NestedStack/data.NestedStackResource (data7552DF31)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'data7552DF31-UPDATE_COMPLETE-2025-03-11T03:26:17.634Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'data7552DF31',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:17.634'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/76f3f913f2bffd4d3b833199e6d331903453796a752c920c60d3aa8456e8a975.json","Parameters":{"referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthIdentityPool5160F57DRef":"us-west-2:6568ef94-3300-4149-bc98-f2266342f0bb","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthUserPool0A15674BRef":"us-west-2_j6BNFhRJd","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthunauthenticatedUserRoleC92E0CFERef":"amplify-sampleprojectapp--amplifyAuthunauthenticate-Rieux6W03N47","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthauthenticatedUserRole11751600Ref":"amplify-sampleprojectapp--amplifyAuthauthenticatedU-y26Xh6dN3Ku9"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec35dd3144-7c45-43b2-bbc0-5c85440a20c6',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'data7552DF31',
                },
                constructPath: 'data.NestedStack/data.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 8,
                formatted: '8/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 9/5 | 11:26:18 PM | UPDATE_COMPLETE_CLEANUP_IN_PROGRESS | AWS::CloudFormation::Stack | amplify-sampleprojectapp-user-sandbox-0e1752a889',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: '98ca81d0-fe28-11ef-9669-02f0541bd1f7',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:18.725'),
                ResourceStatus: 'UPDATE_COMPLETE_CLEANUP_IN_PROGRESS',
                ClientRequestToken: 'exec35dd3144-7c45-43b2-bbc0-5c85440a20c6',
            },
            progress: {
                total: 5,
                completed: 9,
                formatted: '9/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 8/5 | 11:26:19 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | MyCustomResources.NestedStack/MyCustomResources.NestedStackResource (MyCustomResourcesBB610599)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'MyCustomResourcesBB610599-UPDATE_COMPLETE-2025-03-11T03:26:19.805Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'MyCustomResourcesBB610599',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:19.805'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/2b0cd76a5fea2795b3e3b23e2c4cb8eaf621211d2c9e2303e2868ac0be4607bc.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec35dd3144-7c45-43b2-bbc0-5c85440a20c6',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'MyCustomResourcesBB610599',
                },
                constructPath: 'MyCustomResources.NestedStack/MyCustomResources.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 8,
                formatted: '8/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 7/5 | 11:26:21 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | data/amplifyData/ConnectionStack.NestedStack/ConnectionStack.NestedStackResource (amplifyDataConnectionStackNestedStackConnectionStackNestedStackResourceAB0F312B)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: 'amplifyDataConnectionStackNestedStackConnectionStackNestedStackResourceAB0F312B-UPDATE_COMPLETE-2025-03-11T03:26:21.127Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplifyDataConnectionStackNestedStackConnectionStackNestedStackResourceAB0F312B',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataConnectionStackNestedStackConne-L130VQDWLV66/b460a660-fe26-11ef-a711-06b25169542f',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:21.127'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/83ef1d85f437ccf9bf6d749fd82f93b06ce961419558f4618e4da3765db7a706.json","Parameters":{"referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPostNestedStackPostNestedStackResource01FEDE53Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPostPostTable365E2F9CTableArn":"arn:aws:dynamodb:us-west-2:123456789012:table/Post-viu3cvmcnvdx7dufpcbaimflau-NONE","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPersonNestedStackPersonNestedStackResource35C2DCE5Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPersonPersonTableF02A70DBTableArn":"arn:aws:dynamodb:us-west-2:123456789012:table/Person-viu3cvmcnvdx7dufpcbaimflau-NONE","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthIdentityPool5160F57DRef":"us-west-2:6568ef94-3300-4149-bc98-f2266342f0bb","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPersonNestedStackPersonNestedStackResource35C2DCE5Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPersonPersonDataSourceDA3812F2Name":"PersonTable","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPIB1F14B28ApiId":"viu3cvmcnvdx7dufpcbaimflau","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthunauthenticatedUserRoleC92E0CFERef":"amplify-sampleprojectapp--amplifyAuthunauthenticate-Rieux6W03N47","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPINONEDS97A6DFB0Name":"NONE_DS","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthauthenticatedUserRole11751600Ref":"amplify-sampleprojectapp--amplifyAuthauthenticatedU-y26Xh6dN3Ku9","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPostNestedStackPostNestedStackResource01FEDE53Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataPostPostDataSourceED97F46CName":"PostTable"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplifyData","Key":"amplify:friendly-name"},{"Value":"amplify","Key":"created-by"}]}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyDataConnectionStackNestedStackConnectionStackNestedStackResourceAB0F312B',
                },
                constructPath: 'data/amplifyData/ConnectionStack.NestedStack/ConnectionStack.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 7,
                formatted: '7/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 6/5 | 11:26:21 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | data/amplifyData/Post.NestedStack/Post.NestedStackResource (amplifyDataPostNestedStackPostNestedStackResourceB65AFCD3)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: 'amplifyDataPostNestedStackPostNestedStackResourceB65AFCD3-UPDATE_COMPLETE-2025-03-11T03:26:21.541Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplifyDataPostNestedStackPostNestedStackResourceB65AFCD3',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataPostNestedStackPostNestedStackR-Q6BBZ46K7MMS/8f707920-fe26-11ef-b737-06b888091933',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:21.541'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/209a0048df688fe6a8997016c796a69b318d368ae30d416e1656d209d881a050.json","Parameters":{"DynamoDBModelTableReadIOPS":"5","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource9097D1B4Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataAmplifyTableManagerT93BF9650":"arn:aws:lambda:us-west-2:123456789012:function:amplify-sampleprojectapp--TableManagerCustomProvid-dJlbgplvtxTp","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthIdentityPool5160F57DRef":"us-west-2:6568ef94-3300-4149-bc98-f2266342f0bb","DynamoDBEnableServerSideEncryption":"true","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPIB1F14B28ApiId":"viu3cvmcnvdx7dufpcbaimflau","DynamoDBEnablePointInTimeRecovery":"false","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthunauthenticatedUserRoleC92E0CFERef":"amplify-sampleprojectapp--amplifyAuthunauthenticate-Rieux6W03N47","DynamoDBBillingMode":"PAY_PER_REQUEST","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPINONEDS97A6DFB0Name":"NONE_DS","DynamoDBModelTableWriteIOPS":"5","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthauthenticatedUserRole11751600Ref":"amplify-sampleprojectapp--amplifyAuthauthenticatedU-y26Xh6dN3Ku9"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplifyData","Key":"amplify:friendly-name"},{"Value":"amplify","Key":"created-by"}]}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyDataPostNestedStackPostNestedStackResourceB65AFCD3',
                },
                constructPath: 'data/amplifyData/Post.NestedStack/Post.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 6,
                formatted: '6/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 5/5 | 11:26:21 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | data/amplifyData/Person.NestedStack/Person.NestedStackResource (amplifyDataPersonNestedStackPersonNestedStackResource28D60818)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: 'amplifyDataPersonNestedStackPersonNestedStackResource28D60818-UPDATE_COMPLETE-2025-03-11T03:26:21.550Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplifyDataPersonNestedStackPersonNestedStackResource28D60818',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataPersonNestedStackPersonNestedSt-72810Y0AHA1T/8f510a40-fe26-11ef-978a-06d488f07943',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:21.550'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/f7bafc4fb5028045158fda2645f666d5744ff73a7236cbd21784e1b94aaef720.json","Parameters":{"DynamoDBModelTableReadIOPS":"5","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource9097D1B4Outputsamplifysampleprojectappusersandbox0e1752a889dataamplifyDataAmplifyTableManagerT93BF9650":"arn:aws:lambda:us-west-2:123456789012:function:amplify-sampleprojectapp--TableManagerCustomProvid-dJlbgplvtxTp","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthIdentityPool5160F57DRef":"us-west-2:6568ef94-3300-4149-bc98-f2266342f0bb","DynamoDBEnableServerSideEncryption":"true","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPIB1F14B28ApiId":"viu3cvmcnvdx7dufpcbaimflau","DynamoDBEnablePointInTimeRecovery":"false","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthunauthenticatedUserRoleC92E0CFERef":"amplify-sampleprojectapp--amplifyAuthunauthenticate-Rieux6W03N47","DynamoDBBillingMode":"PAY_PER_REQUEST","referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPINONEDS97A6DFB0Name":"NONE_DS","DynamoDBModelTableWriteIOPS":"5","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthauthenticatedUserRole11751600Ref":"amplify-sampleprojectapp--amplifyAuthauthenticatedU-y26Xh6dN3Ku9"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplifyData","Key":"amplify:friendly-name"},{"Value":"amplify","Key":"created-by"}]}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyDataPersonNestedStackPersonNestedStackResource28D60818',
                },
                constructPath: 'data/amplifyData/Person.NestedStack/Person.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 5,
                formatted: '5/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 4/5 | 11:26:21 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | data/amplifyData/AmplifyTableManager.NestedStack/AmplifyTableManager.NestedStackResource (amplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource86290833)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: 'amplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource86290833-UPDATE_COMPLETE-2025-03-11T03:26:21.964Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource86290833',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataAmplifyTableManagerNestedStackA-1JMJ1D9EH891L/60c880e0-fe26-11ef-ab14-0a8a0b41c98f',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:21.964'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/193c5f412086be216636e53fa8e74d99c431936435b456c4b0b7252704331c51.json","Parameters":{"referencetoamplifysampleprojectappusersandbox0e1752a889dataamplifyDataGraphQLAPIB1F14B28ApiId":"viu3cvmcnvdx7dufpcbaimflau"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplifyData","Key":"amplify:friendly-name"},{"Value":"amplify","Key":"created-by"}]}',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'amplifyDataAmplifyTableManagerNestedStackAmplifyTableManagerNestedStackResource86290833',
                },
                constructPath: 'data/amplifyData/AmplifyTableManager.NestedStack/AmplifyTableManager.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 4,
                formatted: '4/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: 'Stack amplify-sampleprojectapp-user-sandbox-0e1752a889 has an ongoing operation in progress and is not stable (UPDATE_COMPLETE_CLEANUP_IN_PROGRESS)',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataPostNestedStackPostNestedStackR-Q6BBZ46K7MMS/8f707920-fe26-11ef-b737-06b888091933"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStacks({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataPersonNestedStackPersonNestedSt-72810Y0AHA1T/8f510a40-fe26-11ef-978a-06d488f07943"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataAmplifyTableManagerNestedStackA-1JMJ1D9EH891L/60c880e0-fe26-11ef-ab14-0a8a0b41c98f"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0-amplifyDataConnectionStackNestedStackConne-L130VQDWLV66/b460a660-fe26-11ef-a711-06b25169542f"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ | 5/5 | 11:26:22 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                EventId: '9ad3bf50-fe28-11ef-ad54-0603cfec563b',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                LogicalResourceId: 'amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:22.137'),
                ResourceStatus: 'UPDATE_COMPLETE',
            },
            progress: {
                total: 5,
                completed: 5,
                formatted: '5/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 4/5 | 11:26:30 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | data.NestedStack/data.NestedStackResource (data7552DF31)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'data7552DF31-UPDATE_COMPLETE-2025-03-11T03:26:30.163Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'data7552DF31',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-data7552DF31-1BIXA5GN4FTBJ/5d3a02f0-fe26-11ef-9db3-0a546862b135',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:30.163'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/76f3f913f2bffd4d3b833199e6d331903453796a752c920c60d3aa8456e8a975.json","Parameters":{"referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthIdentityPool5160F57DRef":"us-west-2:6568ef94-3300-4149-bc98-f2266342f0bb","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthUserPool0A15674BRef":"us-west-2_j6BNFhRJd","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthunauthenticatedUserRoleC92E0CFERef":"amplify-sampleprojectapp--amplifyAuthunauthenticate-Rieux6W03N47","referencetoamplifysampleprojectappusersandbox0e1752a889authNestedStackauthNestedStackResource1C2A2D1EOutputsamplifysampleprojectappusersandbox0e1752a889authamplifyAuthauthenticatedUserRole11751600Ref":"amplify-sampleprojectapp--amplifyAuthauthenticatedU-y26Xh6dN3Ku9"},"Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec35dd3144-7c45-43b2-bbc0-5c85440a20c6',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'data7552DF31',
                },
                constructPath: 'data.NestedStack/data.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 4,
                formatted: '4/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"amplify-sampleprojectapp-user-sandbox-0e1752a889"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-MyCustomResourcesBB610599-O9MPSU7CLLBR/48a46380-fe26-11ef-9a27-02b76afe5d0d"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I0000',
        message: '[sdk info] CloudFormation.DescribeStackEvents({"StackName":"arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847"}) -> OK',
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 3/5 | 11:26:30 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | auth.NestedStack/auth.NestedStackResource (auth179371D7)',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: 'auth179371D7-UPDATE_COMPLETE-2025-03-11T03:26:30.602Z',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'auth179371D7',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889-auth179371D7-1N176OGP9R2MV/489d1080-fe26-11ef-b1c7-0a7dea314847',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:30.602'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ResourceProperties: '{"TemplateURL":"https://s3.us-west-2.amazonaws.com/cdk-hnb659fds-assets-123456789012-us-west-2/53b8945b363f121c80e07e812f42fa6e33a8f3fdbc9b75d4b3374959bb62f47d.json","Tags":[{"Value":"sandbox","Key":"amplify:deployment-type"},{"Value":"amplify","Key":"created-by"}]}',
                ClientRequestToken: 'exec35dd3144-7c45-43b2-bbc0-5c85440a20c6',
            },
            metadata: {
                entry: {
                    type: 'aws:cdk:logicalId',
                    data: 'auth179371D7',
                },
                constructPath: 'auth.NestedStack/auth.NestedStackResource',
            },
            progress: {
                total: 5,
                completed: 3,
                formatted: '3/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5502',
        message: 'amplify-sampleprojectapp-user-sandbox-0e1752a889 | 4/5 | 11:26:30 PM | UPDATE_COMPLETE | AWS::CloudFormation::Stack | amplify-sampleprojectapp-user-sandbox-0e1752a889',
        data: {
            deployment: '66d61396-9cac-4151-b6b5-68a30e5537d3',
            event: {
                StackId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                EventId: '9ffb2cc0-fe28-11ef-9713-02a01569a639',
                StackName: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                LogicalResourceId: 'amplify-sampleprojectapp-user-sandbox-0e1752a889',
                PhysicalResourceId: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
                ResourceType: 'AWS::CloudFormation::Stack',
                Timestamp: new Date('2025-03-11T03:26:30.775'),
                ResourceStatus: 'UPDATE_COMPLETE',
                ClientRequestToken: 'exec35dd3144-7c45-43b2-bbc0-5c85440a20c6',
            },
            progress: {
                total: 5,
                completed: 4,
                formatted: '4/5',
            },
        },
    },
    {
        code: 'CDK_TOOLKIT_I5503',
        message: 'Completed amplify-sampleprojectapp-user-sandbox-0e1752a889',
    },
    {
        code: 'CDK_TOOLKIT_I5900',
        message: '  amplify-sampleprojectapp-user-sandbox-0e1752a889',
        data: {
            type: 'did-deploy-stack',
            noOp: false,
            outputs: {
                webClientId: '1f198iuo1sqf1g76kl9pplu4um',
                socialProviders: '',
                usernameAttributes: '["email"]',
                signupAttributes: '["email"]',
                oauthClientId: '1f198iuo1sqf1g76kl9pplu4um',
                allowUnauthenticatedIdentities: 'true',
                oauthCognitoDomain: '',
                oauthScope: '["profile","phone","email","openid","aws.cognito.signin.user.admin"]',
                oauthRedirectSignOut: '',
                awsAppsyncAdditionalAuthenticationTypes: 'AMAZON_COGNITO_USER_POOLS',
                mfaTypes: '[]',
                mfaConfiguration: 'OFF',
                oauthResponseType: 'code',
                awsAppsyncAuthenticationType: 'AWS_IAM',
                customOutputs: '{"custom":{"some":"keys"},"version":"1.3"}',
                oauthRedirectSignIn: 'https://example.com',
                passwordPolicyMinLength: '8',
                awsAppsyncApiEndpoint: 'https://ystl6ikbafavph56jdnbbbklmu.appsync-api.us-west-2.amazonaws.com/graphql',
                awsAppsyncApiId: 'viu3cvmcnvdx7dufpcbaimflau',
                authRegion: 'us-west-2',
                amplifyApiModelSchemaS3Uri: 's3://amplify-sampleprojectapp--amplifydataamplifycodege-v1bzay4pc2uy/model-schema.graphql',
                groups: '[]',
                awsAppsyncRegion: 'us-west-2',
                deploymentType: 'sandbox',
                passwordPolicyRequirements: '["REQUIRES_NUMBERS","REQUIRES_LOWERCASE","REQUIRES_UPPERCASE","REQUIRES_SYMBOLS"]',
                region: 'us-west-2',
                userPoolId: 'us-west-2_j6BNFhRJd',
                identityPoolId: 'us-west-2:6568ef94-3300-4149-bc98-f2266342f0bb',
                verificationMechanisms: '["email"]',
            },
            stackArn: 'arn:aws:cloudformation:us-west-2:123456789012:stack/amplify-sampleprojectapp-user-sandbox-0e1752a889/4300c130-fe26-11ef-b04b-0624c0eedabf',
        },
    },
    {
        code: 'CDK_TOOLKIT_I5000',
        message: '  Deployment time: 55.45s',
        data: {
            duration: 55447,
        },
    },
];
