// @aws-amplify/backend-cli/lib/commands/generate/generate_command.js
import { AmplifyUserError } from '@aws-amplify/platform-core';
/**
 * An entry point for generate command.
 */
export class GenerateCommand {
    generateOutputsCommand;
    generateFormsCommand;
    generateGraphqlClientCodeCommand;
    generateSchemaCommand;
    commandMiddleware;
    /**
     * @inheritDoc
     */
    command;
    /**
     * @inheritDoc
     */
    describe;
    /**
     * Creates top level entry point for generate command.
     */
    constructor(generateOutputsCommand, generateFormsCommand, generateGraphqlClientCodeCommand, generateSchemaCommand, commandMiddleware) {
        this.generateOutputsCommand = generateOutputsCommand;
        this.generateFormsCommand = generateFormsCommand;
        this.generateGraphqlClientCodeCommand = generateGraphqlClientCodeCommand;
        this.generateSchemaCommand = generateSchemaCommand;
        this.commandMiddleware = commandMiddleware;
        this.command = 'generate';
        this.describe = 'Generates post deployment artifacts';
    }
    /**
     * @inheritDoc
     */
    handler = () => {
        // CommandModule requires handler implementation. But this is never called if top level command
        // is configured to require subcommand.
        // Help is printed by default in that case before ever attempting to call handler.
        throw new Error('Top level generate handler should never be called');
    };
    builder = (yargs) => {
        return (yargs
            .version(false)
            // Cast to erase options types used in internal sub command implementation. Otherwise, compiler fails here.
            .command(this.generateOutputsCommand)
            .command(this.generateFormsCommand)
            .command(this.generateGraphqlClientCodeCommand)
            .command(this.generateSchemaCommand)
            .demandCommand()
            .strictCommands()
            .recommendCommands()
            .option('profile', {
            describe: 'An AWS profile name.',
            type: 'string',
            array: false,
        })
            .check(async (argv) => {
            const stackNameRegex = /^[a-zA-Z][-a-zA-Z0-9]{1,127}$|^arn:[-a-zA-Z0-9:/._+]$/;
            if (argv['stack'] && typeof argv['stack'] === 'string') {
                if (!argv.stack.match(stackNameRegex)) {
                    throw new AmplifyUserError('InvalidCommandInputError', {
                        message: `Invalid --stack name provided: ${argv.stack}`,
                        resolution: 'Check the value of the stack name provided and try again.',
                    });
                }
            }
            return true;
        })
            .middleware([this.commandMiddleware.ensureAwsCredentialAndRegion]));
    };
}
