// @aws-amplify/codegen-ui-react/dist/lib/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.scriptingPatterns = exports.AMPLIFY_JS_V6 = exports.AMPLIFY_JS_V5 = exports.STORAGE_FILE_ALGO_TYPE = exports.STORAGE_FILE_KEY = exports.COMPOSITE_PRIMARY_KEY_PROP_NAME = void 0;
/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
exports.COMPOSITE_PRIMARY_KEY_PROP_NAME = 'id';
exports.STORAGE_FILE_KEY = 'key';
exports.STORAGE_FILE_ALGO_TYPE = 'SHA-1';
exports.AMPLIFY_JS_V5 = '5.0.0';
exports.AMPLIFY_JS_V6 = '6.0.0';
exports.scriptingPatterns = [
    // JavaScript functions
    /eval\s*\(/i,
    /Function\s*\(/i,
    /setTimeout\s*\(/i,
    /setInterval\s*\(/i,
    /new\s+Function/i,
    /import\s*\(/i,
    /require\s*\(/i,
    // DOM manipulation
    /document\./i,
    /window\./i,
    /location\./i,
    // Event handlers
    /on\w+\s*=/i,
    // Dangerous protocols
    /javascript:/i,
    /data:/i,
    /vbscript:/i,
    // Script tags and variations
    /<script/i,
    /<\/script/i,
    /<x:script/i,
    // SVG exploits
    /<svg/i,
    /xlink:href/i,
    // Object prototype attacks
    /\[\s*Symbol\s*\./i,
    /__proto__/i,
    /prototype\s*\./i,
    // Template literal attacks
    /\$\{/i,
    // Base64 indicators
    /base64/i,
];
