// @aws-amplify/backend-cli/lib/commands/notices/notices_acknowledge_command.js
import { printer } from '@aws-amplify/cli-core';
/**
 * Notices acknowledge command.
 */
export class NoticesAcknowledgeCommand {
    noticesController;
    /**
     * @inheritDoc
     */
    command;
    /**
     * @inheritDoc
     */
    describe;
    /**
     * Creates notices command
     */
    constructor(noticesController) {
        this.noticesController = noticesController;
        this.command = 'acknowledge <notice-id>';
        this.describe =
            'Mark a specific notice as acknowledged. ' +
                "Use this command with a notice ID to indicate you've reviewed and understood the notification.";
    }
    /**
     * @inheritDoc
     */
    handler = async (args) => {
        const noticeId = args['notice-id'];
        await this.noticesController.acknowledge(noticeId);
        printer.print(`Notice ${noticeId} has been acknowledged.`);
    };
    /**
     * @inheritDoc
     */
    builder = (yargs) => {
        return yargs
            .version(false)
            .positional('notice-id', {
            describe: 'Id of the notice',
            type: 'string',
            demandOption: true,
        })
            .help();
    };
}
