// @aws-amplify/backend-cli/lib/commands/notices/notices_list_command.js
/**
 * Notices list command.
 */
export class NoticesListCommand {
    noticesController;
    noticesPrinter;
    /**
     * @inheritDoc
     */
    command;
    /**
     * @inheritDoc
     */
    describe;
    /**
     * Creates notices list command
     */
    constructor(noticesController, noticesPrinter) {
        this.noticesController = noticesController;
        this.noticesPrinter = noticesPrinter;
        this.command = 'list';
        this.describe =
            'Displays active notices relevant to your Amplify backend environment. ' +
                'Shows important information about package compatibility, version updates, ' +
                'and potential issues that may affect your development workflow.';
    }
    /**
     * @inheritDoc
     */
    handler = async (args) => {
        const notices = await this.noticesController.getApplicableNotices({
            event: 'listNoticesCommand',
            includeAcknowledged: args.all,
        });
        this.noticesPrinter.print(notices);
    };
    /**
     * @inheritDoc
     */
    builder = (yargs) => {
        return yargs
            .version(false)
            .option('all', {
            type: 'boolean',
            describe: 'Includes already acknowledged notices',
        })
            .help();
    };
}
