// @aws-amplify/graphql-api-construct/lib/internal/metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMetadataCustomOperations = exports.getMetadataAuthorizationModes = exports.getMetadataDataSources = void 0;
const getMetadataDataSources = (definition) => {
    const dataSourceDbTypes = Object.values(definition.dataSourceStrategies).map((strategy) => strategy.dbType.toLocaleLowerCase());
    const customSqlDbTypes = (definition.customSqlDataSourceStrategies ?? []).map((strategy) => strategy.strategy.dbType.toLocaleLowerCase());
    const dataSources = [...new Set([...dataSourceDbTypes, ...customSqlDbTypes])].sort();
    return dataSources.join(',');
};
exports.getMetadataDataSources = getMetadataDataSources;
const getMetadataAuthorizationModes = (authorizationModes) => {
    const configKeyToAuthMode = {
        iamConfig: 'aws_iam',
        oidcConfig: 'openid_connect',
        identityPoolConfig: 'amazon_cognito_identity_pools',
        userPoolConfig: 'amazon_cognito_user_pools',
        apiKeyConfig: 'api_key',
        lambdaConfig: 'aws_lambda',
    };
    const authModes = Object.keys(authorizationModes)
        .map((mode) => configKeyToAuthMode[mode])
        // remove values not found in mapping
        .filter((mode) => !!mode)
        .sort();
    return authModes.join(',');
};
exports.getMetadataAuthorizationModes = getMetadataAuthorizationModes;
const getMetadataCustomOperations = (definition) => {
    const customOperations = [];
    if (definition.schema.includes('type Query')) {
        customOperations.push('queries');
    }
    if (definition.schema.includes('type Mutation')) {
        customOperations.push('mutations');
    }
    if (definition.schema.includes('type Subscription')) {
        customOperations.push('subscriptions');
    }
    return customOperations.join(',');
};
exports.getMetadataCustomOperations = getMetadataCustomOperations;
