// @aws-cdk/toolkit-lib/lib/api/io/private/io-default-messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IoDefaultMessages = void 0;
const util = require("util");
/**
 * Helper class to emit standard log messages to an IoHost
 *
 * It wraps an `IoHelper`, and adds convenience methods to emit default messages
 * for the various log levels.
 */
class IoDefaultMessages {
    ioHelper;
    constructor(ioHelper) {
        this.ioHelper = ioHelper;
    }
    async notify(msg) {
        return this.ioHelper.notify({
            ...msg,
            code: undefined,
        });
    }
    async error(input, ...args) {
        return this.emitMessage('error', input, ...args);
    }
    async warn(input, ...args) {
        return this.emitMessage('warn', input, ...args);
    }
    async warning(input, ...args) {
        return this.emitMessage('warn', input, ...args);
    }
    async info(input, ...args) {
        return this.emitMessage('info', input, ...args);
    }
    async debug(input, ...args) {
        return this.emitMessage('debug', input, ...args);
    }
    async trace(input, ...args) {
        return this.emitMessage('trace', input, ...args);
    }
    async result(input, ...args) {
        return this.emitMessage('result', input, ...args);
    }
    /**
     * Makes a default message object from a level and a message
     */
    msg(level, input, ...args) {
        // Format message if args are provided
        const message = args.length > 0 ? util.format(input, ...args) : input;
        return {
            time: new Date(),
            level,
            message,
            data: undefined,
        };
    }
    async emitMessage(level, input, ...args) {
        return this.ioHelper.notify(this.msg(level, input, ...args));
    }
}
exports.IoDefaultMessages = IoDefaultMessages;
