// @aws-cdk/toolkit-lib/lib/api/io/private/io-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IoHelper = void 0;
exports.asIoHelper = asIoHelper;
const io_default_messages_1 = require("./io-default-messages");
const span_1 = require("./span");
/**
 * A class containing helper tools to interact with IoHost
 */
class IoHelper {
    static fromIoHost(ioHost, action) {
        return new IoHelper({
            notify: (msg) => ioHost.notify({
                ...msg,
                action: action,
            }),
            requestResponse: (msg) => ioHost.requestResponse({
                ...msg,
                action: action,
            }),
        });
    }
    static fromActionAwareIoHost(ioHost) {
        return new IoHelper(ioHost);
    }
    /**
     * Simplified access to emit default messages.
     */
    defaults;
    ioHost;
    constructor(ioHost) {
        this.ioHost = ioHost;
        this.defaults = new io_default_messages_1.IoDefaultMessages(this);
    }
    /**
     * Forward a message to the IoHost, while injection the current action
     */
    notify(msg) {
        return this.ioHost.notify(msg);
    }
    /**
     * Forward a request to the IoHost, while injection the current action
     */
    requestResponse(msg) {
        return this.ioHost.requestResponse(msg);
    }
    /**
     * Create a new marker from a given registry entry
     */
    span(definition) {
        return new span_1.SpanMaker(this, definition, (ioHost) => IoHelper.fromActionAwareIoHost(ioHost));
    }
}
exports.IoHelper = IoHelper;
/**
 * Wraps an IoHost and creates an IoHelper from it
 */
function asIoHelper(ioHost, action) {
    return IoHelper.fromIoHost(ioHost, action);
}
