// @aws-cdk/toolkit-lib/lib/api/io/private/io-host-wrappers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withoutColor = withoutColor;
exports.withoutEmojis = withoutEmojis;
exports.withTrimmedWhitespace = withTrimmedWhitespace;
/**
 * An IoHost wrapper that strips out ANSI colors and styles from the message before
 * sending the message to the given IoHost
 */
function withoutColor(ioHost) {
    return {
        notify: async (msg) => {
            await ioHost.notify({
                ...msg,
                message: stripColor(msg.message),
            });
        },
        requestResponse: async (msg) => {
            return ioHost.requestResponse({
                ...msg,
                message: stripColor(msg.message),
            });
        },
    };
}
function stripColor(msg) {
    return msg.replace(/[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g, '');
}
/**
 * An IoHost wrapper that strips out emojis from the message before
 * sending the message to the given IoHost
 */
function withoutEmojis(ioHost) {
    return {
        notify: async (msg) => {
            await ioHost.notify({
                ...msg,
                message: stripEmojis(msg.message),
            });
        },
        requestResponse: async (msg) => {
            return ioHost.requestResponse({
                ...msg,
                message: stripEmojis(msg.message),
            });
        },
    };
}
function stripEmojis(msg) {
    // https://www.unicode.org/reports/tr51/#def_emoji_presentation
    return msg.replace(/\p{Emoji_Presentation}/gu, '');
}
/**
 * An IoHost wrapper that trims whitespace at the beginning and end of messages.
 * This is required, since after removing emojis and ANSI colors,
 * we might end up with floating whitespace at either end.
 */
function withTrimmedWhitespace(ioHost) {
    return {
        notify: async (msg) => {
            await ioHost.notify({
                ...msg,
                message: msg.message.trim(),
            });
        },
        requestResponse: async (msg) => {
            return ioHost.requestResponse({
                ...msg,
                message: msg.message.trim(),
            });
        },
    };
}
