// @aws-amplify/backend-output-schemas/lib/storage/v1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.storageOutputSchema = void 0;
const zod_1 = require("zod");
const storageAccessActionEnum = zod_1.z.enum([
    'read',
    'get',
    'list',
    'write',
    'delete',
]);
const pathSchema = zod_1.z.record(zod_1.z.string(), zod_1.z.intersection(zod_1.z.object({
    guest: zod_1.z.array(storageAccessActionEnum),
    authenticated: zod_1.z.array(storageAccessActionEnum),
    resource: zod_1.z.array(storageAccessActionEnum),
}), zod_1.z.record(zod_1.z.string(), zod_1.z.array(storageAccessActionEnum))));
const bucketSchema = zod_1.z.object({
    name: zod_1.z.string(),
    bucketName: zod_1.z.string(),
    storageRegion: zod_1.z.string(),
    paths: pathSchema.optional(),
});
exports.storageOutputSchema = zod_1.z.object({
    version: zod_1.z.literal('1'),
    payload: zod_1.z.object({
        bucketName: zod_1.z.string(),
        storageRegion: zod_1.z.string(),
        buckets: zod_1.z.string(zod_1.z.array(bucketSchema)).optional(), // JSON serialized array of bucketSchema
    }),
});
