// @aws-cdk/cloudformation-diff/lib/iam/iam-identity-center.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsoInstanceACAConfig = exports.SsoAssignment = exports.SsoPermissionSet = void 0;
// namespace object imports won't work in the bundle for function exports
// eslint-disable-next-line @typescript-eslint/no-require-imports
const deepEqual = require('fast-deep-equal');
class SsoPermissionSet {
    constructor(props) {
        this.cfnLogicalId = props.cfnLogicalId;
        this.name = props.name;
        this.ssoInstanceArn = props.ssoInstanceArn;
        this.ssoPermissionsBoundary = props.ssoPermissionsBoundary;
        this.ssoCustomerManagedPolicyReferences = props.ssoCustomerManagedPolicyReferences;
    }
    equal(other) {
        return deepEqual(this, other);
    }
}
exports.SsoPermissionSet = SsoPermissionSet;
class SsoAssignment {
    constructor(props) {
        this.cfnLogicalId = props.cfnLogicalId;
        this.ssoInstanceArn = props.ssoInstanceArn;
        this.permissionSetArn = props.permissionSetArn;
        this.principalId = props.principalId;
        this.principalType = props.principalType;
        this.targetId = props.targetId;
        this.targetType = props.targetType;
    }
    equal(other) {
        return deepEqual(this, other);
    }
}
exports.SsoAssignment = SsoAssignment;
class SsoInstanceACAConfig {
    constructor(props) {
        this.cfnLogicalId = props.cfnLogicalId;
        this.ssoInstanceArn = props.ssoInstanceArn;
        this.accessControlAttributes = props.accessControlAttributes;
    }
    equal(other) {
        return deepEqual(this, other);
    }
}
exports.SsoInstanceACAConfig = SsoInstanceACAConfig;
