// @aws-cdk/cx-api/lib/private/flag-modeling.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAGIC_V2NEXT = exports.FlagType = void 0;
exports.compareVersions = compareVersions;
var FlagType;
(function (FlagType) {
    /**
     * Change the default behavior of the API
     *
     * The old behavior is still valid, and possible to achieve with source
     * code changes, but we recommend the new behavior instead.
     *
     * Also valid for changes that don't affect CloudFormation, but the CXAPI
     * contract.
     */
    FlagType[FlagType["ApiDefault"] = 0] = "ApiDefault";
    /**
     * Address a bug in a way that requires contract breaking or has availability implications for existing infrastructure
     *
     * The old behavior is not recommended, and shouldn't have been possible in the first place.
     * We only have this flag because we can't roll out the fix to everyone
     * automatically for fear of breakage.
     */
    FlagType[FlagType["BugFix"] = 1] = "BugFix";
    /**
     * Advertise the presence of this context option in `cdk.json`
     */
    FlagType[FlagType["VisibleContext"] = 2] = "VisibleContext";
    /**
     * Use this type for flags that are to be removed on a set date
     */
    FlagType[FlagType["Temporary"] = 3] = "Temporary";
})(FlagType || (exports.FlagType = FlagType = {}));
/**
 * The magic value that will be substituted at version bump time with the actual
 * new V2 version.
 *
 * Do not import this constant in the `features.ts` file, or the substitution
 * process won't work.
 */
exports.MAGIC_V2NEXT = 'V2NEXT';
/**
 * Compare two versions, returning -1, 0, or 1.
 */
function compareVersions(a, b) {
    if (a === b) {
        return 0;
    }
    if (a === undefined) {
        return -1;
    }
    if (b === undefined) {
        return 1;
    }
    const as = a.split('.').map(x => parseInt(x, 10));
    const bs = b.split('.').map(x => parseInt(x, 10));
    if (a === exports.MAGIC_V2NEXT) {
        return bs[0] <= 2 ? 1 : -1;
    }
    if (b === exports.MAGIC_V2NEXT) {
        return as[0] <= 2 ? -1 : 1;
    }
    for (let i = 0; i < Math.min(as.length, bs.length); i++) {
        if (as[i] < bs[i]) {
            return -1;
        }
        if (as[i] > bs[i]) {
            return 1;
        }
    }
    return as.length - bs.length;
}
