// cdk-assets/lib/publishing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetPublishing = void 0;
const docker_1 = require("./private/docker");
const handlers_1 = require("./private/handlers");
const p_limit_1 = require("./private/p-limit");
const progress_1 = require("./progress");
class AssetPublishing {
    constructor(manifest, options) {
        this.manifest = manifest;
        this.options = options;
        /**
         * The message for the IPublishProgress interface
         */
        this.message = 'Starting';
        this.failures = new Array();
        this.completedOperations = 0;
        this.aborted = false;
        this.handlerCache = new Map();
        this.assets = manifest.entries;
        this.totalOperations = this.assets.length;
        this.publishInParallel = options.publishInParallel ?? false;
        this.buildAssets = options.buildAssets ?? true;
        this.publishAssets = options.publishAssets ?? true;
        const self = this;
        this.handlerHost = {
            aws: this.options.aws,
            get aborted() {
                return self.aborted;
            },
            emitMessage(t, m) {
                self.progressEvent(t, m);
            },
            dockerFactory: new docker_1.DockerFactory(),
        };
    }
    /**
     * Publish all assets from the manifest
     */
    async publish(options = {}) {
        if (this.publishInParallel) {
            const limit = (0, p_limit_1.pLimit)(20);
            // eslint-disable-next-line @cdklabs/promiseall-no-unbounded-parallelism
            await Promise.all(this.assets.map((asset) => limit(async () => this.publishAsset(asset, options))));
        }
        else {
            for (const asset of this.assets) {
                if (!(await this.publishAsset(asset, options))) {
                    break;
                }
            }
        }
        if ((this.options.throwOnError ?? true) && this.failures.length > 0) {
            throw new Error(`Error publishing: ${this.failures.map((e) => e.error.message)}`);
        }
    }
    /**
     * Build a single asset from the manifest
     */
    async buildEntry(asset) {
        try {
            if (this.progressEvent(progress_1.EventType.START, `Building ${asset.displayName(false)}`)) {
                return false;
            }
            const handler = this.assetHandler(asset);
            await handler.build();
            if (this.aborted) {
                throw new Error('Aborted');
            }
            this.completedOperations++;
            if (this.progressEvent(progress_1.EventType.SUCCESS, `Built ${asset.displayName(false)}`)) {
                return false;
            }
        }
        catch (e) {
            this.failures.push({ asset, error: e });
            this.completedOperations++;
            if (this.progressEvent(progress_1.EventType.FAIL, e.message)) {
                return false;
            }
        }
        return true;
    }
    /**
     * Publish a single asset from the manifest
     */
    async publishEntry(asset, options = {}) {
        try {
            if (this.progressEvent(progress_1.EventType.START, `Publishing ${asset.displayName(true)}`)) {
                return false;
            }
            const handler = this.assetHandler(asset);
            await handler.publish(options);
            if (this.aborted) {
                throw new Error('Aborted');
            }
            this.completedOperations++;
            if (this.progressEvent(progress_1.EventType.SUCCESS, `Published ${asset.displayName(true)}`)) {
                return false;
            }
        }
        catch (e) {
            this.failures.push({ asset, error: e });
            this.completedOperations++;
            if (this.progressEvent(progress_1.EventType.FAIL, e.message)) {
                return false;
            }
        }
        return true;
    }
    /**
     * Return whether a single asset is published
     */
    isEntryPublished(asset) {
        const handler = this.assetHandler(asset);
        return handler.isPublished();
    }
    /**
     * publish an asset (used by 'publish()')
     * @param asset - The asset to publish
     * @returns false when publishing should stop
     */
    async publishAsset(asset, options = {}) {
        try {
            if (this.progressEvent(progress_1.EventType.START, `Publishing ${asset.displayName(true)}`)) {
                return false;
            }
            const handler = this.assetHandler(asset);
            if (this.buildAssets) {
                await handler.build();
            }
            if (this.publishAssets) {
                await handler.publish(options);
            }
            if (this.aborted) {
                throw new Error('Aborted');
            }
            this.completedOperations++;
            if (this.progressEvent(progress_1.EventType.SUCCESS, `Published ${asset.displayName(true)}`)) {
                return false;
            }
        }
        catch (e) {
            this.failures.push({ asset, error: e });
            this.completedOperations++;
            if (this.progressEvent(progress_1.EventType.FAIL, e.message)) {
                return false;
            }
        }
        return true;
    }
    get percentComplete() {
        if (this.totalOperations === 0) {
            return 100;
        }
        return Math.floor((this.completedOperations / this.totalOperations) * 100);
    }
    abort() {
        this.aborted = true;
    }
    get hasFailures() {
        return this.failures.length > 0;
    }
    /**
     * Publish a progress event to the listener, if present.
     *
     * Returns whether an abort is requested. Helper to get rid of repetitive code in publish().
     */
    progressEvent(event, message) {
        this.message = message;
        if (this.options.progressListener) {
            this.options.progressListener.onPublishEvent(event, this);
        }
        return this.aborted;
    }
    assetHandler(asset) {
        const existing = this.handlerCache.get(asset);
        if (existing) {
            return existing;
        }
        if (this.options.quiet !== undefined && this.options.subprocessOutputDestination) {
            throw new Error('Cannot set both quiet and subprocessOutputDestination. Please use only subprocessOutputDestination');
        }
        const subprocessOutputDestination = this.options.subprocessOutputDestination ?? (this.options.quiet ? 'ignore' : 'stdio');
        const ret = (0, handlers_1.makeAssetHandler)(this.manifest, asset, this.handlerHost, {
            subprocessOutputDestination,
        });
        this.handlerCache.set(asset, ret);
        return ret;
    }
}
exports.AssetPublishing = AssetPublishing;
