// @aws-cdk/toolkit-lib/lib/api/refactoring/cloudformation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceMapping = exports.ResourceLocation = void 0;
/**
 * This class mirrors the `ResourceLocation` interface from CloudFormation,
 * but is richer, since it has a reference to the stack object, rather than
 * merely the stack name.
 */
class ResourceLocation {
    stack;
    logicalResourceId;
    constructor(stack, logicalResourceId) {
        this.stack = stack;
        this.logicalResourceId = logicalResourceId;
    }
    toPath() {
        const stack = this.stack;
        const resource = stack.template.Resources?.[this.logicalResourceId];
        const result = resource?.Metadata?.['aws:cdk:path'];
        if (result != null) {
            return result;
        }
        // If the path is not available, we can use stack name and logical ID
        return `${stack.stackName}.${this.logicalResourceId}`;
    }
    getType() {
        const resource = this.stack.template.Resources?.[this.logicalResourceId ?? ''];
        return resource?.Type ?? 'Unknown';
    }
    equalTo(other) {
        return this.logicalResourceId === other.logicalResourceId && this.stack.stackName === other.stack.stackName;
    }
}
exports.ResourceLocation = ResourceLocation;
/**
 * A mapping between a source and a destination location.
 */
class ResourceMapping {
    source;
    destination;
    constructor(source, destination) {
        this.source = source;
        this.destination = destination;
    }
    toTypedMapping() {
        return {
            // the type is the same in both source and destination,
            // so we can use either one
            type: this.source.getType(),
            sourcePath: this.source.toPath(),
            destinationPath: this.destination.toPath(),
        };
    }
}
exports.ResourceMapping = ResourceMapping;
