// @aws-amplify/backend-function/lib/function_construct_base.js
import { Construct } from 'constructs';
import { Stack } from 'aws-cdk-lib';
import { functionOutputKey, } from '@aws-amplify/backend-output-schemas';
import { AttributionMetadataStorage } from '@aws-amplify/backend-output-storage';
import { fileURLToPath } from 'node:url';
const functionStackType = 'function-Lambda';
/**
 * A base class for function constructs.
 */
export class AmplifyFunctionBase extends Construct {
    outputStorageStrategy;
    stack;
    /**
     * Creates base function construct.
     */
    constructor(scope, id, outputStorageStrategy) {
        super(scope, id);
        this.outputStorageStrategy = outputStorageStrategy;
        this.stack = Stack.of(scope);
        new AttributionMetadataStorage().storeAttributionMetadata(Stack.of(this), functionStackType, fileURLToPath(new URL('../package.json', import.meta.url)));
    }
    storeOutput = () => {
        this.outputStorageStrategy.appendToBackendOutputList(functionOutputKey, {
            version: '1',
            payload: {
                definedFunctions: this.resources.lambda.functionName,
            },
        });
    };
}
