// @aws-cdk/toolkit-lib/lib/api/stack-events/stack-progress-monitor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StackProgressMonitor = void 0;
const util = require("util");
const util_1 = require("../../util");
/**
 * Monitors stack progress.s
 */
class StackProgressMonitor {
    /**
     * Previous completion state observed by logical ID
     *
     * We use this to detect that if we see a DELETE_COMPLETE after a
     * CREATE_COMPLETE, it's actually a rollback and we should DECREASE
     * resourcesDone instead of increase it
     */
    resourcesPrevCompleteState = {};
    /**
     * Count of resources that have reported a _COMPLETE status
     */
    resourcesDone = 0;
    /**
     * How many digits we need to represent the total count (for lining up the status reporting)
     */
    resourceDigits = 0;
    /**
     * Number of expected resources in the monitor.
     */
    resourcesTotal;
    constructor(resourcesTotal) {
        // +1 because the stack also emits a "COMPLETE" event at the end, and that wasn't
        // counted yet. This makes it line up with the amount of events we expect.
        this.resourcesTotal = resourcesTotal ? resourcesTotal + 1 : undefined;
        // How many digits does this number take to represent?
        this.resourceDigits = this.resourcesTotal ? Math.ceil(Math.log10(this.resourcesTotal)) : 0;
    }
    /**
     * Report the stack progress
     */
    get progress() {
        return {
            total: this.total,
            completed: this.completed,
            formatted: this.formatted,
        };
    }
    /**
     * The total number of progress monitored resources.
     */
    get total() {
        return this.resourcesTotal;
    }
    /**
     * The number of completed resources.
     */
    get completed() {
        return this.resourcesDone;
    }
    /**
     * Report the current progress as a [34/42] string, or just [34] if the total is unknown
     */
    get formatted() {
        if (this.resourcesTotal == null) {
            // Don't have total, show simple count and hope the human knows
            return (0, util_1.padLeft)(3, util.format('%s', this.resourcesDone)); // max 500 resources
        }
        return util.format('%s/%s', (0, util_1.padLeft)(this.resourceDigits, this.resourcesDone.toString()), (0, util_1.padLeft)(this.resourceDigits, this.resourcesTotal.toString()));
    }
    /**
     * Process as stack event and update the progress state.
     */
    process(event) {
        const status = event.ResourceStatus;
        if (!status || !event.LogicalResourceId) {
            return;
        }
        if (status.endsWith('_COMPLETE_CLEANUP_IN_PROGRESS')) {
            this.resourcesDone++;
        }
        if (status.endsWith('_COMPLETE')) {
            const prevState = this.resourcesPrevCompleteState[event.LogicalResourceId];
            if (!prevState) {
                this.resourcesDone++;
            }
            else {
                // If we completed this before and we're completing it AGAIN, means we're rolling back.
                // Protect against silly underflow.
                this.resourcesDone--;
                if (this.resourcesDone < 0) {
                    this.resourcesDone = 0;
                }
            }
            this.resourcesPrevCompleteState[event.LogicalResourceId] = status;
        }
    }
}
exports.StackProgressMonitor = StackProgressMonitor;
