// @aws-amplify/backend/lib/engine/backend-secret/backend_secret_fetcher_factory.js
import { CustomResource, Lazy } from 'aws-cdk-lib';
/**
 * Resource provider ID for the backend secret resource.
 */
export const SECRET_RESOURCE_PROVIDER_ID = 'AmplifySecretFetcherResourceProvider';
class SecretFetcherCustomResource extends CustomResource {
    secrets;
    constructor(scope, id, props, secrets) {
        super(scope, id, {
            ...props,
        });
        this.secrets = secrets;
    }
    addSecret = (secretName) => {
        this.secrets.add(secretName);
    };
}
/**
 * Type of the backend custom CFN resource.
 */
const SECRET_RESOURCE_TYPE = `Custom::AmplifySecretFetcherResource`;
/**
 * The factory to create backend secret-fetcher resource.
 */
export class BackendSecretFetcherFactory {
    secretProviderFactory;
    /**
     * Creates a backend secret-fetcher resource factory.
     */
    constructor(secretProviderFactory) {
        this.secretProviderFactory = secretProviderFactory;
    }
    /**
     * Returns a resource if it exists in the input scope. Otherwise,
     * creates a new one.
     */
    getOrCreate = (scope, secretName, backendIdentifier) => {
        const secretResourceId = `AmplifySecretFetcherResource`;
        const existingResource = scope.node.tryFindChild(secretResourceId);
        if (existingResource) {
            existingResource.addSecret(secretName);
            return existingResource;
        }
        const secrets = new Set();
        secrets.add(secretName);
        const provider = this.secretProviderFactory.getOrCreateInstance(scope, SECRET_RESOURCE_PROVIDER_ID, backendIdentifier);
        // Sandbox deployment passes down the secret's last updated timestamp to
        // trigger secret update. It is to optimize sandbox deployment time by
        // leveraging cdk hotswap.
        const secretLastUpdated = scope.node.tryGetContext('secretLastUpdated') ?? Date.now();
        const customResourceProps = {
            namespace: backendIdentifier.namespace,
            name: backendIdentifier.name,
            type: backendIdentifier.type,
            secretNames: Lazy.list({
                produce: () => {
                    return Array.from(secrets);
                },
            }),
        };
        return new SecretFetcherCustomResource(scope, secretResourceId, {
            serviceToken: provider.serviceToken,
            properties: {
                ...customResourceProps,
                secretLastUpdated, // this property is only to trigger resource update event.
            },
            resourceType: SECRET_RESOURCE_TYPE,
        }, secrets);
    };
}
