// @aws-amplify/graphql-conversation-transformer/lib/graphql-types/conversation-model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createConversationModel = void 0;
const graphql_directives_1 = require("@aws-amplify/graphql-directives");
const graphql_1 = require("graphql");
const graphql_transformer_common_1 = require("graphql-transformer-common");
const createConversationModel = (conversationName, messageName, referenceFieldName) => {
    const authDirective = createConversationAuthDirective();
    const modelDirective = createConversationModelDirective();
    const hasManyMessagesDirective = createConversationModelMessagesFieldHasManyDirective(referenceFieldName);
    const messagesField = createConversationModelMessagesField(hasManyMessagesDirective, messageName);
    const model = makeConversationModel(conversationName, messagesField, [authDirective, modelDirective]);
    return {
        authDirective,
        modelDirective,
        hasManyMessagesDirective,
        messagesField,
        model,
    };
};
exports.createConversationModel = createConversationModel;
const createConversationAuthDirective = () => {
    const rules = (0, graphql_transformer_common_1.makeArgument)('rules', {
        kind: graphql_1.Kind.LIST,
        values: [
            {
                kind: graphql_1.Kind.OBJECT,
                fields: [
                    {
                        kind: graphql_1.Kind.OBJECT_FIELD,
                        name: { kind: graphql_1.Kind.NAME, value: 'allow' },
                        value: { kind: graphql_1.Kind.ENUM, value: 'owner' },
                    },
                    {
                        kind: graphql_1.Kind.OBJECT_FIELD,
                        name: { kind: graphql_1.Kind.NAME, value: 'ownerField' },
                        value: { kind: graphql_1.Kind.STRING, value: 'owner' },
                    },
                ],
            },
        ],
    });
    return (0, graphql_transformer_common_1.makeDirective)('auth', [rules]);
};
const createConversationModelDirective = () => {
    const subscriptionsOffValue = {
        kind: graphql_1.Kind.OBJECT,
        fields: [
            {
                kind: graphql_1.Kind.OBJECT_FIELD,
                name: { kind: graphql_1.Kind.NAME, value: 'level' },
                value: { kind: graphql_1.Kind.ENUM, value: 'off' },
            },
        ],
    };
    return (0, graphql_transformer_common_1.makeDirective)('model', [(0, graphql_transformer_common_1.makeArgument)('subscriptions', subscriptionsOffValue)]);
};
const createConversationModelMessagesFieldHasManyDirective = (fieldName) => {
    const referencesArg = (0, graphql_transformer_common_1.makeArgument)('references', (0, graphql_transformer_common_1.makeValueNode)(fieldName));
    return (0, graphql_transformer_common_1.makeDirective)(graphql_directives_1.HasManyDirective.name, [referencesArg]);
};
const createConversationModelMessagesField = (hasManyDirective, typeName) => {
    return (0, graphql_transformer_common_1.makeField)('messages', [], (0, graphql_transformer_common_1.makeListType)((0, graphql_transformer_common_1.makeNamedType)(typeName)), [hasManyDirective]);
};
const makeConversationModel = (modelName, messagesField, typeLevelDirectives) => {
    const id = (0, graphql_transformer_common_1.makeField)('id', [], (0, graphql_transformer_common_1.wrapNonNull)((0, graphql_transformer_common_1.makeNamedType)('ID')));
    const name = (0, graphql_transformer_common_1.makeField)('name', [], (0, graphql_transformer_common_1.makeNamedType)('String'));
    const metadata = (0, graphql_transformer_common_1.makeField)('metadata', [], (0, graphql_transformer_common_1.makeNamedType)('AWSJSON'));
    return {
        ...(0, graphql_transformer_common_1.blankObject)(modelName),
        fields: [id, name, metadata, messagesField],
        directives: typeLevelDirectives,
    };
};
