// @aws-amplify/backend-output-schemas/lib/auth/v1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.authOutputSchema = void 0;
const zod_1 = require("zod");
// The values stored in the stack output is converted to string.
exports.authOutputSchema = zod_1.z.object({
    version: zod_1.z.literal('1'),
    payload: zod_1.z.object({
        authRegion: zod_1.z.string(),
        userPoolId: zod_1.z.string(),
        webClientId: zod_1.z.string(),
        identityPoolId: zod_1.z.string(),
        allowUnauthenticatedIdentities: zod_1.z.optional(zod_1.z.string()), // boolean as string 'true' | 'false'
        usernameAttributes: zod_1.z.string().optional(), // JSON array as string
        signupAttributes: zod_1.z.string().optional(), // JSON array as string
        passwordPolicyMinLength: zod_1.z.string().optional(),
        passwordPolicyRequirements: zod_1.z.string().optional(), // JSON array as string
        mfaConfiguration: zod_1.z.string().optional(),
        mfaTypes: zod_1.z.string().optional(), // JSON array as string
        verificationMechanisms: zod_1.z.string().optional(), // JSON array as string
        socialProviders: zod_1.z.string().optional(), // JSON array as string
        oauthCognitoDomain: zod_1.z.string().optional(),
        oauthScope: zod_1.z.string().optional(), // JSON array as string
        oauthRedirectSignIn: zod_1.z.string().optional(),
        oauthRedirectSignOut: zod_1.z.string().optional(),
        oauthClientId: zod_1.z.string().optional(),
        oauthResponseType: zod_1.z.string().optional(),
        groups: zod_1.z.string().optional(), // JSON array as string
    }),
});
