// @aws-amplify/backend-cli/lib/commands/sandbox/sandbox-seed/sandbox_seed_command.js
import path from 'path';
import { existsSync } from 'fs';
import { AmplifyError, AmplifyUserError } from '@aws-amplify/platform-core';
import { format, printer } from '@aws-amplify/cli-core';
import { tsImport } from 'tsx/esm/api';
import { pathToFileURL } from 'url';
/**
 * Command that runs seed in sandbox environment
 */
export class SandboxSeedCommand {
    backendIDResolver;
    seedSubCommands;
    /**
     * @inheritDoc
     */
    command;
    /**
     * @inheritDoc
     */
    describe;
    /**
     * Seeds sandbox environment.
     */
    constructor(backendIDResolver, seedSubCommands) {
        this.backendIDResolver = backendIDResolver;
        this.seedSubCommands = seedSubCommands;
        this.command = 'seed';
        this.describe = 'Seeds sandbox environment';
    }
    /**
     * @inheritDoc
     */
    handler = async (args) => {
        printer.print(`${format.color('seed is running...', 'Blue')}`);
        const backendID = await this.backendIDResolver.resolve(args.identifier);
        const seedPath = path.join('amplify', 'seed', 'seed.ts');
        process.env.AMPLIFY_BACKEND_IDENTIFIER = JSON.stringify(backendID);
        try {
            await tsImport(pathToFileURL(seedPath).toString(), import.meta.url);
        }
        catch (e) {
            const error = e;
            if (error.message.includes('Cannot find module') &&
                error.message.includes('seed')) {
                throw new AmplifyUserError('SeedScriptNotFoundError', {
                    message: `There is no file that corresponds to ${seedPath}`,
                    resolution: `Please make a file that corresponds to ${seedPath} and put your seed logic in it`,
                }, error);
            }
            else {
                if (AmplifyError.isAmplifyError(e)) {
                    throw e;
                }
                throw new AmplifyUserError('SeedingFailedError', {
                    message: 'Seed failed to complete',
                    resolution: 'Check the Caused by error and fix any issues in your seed script',
                }, e instanceof Error ? error : undefined);
            }
        }
        printer.printNewLine();
        printer.print(`${format.success('✔')} seed has successfully completed`);
    };
    /**
     * @inheritDoc
     */
    builder = (yargs) => {
        return yargs.command(this.seedSubCommands).check(() => {
            const seedPath = path.join(process.cwd(), 'amplify', 'seed', 'seed.ts');
            if (!existsSync(seedPath)) {
                throw new AmplifyUserError('SeedScriptNotFoundError', {
                    message: `There is no file that corresponds to ${seedPath}`,
                    resolution: `Please make a file that corresponds to ${seedPath} and put your seed logic in it`,
                });
            }
            return true;
        });
    };
}
