// @aws-cdk/toolkit-lib/lib/payloads/hotswap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NonHotswappableReason = void 0;
var NonHotswappableReason;
(function (NonHotswappableReason) {
    /**
     * Tags are not hotswappable
     */
    NonHotswappableReason["TAGS"] = "tags";
    /**
     * Changed resource properties are not hotswappable on this resource type
     */
    NonHotswappableReason["PROPERTIES"] = "properties";
    /**
     * A stack output has changed
     */
    NonHotswappableReason["OUTPUT"] = "output";
    /**
     * A dependant resource is not hotswappable
     */
    NonHotswappableReason["DEPENDENCY_UNSUPPORTED"] = "dependency-unsupported";
    /**
     * The resource type is not hotswappable
     */
    NonHotswappableReason["RESOURCE_UNSUPPORTED"] = "resource-unsupported";
    /**
     * The resource is created in the deployment
     */
    NonHotswappableReason["RESOURCE_CREATION"] = "resource-creation";
    /**
     * The resource is removed in the deployment
     */
    NonHotswappableReason["RESOURCE_DELETION"] = "resource-deletion";
    /**
     * The resource identified by the logical id has its type changed
     */
    NonHotswappableReason["RESOURCE_TYPE_CHANGED"] = "resource-type-changed";
    /**
     * The nested stack is created in the deployment
     */
    NonHotswappableReason["NESTED_STACK_CREATION"] = "nested-stack-creation";
})(NonHotswappableReason || (exports.NonHotswappableReason = NonHotswappableReason = {}));
