// @aws-amplify/backend-cli/lib/backend-identifier/backend_identifier_resolver.js
import { BackendIdentifierConversions } from '@aws-amplify/platform-core';
/**
 * Translates args to BackendIdentifier.
 * Throws if translation can't be made (this should never happen if command validation works correctly).
 */
export class AppBackendIdentifierResolver {
    namespaceResolver;
    /**
     * Instantiates BackendIdentifierResolver
     */
    constructor(namespaceResolver) {
        this.namespaceResolver = namespaceResolver;
    }
    resolveDeployedBackendIdentifier = async (args) => {
        if (args.stack) {
            return { stackName: args.stack };
        }
        else if (args.appId && args.branch) {
            return {
                namespace: args.appId,
                name: args.branch,
                type: 'branch',
            };
        }
        else if (args.branch) {
            return {
                appName: await this.namespaceResolver.resolve(),
                branchName: args.branch,
            };
        }
        return undefined;
    };
    resolveBackendIdentifier = async (args) => {
        if (args.stack) {
            return BackendIdentifierConversions.fromStackName(args.stack);
        }
        else if (args.appId && args.branch) {
            return {
                namespace: args.appId,
                name: args.branch,
                type: 'branch',
            };
        }
        else if (args.branch) {
            return {
                namespace: await this.namespaceResolver.resolve(),
                name: args.branch,
                type: 'branch',
            };
        }
        return undefined;
    };
}
