// @aws-amplify/graphql-conversation-transformer/lib/resolvers/assistant-response-stream-pipeline-definition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.assistantResponseStreamPipelineDefinition = void 0;
const graphql_transformer_core_1 = require("@aws-amplify/graphql-transformer-core");
const resolver_function_definition_1 = require("./resolver-function-definition");
const graphql_transformer_common_1 = require("graphql-transformer-common");
exports.assistantResponseStreamPipelineDefinition = {
    requestSlots: [init(), auth(), verifySessionOwner()],
    dataSlot: data(),
    responseSlots: [],
    field: (config) => ({ typeName: 'Mutation', fieldName: fieldName(config) }),
};
function init() {
    return (0, resolver_function_definition_1.createResolverFunctionDefinition)({
        slotName: 'init',
        fileName: 'init-resolver-fn.template.js',
        generateTemplate: (_, code) => graphql_transformer_core_1.MappingTemplate.inlineTemplateFromString(code),
        substitutions: (_, ctx) => ({
            GRAPHQL_API_ENDPOINT: ctx.api.graphqlUrl,
        }),
    });
}
function auth() {
    return (0, resolver_function_definition_1.createResolverFunctionDefinition)({
        slotName: 'auth',
        fileName: 'auth-resolver-fn.template.js',
        generateTemplate: templateGenerator('auth'),
    });
}
function verifySessionOwner() {
    return (0, resolver_function_definition_1.createResolverFunctionDefinition)({
        slotName: 'verifySessionOwner',
        fileName: 'verify-session-owner-resolver-fn.template.js',
        generateTemplate: templateGenerator('verify-session-owner'),
        dataSource: (config) => config.dataSources.conversationTableDataSource,
        substitutions: () => ({
            CONVERSATION_ID_PARENT: 'ctx.args.input',
        }),
    });
}
function data() {
    return (0, resolver_function_definition_1.createResolverFunctionDefinition)({
        slotName: 'data',
        fileName: 'assistant-streaming-mutation-resolver-fn.template.js',
        generateTemplate: templateGenerator('persist-message'),
        dataSource: (config) => config.dataSources.messageTableDataSource,
        substitutions: (config) => ({
            CONVERSATION_MESSAGE_TYPE_NAME: `ConversationMessage${(0, graphql_transformer_common_1.toUpper)(config.field.name.value)}`,
        }),
    });
}
function fieldName(config) {
    return config.assistantResponseStreamingMutation.field.name.value;
}
function templateGenerator(slotName) {
    return (0, resolver_function_definition_1.createS3AssetMappingTemplateGenerator)('Mutation', slotName, fieldName);
}
