// @aws-amplify/graphql-conversation-transformer/lib/resolvers/list-conversations-init-resolver-definition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listConversationsInitFunctionDefinition = void 0;
const name_values_1 = require("../graphql-types/name-values");
const resolver_function_definition_1 = require("./resolver-function-definition");
exports.listConversationsInitFunctionDefinition = (0, resolver_function_definition_1.createResolverFunctionDefinition)({
    slotName: 'init',
    fileName: 'list-conversations-init-set-index-resolver-fn.template.js',
    generateTemplate: (0, resolver_function_definition_1.createS3AssetMappingTemplateGenerator)('Query', 'list-conversations-init', (config) => config.field.name.value),
    substitutions: (config) => ({
        INDEX_NAME: name_values_1.LIST_CONVERSATIONS_INDEX_NAME,
        MODEL_QUERY_EXPRESSION: `{
      expression: '#typename = :typename',
      expressionNames: {
        '#typename': '__typename',
      },
      expressionValues: util.dynamodb.toMapValues({
        ':typename': '${config.conversation.model.name.value}',
      }),
    }`,
        SORT_DIRECTION: 'DESC',
    }),
});
