// @cdklabs/tskb/lib/relationship.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRelationshipCollection = exports.relationshipCollection = exports.NO_RELATIONSHIPS = void 0;
const NO_RELATIONSHIPS = () => ({});
exports.NO_RELATIONSHIPS = NO_RELATIONSHIPS;
function relationshipCollection(fromField, toField) {
    const forward = new Map();
    const backward = new Map();
    function add(fromId, toId, attrs) {
        let f = forward.get(fromId);
        if (!f) {
            f = [];
            forward.set(fromId, f);
        }
        let b = backward.get(toId);
        if (!b) {
            b = [];
            backward.set(toId, b);
        }
        // Behaves like a set, only add new relationship if it is structurally distinct
        const forwardRel = { $id: toId, ...attrs };
        const forwardRelStr = JSON.stringify(forwardRel);
        const existingRelationship = f.find((x) => JSON.stringify(x) === forwardRelStr);
        if (!existingRelationship) {
            f.push(forwardRel);
            b.push({ $id: fromId, ...attrs });
        }
    }
    return {
        type: 'rel',
        fromColl: fromField,
        toColl: toField,
        forward,
        backward,
        add(from, to, attributes) {
            add(from.$id, to.$id, attributes);
        },
        dehydrate() {
            return {
                type: 'rel',
                forward: Object.fromEntries(forward.entries()),
            };
        },
        hydrateFrom(x) {
            forward.clear();
            backward.clear();
            for (const [fromId, targets] of Object.entries(x.forward)) {
                for (const target of targets) {
                    add(fromId, target.$id, removeId(target));
                }
            }
        },
    };
}
exports.relationshipCollection = relationshipCollection;
function isRelationshipCollection(x) {
    return typeof x === 'object' && !!x && x.type === 'rel';
}
exports.isRelationshipCollection = isRelationshipCollection;
function removeId(x) {
    const ret = { ...x };
    delete ret.$id;
    return ret;
}
